{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ccb9eb27-ea38-419b-a434-29059ecb0963",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tbellagio/.local/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.26.4\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import os\n",
    "import pickle\n",
    "import dask.dataframe as dd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import random\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.metrics import r2_score\n",
    "from scipy.stats import spearmanr\n",
    "import numpy as np\n",
    "import cvxpy as cp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb8fe8bc-59ff-4cd3-a204-ebde582e0098",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "75e5bc84-8745-41df-a29a-f70cb08eb9d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## bring the founder \n",
    "af_founder = pd.read_csv('../leave_1_out/af_founder.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dd1e97e0-9bc2-41df-a20f-28c9bd87ab3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "delta_p_normed_file = 'delta_p_normed.csv'\n",
    "delta_p_normed = dd.read_csv(delta_p_normed_file, sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4cd845f5-59e5-4db8-86eb-e3ee1c7952c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "delta_p_normed = delta_p_normed.compute() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7bd6d385-90e6-469a-8122-5515f81f7692",
   "metadata": {},
   "outputs": [],
   "source": [
    "delta_p_normed = delta_p_normed.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "71a1e6e9-4f7c-46f3-927f-61534914a49b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_counts = pd.read_csv('../leave_1_out/genotype_counts_fullgenome231ecotypes.csv').drop('Unnamed: 0', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fd90732f-7a27-43ce-b2cd-c57006330d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_ecotypes = '/carnegie/nobackup/scratch/xwu/grenet/merged_frequency/merged_ecotype_frequency.txt'\n",
    "real_ef = pd.read_csv(path_ecotypes, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9d0b4081-6000-465e-bf2b-82d1c140728b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('splits_samples.pkl', 'rb') as file:\n",
    "    splits_samples = pickle.load(file)\n",
    "\n",
    "len(splits_samples) ## 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e621ff8-b97b-4798-8b80-f4780c320aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_files(directory, filename):\n",
    "    \"\"\"\n",
    "    Search for files with a specific name within all subdirectories of a given directory.\n",
    "\n",
    "    Parameters:\n",
    "        directory (str): The root directory from which to start the search.\n",
    "        filename (str): The name of the file to search for.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of full paths to the files that match the given filename.\n",
    "    \"\"\"\n",
    "    matches = []  # List to store the full paths of matching files\n",
    "    for root, dirs, files in os.walk(directory):\n",
    "        if filename in files:\n",
    "            full_path = os.path.join(root, filename)\n",
    "            matches.append(full_path)\n",
    "    return matches\n",
    "\n",
    "# Usage example\n",
    "root_directory = '/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/'  # Update this to the directory you want to search\n",
    "target_filename = 'predicted_allele_freq_bio1.csv'\n",
    "predicted_delta_p_normed_bio1_files = find_files(root_directory, target_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4c05c2c9-099d-4397-af8c-2878f0daaa2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Usage example\n",
    "root_directory = '/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/'  # Update this to the directory you want to search\n",
    "target_filename = 'snps_indices_bio1.csv'\n",
    "snps_indices_bio1_files = find_files(root_directory, target_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1138c516-0bc7-40f2-b92b-4f19dd5ed853",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_number = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3f631cac-20f5-4891-ab66-73a971e0dc13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_8/snps_indices_bio1.csv'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_indices_bio1_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a656853-fcfe-4cf2-9b00-ae05083b10d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "06f1aabc-5f27-4cf4-aa2e-25145e7146e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done0\n",
      "done1\n",
      "done2\n",
      "done3\n",
      "done4\n",
      "done5\n",
      "done6\n",
      "done7\n",
      "No snps split 8\n",
      "done9\n",
      "done10\n",
      "done11\n",
      "done12\n",
      "done13\n",
      "done14\n",
      "done15\n",
      "done16\n",
      "done17\n",
      "done18\n",
      "done19\n",
      "done20\n",
      "done21\n",
      "done22\n",
      "done23\n",
      "done24\n",
      "done25\n",
      "done26\n",
      "done27\n",
      "done28\n",
      "done29\n",
      "done30\n",
      "done31\n",
      "done32\n",
      "done33\n",
      "done34\n",
      "done35\n",
      "done36\n",
      "done37\n",
      "done38\n",
      "done39\n",
      "done40\n",
      "done41\n",
      "done42\n",
      "done43\n",
      "done44\n",
      "done45\n",
      "done46\n",
      "done47\n",
      "done48\n",
      "done49\n",
      "done50\n",
      "done51\n",
      "done52\n",
      "done53\n",
      "done54\n",
      "done55\n",
      "done56\n",
      "done57\n",
      "done58\n",
      "done59\n",
      "done60\n",
      "done61\n",
      "done62\n",
      "done63\n",
      "done64\n",
      "done65\n",
      "done66\n",
      "done67\n",
      "done68\n",
      "done69\n",
      "done70\n",
      "done71\n",
      "done72\n",
      "done73\n",
      "done74\n",
      "done75\n",
      "done76\n",
      "done77\n",
      "done78\n",
      "done79\n",
      "done80\n",
      "done81\n",
      "done82\n",
      "done83\n",
      "done84\n",
      "done85\n",
      "done86\n",
      "done87\n",
      "done88\n",
      "done89\n",
      "done90\n",
      "done91\n",
      "done92\n",
      "done93\n",
      "done94\n",
      "done95\n",
      "done96\n",
      "done97\n",
      "done98\n",
      "done99\n"
     ]
    }
   ],
   "source": [
    "for split_number in range(100):\n",
    "    predicted_delta_p_normed_bio1_file = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_{split_number}/predicted_allele_freq_bio1.csv'\n",
    "    snps_indices_bio1_file = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_{split_number}/snps_indices_bio1.csv'\n",
    "    clumping_file = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_{split_number}/output_clumping.clumped'\n",
    "    dict_rpos_index0_file = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_{split_number}/dict_rpos_index0.pkl'\n",
    "\n",
    "    output_file_af = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_{split_number}/r2_pred_allele_freq.csv' \n",
    "    output_file_ef = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_{split_number}/r2_pred_ecotype_freq.csv' \n",
    "    output_file_sp = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out_sv/split_number_{split_number}/spcorr_pred_ecotype_freq.csv' \n",
    "    \n",
    "    #split1 \n",
    "    train_samples = splits_samples[split_number][0]\n",
    "    test_samples = splits_samples[split_number][1]\n",
    "    \n",
    "    predicted_delta_p_normed_bio1 = pd.read_csv(predicted_delta_p_normed_bio1_file).drop('Unnamed: 0',axis=1)\n",
    "\n",
    "    snps_indices_bio1_old = pd.read_csv(snps_indices_bio1_file)\n",
    "\n",
    "\n",
    "    if len(snps_indices_bio1_old) == 0:\n",
    "        print(f'No snps split {split_number}')\n",
    "    else:  \n",
    "        \n",
    "        ## load the dictionary with the index0 and real snp id\n",
    "        with open(dict_rpos_index0_file, 'rb') as file:\n",
    "            dict_rpos_index0 = pickle.load(file)\n",
    "    \n",
    "        # load the result of clumping\n",
    "        clumping = pd.read_csv(clumping_file, delim_whitespace=True)\n",
    "        # use the dictionary to get a column with the index of the clumped snps \n",
    "        clumping['index0'] = clumping['SNP'].replace(dict_rpos_index0)\n",
    "        \n",
    "        snps_0index = clumping['index0']\n",
    "        snps_0index = snps_0index.to_list()\n",
    "        \n",
    "        genotype_counts_filt = genotype_counts.iloc[snps_0index, :]\n",
    "        \n",
    "        ## retrieving the old index to filter the allele freq prediciton\n",
    "        snps_1index = np.array(snps_0index) + 1\n",
    "        \n",
    "        snps_indices_bio1_old = snps_indices_bio1_old[snps_indices_bio1_old['x'].isin(snps_1index)]\n",
    "        \n",
    "        predicted_delta_p_normed_bio1.columns = predicted_delta_p_normed_bio1.columns.str.replace('V','').astype(int)\n",
    "        \n",
    "        snps_indices_bio1_old.columns = ['old_index1short', 'snps_indices_bio1_old_full']\n",
    "        \n",
    "        predicted_delta_p_normed_bio1 = predicted_delta_p_normed_bio1[snps_indices_bio1_old['old_index1short']]\n",
    "            \n",
    "        ## delta_p_normed train \n",
    "        delta_p_normed_train = delta_p_normed[train_samples]\n",
    "        #delta_p_normed_mean = delta_p_normed_train.mean(axis=1)\n",
    "        #mean_af_sel = delta_p_normed_mean.iloc[snps_0index,]\n",
    "        \n",
    "        ## allele freq test\n",
    "        delta_p_normed_test = delta_p_normed[test_samples]\n",
    "        \n",
    "        delta_p_normed_sel = delta_p_normed_test.iloc[snps_0index,:]\n",
    "        \n",
    "        delta_p_normed_sel = delta_p_normed_sel.sort_index()\n",
    "        \n",
    "        ## subract the mean to see what i get \n",
    "        #predicted_delta_p_normed_bio1 = predicted_delta_p_normed_bio1.sub(mean_af_sel.values,axis=1)\n",
    "        \n",
    "        predicted_delta_p_normed_bio1t = predicted_delta_p_normed_bio1.T\n",
    "        \n",
    "        predicted_delta_p_normed_bio1t.columns = delta_p_normed_sel.columns\n",
    "        \n",
    "        r2_af = {}\n",
    "        for i in delta_p_normed_sel.columns:\n",
    "            # Reshape the data to fit the model\n",
    "            X = delta_p_normed_sel[i].values.reshape(-1, 1)  # Independent variable\n",
    "            y = predicted_delta_p_normed_bio1t[i].values # Dependent variable\n",
    "        \n",
    "            # Fit the linear model\n",
    "            model = LinearRegression().fit(X, y)\n",
    "        \n",
    "            # Predict the values\n",
    "            y_pred = model.predict(X)\n",
    "        \n",
    "            # Calculate R^2\n",
    "            r_squared = r2_score(y, y_pred)\n",
    "            r2_af[i] = r_squared\n",
    "        \n",
    "        r2_af = pd.DataFrame(list(r2_af.items()), columns=['Index', 'Value'])\n",
    "        r2_af.columns = ['sample', 'r2'] \n",
    "        \n",
    "        r2_af.to_csv(output_file_af)\n",
    "        \n",
    "        \n",
    "        af_founder_sel = af_founder.iloc[snps_0index]\n",
    "        \n",
    "        af_founder_sel = np.array(af_founder_sel['af'])\n",
    "        \n",
    "        predicted_allele_freq_bio1 = (predicted_delta_p_normed_bio1.multiply(af_founder_sel, axis=1)) + af_founder_sel\n",
    "        \n",
    "        genotype_counts_array = genotype_counts_filt.to_numpy()\n",
    "        \n",
    "        pred_ecotypes_frequency = {}\n",
    "        for sample in range(predicted_allele_freq_bio1.shape[0]):\n",
    "            ## just select the frist sample\n",
    "            one_sample = predicted_allele_freq_bio1.iloc[sample,:]\n",
    "            \n",
    "            y = np.asarray(one_sample).T * 2\n",
    "            \n",
    "            h = cp.Variable(231)\n",
    "            \n",
    "            product = np.asarray(genotype_counts_array) @ h \n",
    "            \n",
    "            diff = product - y\n",
    "            \n",
    "            constraints = [0 <= h, sum(h) == 1]\n",
    "            \n",
    "            problem = cp.Problem(cp.Minimize(cp.norm(diff)), constraints)\n",
    "            \n",
    "            # Solve the optimization problem using the SCS solver, with verbose output disabled.\n",
    "            problem.solve(solver='SCS', verbose=False,eps=1e-6)\n",
    "            \n",
    "            pred_ecotypes_frequency[sample] = h.value\n",
    "        \n",
    "        \n",
    "        ## get the real ecotype freq \n",
    "        \n",
    "        real_ef_test = real_ef[test_samples]\n",
    "        \n",
    "        pred_ecotypes_frequency = pd.DataFrame(pred_ecotypes_frequency)\n",
    "        \n",
    "        pred_ecotypes_frequency.columns = real_ef_test.columns\n",
    "        \n",
    "        r2_ef = {}\n",
    "        spcorrelation_ef = {}\n",
    "        \n",
    "        for i in real_ef_test.columns:\n",
    "        \n",
    "            # Get the ranks of the data\n",
    "            X_ranked = real_ef_test[i].sort_values().reset_index()['index'] #.reshape(-1, 1)  # Independent variable\n",
    "            y_ranked = pred_ecotypes_frequency[i].sort_values().reset_index()['index']  # Dependent variable\n",
    "        \n",
    "            # Compute Spearman's rank correlation\n",
    "            correlation, _ = spearmanr(X_ranked, y_ranked)\n",
    "            spcorrelation_ef[i] = correlation\n",
    "        \n",
    "            # Reshape the data to fit the model\n",
    "            X = real_ef_test[i].values.reshape(-1, 1)  # Independent variable\n",
    "            y = pred_ecotypes_frequency[i].values  # Dependent variable\n",
    "        \n",
    "            # Fit the linear model\n",
    "            model = LinearRegression().fit(X, y)\n",
    "        \n",
    "            # Predict the values\n",
    "            y_pred = model.predict(X)\n",
    "        \n",
    "            # Calculate R^2\n",
    "            r_squared = r2_score(y, y_pred)\n",
    "            r2_ef[i] = r_squared\n",
    "        \n",
    "        \n",
    "        spcorrelation_ef = pd.DataFrame(list(spcorrelation_ef.items()), columns=['Index', 'Value'])\n",
    "        spcorrelation_ef.columns = ['sample', 'r2'] \n",
    "        \n",
    "        r2_ef = pd.DataFrame(list(r2_ef.items()), columns=['Index', 'Value'])\n",
    "        r2_ef.columns = ['sample', 'r2'] \n",
    "        \n",
    "        spcorrelation_ef.to_csv(output_file_sp)\n",
    "        r2_ef.to_csv(output_file_ef)\n",
    "        \n",
    "        print(f'done{split_number}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d07bac49-4a0f-4d83-b451-75ede9f51337",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cfa258e-bba4-4481-8218-34ccd70de5ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56ecfbdd-9cde-4e17-bc06-e177676ffc9c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d80cd43-f2df-4f90-9f1b-40f248f83c49",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3940c1c7-788a-4d37-82e5-885142bb465b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f508ff22-0a67-45c3-89c0-cd834d38e904",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        V1        V2        V3        V4        V5        V6        V7  \\\n",
      "0  0.06982  0.055864  0.049478  0.041729  0.585285  0.024016  0.003325   \n",
      "1  0.06982  0.055864  0.049478  0.041729  0.585285  0.024016  0.003325   \n",
      "2  0.06982  0.055864  0.049478  0.041729  0.585285  0.024016  0.003325   \n",
      "3  0.06982  0.055864  0.049478  0.041729  0.585285  0.024016  0.003325   \n",
      "4  0.06982  0.055864  0.049478  0.041729  0.585285  0.024016  0.003325   \n",
      "\n",
      "         V8        V9       V10  ...      V673      V674      V675      V676  \\\n",
      "0  0.014479  0.656701  0.638682  ...  0.144132  0.524631  0.524631  0.524631   \n",
      "1  0.014479  0.656701  0.638682  ...  0.144132  0.524631  0.524631  0.524631   \n",
      "2  0.014479  0.656701  0.638682  ...  0.144132  0.524631  0.524631  0.524631   \n",
      "3  0.014479  0.656701  0.638682  ...  0.144132  0.524631  0.524631  0.524631   \n",
      "4  0.014479  0.656701  0.638682  ...  0.144132  0.524631  0.524631  0.524631   \n",
      "\n",
      "       V677      V678      V679      V680      V681      V682  \n",
      "0  0.172696  0.155313  0.370611  0.309363  0.089572  0.054044  \n",
      "1  0.172696  0.155313  0.370611  0.309363  0.089572  0.054044  \n",
      "2  0.172696  0.155313  0.370611  0.309363  0.089572  0.054044  \n",
      "3  0.172696  0.155313  0.370611  0.309363  0.089572  0.054044  \n",
      "4  0.172696  0.155313  0.370611  0.309363  0.089572  0.054044  \n",
      "\n",
      "[5 rows x 682 columns]\n",
      "       0  1  2  3  4  5  6  7  8  9  ...  221  222  223  224  225  226  227  \\\n",
      "4163   0  0  0  0  0  0  2  0  0  0  ...    0    0    0    0    0    0    0   \n",
      "4168   0  0  0  0  0  0  2  0  0  0  ...    0    0    0    0    0    0    0   \n",
      "19041  0  0  2  0  0  0  0  0  0  2  ...    0    0    0    0    0    0    0   \n",
      "19828  0  0  0  0  0  0  0  0  0  2  ...    0    0    0    0    0    0    0   \n",
      "19866  2  0  0  2  2  2  0  0  2  0  ...    0    2    0    0    0    0    0   \n",
      "\n",
      "       228  229  230  \n",
      "4163     0    0    0  \n",
      "4168     0    0    0  \n",
      "19041    0    0    0  \n",
      "19828    0    0    0  \n",
      "19866    2    0    2  \n",
      "\n",
      "[5 rows x 231 columns]\n",
      "done0\n",
      "         V1        V2        V3        V4        V5        V6        V7  \\\n",
      "0  0.067783  0.067783  0.074733  0.104345  0.081487  0.589872  0.192696   \n",
      "1  0.067783  0.067783  0.074733  0.104345  0.081487  0.589872  0.192696   \n",
      "2  0.067783  0.067783  0.074733  0.104345  0.081487  0.589872  0.192696   \n",
      "3  0.067783  0.067783  0.074733  0.104345  0.081487  0.589872  0.192696   \n",
      "4  0.067783  0.067783  0.074733  0.104345  0.081487  0.589872  0.192696   \n",
      "\n",
      "         V8        V9       V10  ...     V2887     V2888     V2889     V2890  \\\n",
      "0  0.195251  0.486142  0.342619  ...  0.074728  0.028341  0.029955  0.029955   \n",
      "1  0.195251  0.486142  0.342619  ...  0.074728  0.028341  0.029955  0.029955   \n",
      "2  0.195251  0.486142  0.342619  ...  0.074728  0.028341  0.029955  0.029955   \n",
      "3  0.195251  0.486142  0.342619  ...  0.074728  0.028341  0.029955  0.029955   \n",
      "4  0.195251  0.486142  0.342619  ...  0.074728  0.028341  0.029955  0.029955   \n",
      "\n",
      "      V2891     V2892     V2893     V2894     V2895     V2896  \n",
      "0  0.029955  0.029955  0.028341  0.038363  0.028341  0.028341  \n",
      "1  0.029955  0.029955  0.028341  0.038363  0.028341  0.028341  \n",
      "2  0.029955  0.029955  0.028341  0.038363  0.028341  0.028341  \n",
      "3  0.029955  0.029955  0.028341  0.038363  0.028341  0.028341  \n",
      "4  0.029955  0.029955  0.028341  0.038363  0.028341  0.028341  \n",
      "\n",
      "[5 rows x 2896 columns]\n",
      "      0  1  2  3  4  5  6  7  8  9  ...  221  222  223  224  225  226  227  \\\n",
      "1268  0  0  0  0  0  0  2  0  0  0  ...    0    0    0    0    0    0    0   \n",
      "1617  0  0  0  0  0  0  2  0  0  0  ...    0    0    0    0    0    0    0   \n",
      "1819  0  0  0  0  0  0  2  0  0  0  ...    0    0    0    0    0    0    0   \n",
      "4163  0  0  0  0  0  0  2  0  0  0  ...    0    0    0    0    0    0    0   \n",
      "4168  0  0  0  0  0  0  2  0  0  0  ...    0    0    0    0    0    0    0   \n",
      "\n",
      "      228  229  230  \n",
      "1268    0    0    0  \n",
      "1617    0    0    0  \n",
      "1819    0    0    0  \n",
      "4163    0    0    0  \n",
      "4168    0    0    0  \n",
      "\n",
      "[5 rows x 231 columns]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 44\u001b[0m\n\u001b[1;32m     41\u001b[0m     problem \u001b[38;5;241m=\u001b[39m cp\u001b[38;5;241m.\u001b[39mProblem(cp\u001b[38;5;241m.\u001b[39mMinimize(cp\u001b[38;5;241m.\u001b[39mnorm(diff)), constraints)\n\u001b[1;32m     43\u001b[0m     \u001b[38;5;66;03m# Solve the optimization problem using the SCS solver, with verbose output disabled.\u001b[39;00m\n\u001b[0;32m---> 44\u001b[0m     \u001b[43mproblem\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msolve\u001b[49m\u001b[43m(\u001b[49m\u001b[43msolver\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mSCS\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43meps\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1e-6\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     46\u001b[0m     pred_ecotypes_frequency[sample] \u001b[38;5;241m=\u001b[39m h\u001b[38;5;241m.\u001b[39mvalue\n\u001b[1;32m     49\u001b[0m \u001b[38;5;66;03m## get the real ecotype freq \u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/problems/problem.py:503\u001b[0m, in \u001b[0;36mProblem.solve\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    501\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    502\u001b[0m     solve_func \u001b[38;5;241m=\u001b[39m Problem\u001b[38;5;241m.\u001b[39m_solve\n\u001b[0;32m--> 503\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msolve_func\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/problems/problem.py:1073\u001b[0m, in \u001b[0;36mProblem._solve\u001b[0;34m(self, solver, warm_start, verbose, gp, qcp, requires_grad, enforce_dpp, ignore_dpp, canon_backend, **kwargs)\u001b[0m\n\u001b[1;32m   1070\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39munpack(chain\u001b[38;5;241m.\u001b[39mretrieve(soln))\n\u001b[1;32m   1071\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalue\n\u001b[0;32m-> 1073\u001b[0m data, solving_chain, inverse_data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_problem_data\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1074\u001b[0m \u001b[43m    \u001b[49m\u001b[43msolver\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43menforce_dpp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mignore_dpp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcanon_backend\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m   1075\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1077\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m verbose:\n\u001b[1;32m   1078\u001b[0m     \u001b[38;5;28mprint\u001b[39m(_NUM_SOLVER_STR)\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/problems/problem.py:696\u001b[0m, in \u001b[0;36mProblem.get_problem_data\u001b[0;34m(self, solver, gp, enforce_dpp, ignore_dpp, verbose, canon_backend, solver_opts)\u001b[0m\n\u001b[1;32m    693\u001b[0m     s\u001b[38;5;241m.\u001b[39mLOGGER\u001b[38;5;241m.\u001b[39minfo(\n\u001b[1;32m    694\u001b[0m              \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCompiling problem (target solver=\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m).\u001b[39m\u001b[38;5;124m'\u001b[39m, solver_name)\n\u001b[1;32m    695\u001b[0m     s\u001b[38;5;241m.\u001b[39mLOGGER\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReduction chain: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m, reduction_chain_str)\n\u001b[0;32m--> 696\u001b[0m data, inverse_data \u001b[38;5;241m=\u001b[39m \u001b[43msolving_chain\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    697\u001b[0m safe_to_cache \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    698\u001b[0m     \u001b[38;5;28misinstance\u001b[39m(data, \u001b[38;5;28mdict\u001b[39m)\n\u001b[1;32m    699\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m s\u001b[38;5;241m.\u001b[39mPARAM_PROB \u001b[38;5;129;01min\u001b[39;00m data\n\u001b[1;32m    700\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(reduction, EvalParams)\n\u001b[1;32m    701\u001b[0m                 \u001b[38;5;28;01mfor\u001b[39;00m reduction \u001b[38;5;129;01min\u001b[39;00m solving_chain\u001b[38;5;241m.\u001b[39mreductions)\n\u001b[1;32m    702\u001b[0m )\n\u001b[1;32m    703\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compilation_time \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/reductions/chain.py:76\u001b[0m, in \u001b[0;36mChain.apply\u001b[0;34m(self, problem, verbose)\u001b[0m\n\u001b[1;32m     74\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m verbose:\n\u001b[1;32m     75\u001b[0m         s\u001b[38;5;241m.\u001b[39mLOGGER\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mApplying reduction \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28mtype\u001b[39m(r)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m)\n\u001b[0;32m---> 76\u001b[0m     problem, inv \u001b[38;5;241m=\u001b[39m \u001b[43mr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mproblem\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     77\u001b[0m     inverse_data\u001b[38;5;241m.\u001b[39mappend(inv)\n\u001b[1;32m     78\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m problem, inverse_data\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/reductions/dcp2cone/cone_matrix_stuffing.py:379\u001b[0m, in \u001b[0;36mConeMatrixStuffing.apply\u001b[0;34m(self, problem)\u001b[0m\n\u001b[1;32m    377\u001b[0m \u001b[38;5;66;03m# Batch expressions together, then split apart.\u001b[39;00m\n\u001b[1;32m    378\u001b[0m expr_list \u001b[38;5;241m=\u001b[39m [arg \u001b[38;5;28;01mfor\u001b[39;00m c \u001b[38;5;129;01min\u001b[39;00m ordered_cons \u001b[38;5;28;01mfor\u001b[39;00m arg \u001b[38;5;129;01min\u001b[39;00m c\u001b[38;5;241m.\u001b[39margs]\n\u001b[0;32m--> 379\u001b[0m params_to_problem_data \u001b[38;5;241m=\u001b[39m \u001b[43mextractor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43maffine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexpr_list\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    381\u001b[0m inverse_data\u001b[38;5;241m.\u001b[39mminimize \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mtype\u001b[39m(problem\u001b[38;5;241m.\u001b[39mobjective) \u001b[38;5;241m==\u001b[39m Minimize\n\u001b[1;32m    382\u001b[0m variables \u001b[38;5;241m=\u001b[39m problem\u001b[38;5;241m.\u001b[39mvariables()\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/utilities/coeff_extractor.py:72\u001b[0m, in \u001b[0;36mCoeffExtractor.affine\u001b[0;34m(self, expr)\u001b[0m\n\u001b[1;32m     70\u001b[0m num_rows \u001b[38;5;241m=\u001b[39m \u001b[38;5;28msum\u001b[39m([e\u001b[38;5;241m.\u001b[39msize \u001b[38;5;28;01mfor\u001b[39;00m e \u001b[38;5;129;01min\u001b[39;00m expr_list])\n\u001b[1;32m     71\u001b[0m op_list \u001b[38;5;241m=\u001b[39m [e\u001b[38;5;241m.\u001b[39mcanonical_form[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m e \u001b[38;5;129;01min\u001b[39;00m expr_list]\n\u001b[0;32m---> 72\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mcanonInterface\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_problem_matrix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mop_list\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     73\u001b[0m \u001b[43m                                         \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mx_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     74\u001b[0m \u001b[43m                                         \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mid_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     75\u001b[0m \u001b[43m                                         \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparam_to_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     76\u001b[0m \u001b[43m                                         \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparam_id_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     77\u001b[0m \u001b[43m                                         \u001b[49m\u001b[43mnum_rows\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     78\u001b[0m \u001b[43m                                         \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcanon_backend\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/cvxcore/python/canonInterface.py:297\u001b[0m, in \u001b[0;36mget_problem_matrix\u001b[0;34m(linOps, var_length, id_to_col, param_to_size, param_to_col, constr_length, canon_backend)\u001b[0m\n\u001b[1;32m    295\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m canon_backend \u001b[38;5;241m==\u001b[39m s\u001b[38;5;241m.\u001b[39mCPP_CANON_BACKEND:\n\u001b[1;32m    296\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcvxpy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcvxcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpython\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcppbackend\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m build_matrix\n\u001b[0;32m--> 297\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mbuild_matrix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mid_to_col\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam_to_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam_to_col\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvar_length\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconstr_length\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlinOps\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    299\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m canon_backend \u001b[38;5;129;01min\u001b[39;00m {s\u001b[38;5;241m.\u001b[39mSCIPY_CANON_BACKEND, s\u001b[38;5;241m.\u001b[39mRUST_CANON_BACKEND,\n\u001b[1;32m    300\u001b[0m                        s\u001b[38;5;241m.\u001b[39mNUMPY_CANON_BACKEND}:\n\u001b[1;32m    301\u001b[0m     param_size_plus_one \u001b[38;5;241m=\u001b[39m \u001b[38;5;28msum\u001b[39m(param_to_size\u001b[38;5;241m.\u001b[39mvalues())\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/cvxcore/python/cppbackend.py:56\u001b[0m, in \u001b[0;36mbuild_matrix\u001b[0;34m(id_to_col, param_to_size, param_to_col, var_length, constr_length, linOps)\u001b[0m\n\u001b[1;32m     54\u001b[0m linPy_to_linC \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m     55\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m lin \u001b[38;5;129;01min\u001b[39;00m linOps:\n\u001b[0;32m---> 56\u001b[0m     \u001b[43mbuild_lin_op_tree\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlin\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlinPy_to_linC\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     57\u001b[0m     tree \u001b[38;5;241m=\u001b[39m linPy_to_linC[lin]\n\u001b[1;32m     58\u001b[0m     lin_vec\u001b[38;5;241m.\u001b[39mpush_back(tree)\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/cvxcore/python/cppbackend.py:225\u001b[0m, in \u001b[0;36mbuild_lin_op_tree\u001b[0;34m(root_linPy, linPy_to_linC)\u001b[0m\n\u001b[1;32m    223\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m post_order_stack:\n\u001b[1;32m    224\u001b[0m     linPy \u001b[38;5;241m=\u001b[39m post_order_stack\u001b[38;5;241m.\u001b[39mpop()\n\u001b[0;32m--> 225\u001b[0m     \u001b[43mmake_linC_from_linPy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlinPy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlinPy_to_linC\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/cvxcore/python/cppbackend.py:175\u001b[0m, in \u001b[0;36mmake_linC_from_linPy\u001b[0;34m(linPy, linPy_to_linC)\u001b[0m\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m argPy \u001b[38;5;129;01min\u001b[39;00m linPy\u001b[38;5;241m.\u001b[39margs:\n\u001b[1;32m    174\u001b[0m     lin_args_vec\u001b[38;5;241m.\u001b[39mpush_back(linPy_to_linC[argPy])\n\u001b[0;32m--> 175\u001b[0m linC \u001b[38;5;241m=\u001b[39m \u001b[43mcvxcore\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mLinOp\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlin_args_vec\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    176\u001b[0m linPy_to_linC[linPy] \u001b[38;5;241m=\u001b[39m linC\n\u001b[1;32m    178\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m linPy\u001b[38;5;241m.\u001b[39mdata \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/cvxpy/cvxcore/python/cvxcore.py:153\u001b[0m, in \u001b[0;36mLinOp.__init__\u001b[0;34m(self, type, shape, args)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28mtype\u001b[39m, shape, args):\n\u001b[0;32m--> 153\u001b[0m     \u001b[43m_cvxcore\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mLinOp_swiginit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_cvxcore\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnew_LinOp\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for split_number in range(100):\n",
    "    predicted_allele_freq_bio1_file = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out/split_number_{split_number}/predicted_allele_freq_bio1.csv'\n",
    "    snps_indices_bio1_file = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out/split_number_{split_number}/snps_indices_bio1.csv'\n",
    "    output_file_ef = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out/split_number_{split_number}/r2_pred_ecotype_freq_wmean.csv' \n",
    "    output_file_af = f'/carnegie/nobackup/scratch/tbellagio/gea_grene-net/leave_1_out/split_number_{split_number}/r2_pred_allele_freq_wmean.csv' \n",
    "\n",
    "    \n",
    "    #split1 \n",
    "    train_samples = splits_samples[split_number][0]\n",
    "    test_samples = splits_samples[split_number][1]\n",
    "    \n",
    "    predicted_allele_freq_bio1 = pd.read_csv(predicted_allele_freq_bio1_file).drop('Unnamed: 0',axis=1)\n",
    "    \n",
    "    snps_indices_bio1 = pd.read_csv(snps_indices_bio1_file)\n",
    "    print(predicted_allele_freq_bio1.head())    \n",
    "    snps_0index = snps_indices_bio1['x'] - 1 \n",
    "    \n",
    "    snps_0index = snps_0index.to_list()\n",
    "    \n",
    "    genotype_counts_filt = genotype_counts.iloc[snps_0index, :]\n",
    "    print(genotype_counts_filt.head())\n",
    "    genotype_counts_array = genotype_counts_filt.to_numpy()\n",
    "    \n",
    "    pred_ecotypes_frequency = {}\n",
    "    for sample in range(predicted_allele_freq_bio1.shape[0]):\n",
    "        ## just select the frist sample\n",
    "        one_sample = predicted_allele_freq_bio1.iloc[sample,:]\n",
    "        \n",
    "        y = np.asarray(one_sample).T * 2\n",
    "        \n",
    "        h = cp.Variable(231)\n",
    "    \n",
    "        genotype_counts_array = genotype_counts_filt.to_numpy()\n",
    "        \n",
    "        product = np.asarray(genotype_counts_array) @ h \n",
    "        \n",
    "        diff = product - y\n",
    "        \n",
    "        constraints = [0 <= h, sum(h) == 1]\n",
    "        \n",
    "        problem = cp.Problem(cp.Minimize(cp.norm(diff)), constraints)\n",
    "        \n",
    "        # Solve the optimization problem using the SCS solver, with verbose output disabled.\n",
    "        problem.solve(solver='SCS', verbose=False,eps=1e-6)\n",
    "        \n",
    "        pred_ecotypes_frequency[sample] = h.value\n",
    "    \n",
    "    \n",
    "    ## get the real ecotype freq \n",
    "    \n",
    "    real_ef_test = real_ef[test_samples]\n",
    "    \n",
    "    pred_ecotypes_frequency = pd.DataFrame(pred_ecotypes_frequency)\n",
    "    \n",
    "    pred_ecotypes_frequency.columns = real_ef_test.columns\n",
    "    \n",
    "    r2_ef = {}\n",
    "    \n",
    "    for i in real_ef_test.columns:\n",
    "        # Reshape the data to fit the model\n",
    "        X = real_ef_test[i].values.reshape(-1, 1)  # Independent variable\n",
    "        y = pred_ecotypes_frequency[i].values  # Dependent variable\n",
    "    \n",
    "        # Fit the linear model\n",
    "        model = LinearRegression().fit(X, y)\n",
    "    \n",
    "        # Predict the values\n",
    "        y_pred = model.predict(X)\n",
    "    \n",
    "        # Calculate R^2\n",
    "        r_squared = r2_score(y, y_pred)\n",
    "        r2_ef[i] = r_squared\n",
    "    \n",
    "    r2_ef = pd.DataFrame(list(r2_ef.items()), columns=['Index', 'Value'])\n",
    "    r2_ef.columns = ['sample', 'r2'] \n",
    "    \n",
    "    r2_ef.to_csv(output_file_ef)\n",
    "    \n",
    "    ## allele_freq train \n",
    "    allele_freq_train = allele_freq[train_samples]\n",
    "    allele_freq_mean = allele_freq_train.mean(axis=1)\n",
    "    #mean_af_sel = allele_freq_mean.iloc[snps_0index,]\n",
    "\n",
    "    ## allele freq test\n",
    "    allele_freq_test = allele_freq[test_samples]\n",
    "\n",
    "    allele_freq_sel = allele_freq_test.iloc[snps_0index,:]\n",
    "    \n",
    "    allele_freq_sel = allele_freq_sel.sort_index()\n",
    "    \n",
    "    ## subract the mean to see what i get \n",
    "    #predicted_allele_freq_bio1 = predicted_allele_freq_bio1.sub(mean_af_sel.values,axis=1)\n",
    "    \n",
    "    predicted_allele_freq_bio1t = predicted_allele_freq_bio1.T\n",
    "    \n",
    "    predicted_allele_freq_bio1t.columns = allele_freq_sel.columns\n",
    "    \n",
    "    r2_af = {}\n",
    "    for i in allele_freq_sel.columns:\n",
    "        # Reshape the data to fit the model\n",
    "        X = allele_freq_sel[i].values.reshape(-1, 1)  # Independent variable\n",
    "        y = predicted_allele_freq_bio1t[i].values # Dependent variable\n",
    "    \n",
    "        # Fit the linear model\n",
    "        model = LinearRegression().fit(X, y)\n",
    "    \n",
    "        # Predict the values\n",
    "        y_pred = model.predict(X)\n",
    "    \n",
    "        # Calculate R^2\n",
    "        r_squared = r2_score(y, y_pred)\n",
    "        r2_af[i] = r_squared\n",
    "    \n",
    "    r2_af = pd.DataFrame(list(r2_af.items()), columns=['Index', 'Value'])\n",
    "    r2_af.columns = ['sample', 'r2'] \n",
    "    \n",
    "    r2_af.to_csv(output_file_af)\n",
    "    print(f'done{split_number}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5946c4fd-218d-4ff2-bdb3-b93f998c50e0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (pipeline_snakemake)",
   "language": "python",
   "name": "pipeline_snakemake"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
