{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e586a58a-a8f4-4c5b-bce7-df30076468b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_443869/2430059226.py:4: DeprecationWarning: \n",
      "Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),\n",
      "(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)\n",
      "but was not found to be installed on your system.\n",
      "If this would cause problems for you,\n",
      "please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466\n",
      "        \n",
      "  import pandas as pd\n"
     ]
    }
   ],
   "source": [
    "## last gen \n",
    "from scipy.stats import spearmanr, pearsonr, kendalltau\n",
    "\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bddd1fbf-ba21-4363-bb54-f9803e19d2db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import allel\n",
    "#vcf = allel.read_vcf('../key_files/greneNet_final_v1.1_LDpruned.recode.vcf')\n",
    "#ecotypes_names = vcf['samples']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "97d26b1b-0a53-456a-8105-16fb738ee3a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.Series(ecotypes_names).to_csv('../key_files/founder_ecotype_names.csv', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "84fec678-a4d9-43ea-945b-70289f02653d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ecotypes_names = pd.read_csv('../key_files/founder_ecotype_names.csv')['0'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "42b99dac-dd83-4814-9ed5-b158cf8e598c",
   "metadata": {},
   "outputs": [],
   "source": [
    "clim_sites_during_exp = pd.read_csv('../key_files/bioclimvars_experimental_sites_era5.csv')\n",
    "#clim_sites_during_exp = pd.read_csv('../key_files/bioclimvars_sites_era5_year_2018.csv')\n",
    "\n",
    "\n",
    "#sites_test = pd.Series(test).str.split('_').str[0].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5f4d72fb-7504-4548-b32a-ddc0f02e82c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_sites = pd.read_csv('../key_files/final_gen.csv')['site'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5c27b6b2-c3e2-4356-896b-e5cb070afc3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort_bio1 = clim_sites_during_exp\n",
    "clim_sites_during_exp_s = clim_sites_during_exp[clim_sites_during_exp['site'].isin(unique_sites)]\n",
    "\n",
    "sort_bio1 = clim_sites_during_exp_s[['site', 'bio1']].sort_values('bio1')['site'].astype(str).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "84557870-c48c-42cb-b9a8-217a27c8455b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sort_bio1_int = clim_sites_during_exp_s[['site', 'bio1']].sort_values('bio1')['site'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5a245dff-9fc3-4023-92dc-4582f54816c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_gen = pd.read_csv('../key_files/final_gen.csv')['sample_name'].to_list()\n",
    "delta_ef=pd.read_csv('../key_files/delta_ecotype_freq.txt', sep = '\\t', usecols = final_gen)#[['log_p1_p0','ecotype', 'site','plot']]\n",
    "\n",
    "ef=pd.read_csv('../key_files/merged_ecotype_frequency.txt', sep = '\\t', usecols = final_gen)#[['log_p1_p0','ecotype', 'site','plot']]\n",
    "delta_ef.index = ecotypes_names\n",
    "ef.index = ecotypes_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cb4259a0-58aa-403c-b470-8e0e5e6c7207",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_sites = pd.read_csv('../key_files/final_gen.csv')['site'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "eafe64e6-d5b9-4a04-a419-6a34ba2f075a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 1,  2,  4,  5,  6,  9, 10, 11, 12, 13, 23, 24, 25, 26, 27, 28, 32,\n",
       "       33, 37, 42, 43, 45, 46, 48, 49, 52, 53, 54, 55, 57, 60])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "15bbed32-bf4a-49ac-8ec9-e5ef9a9550b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '../jacknife_lfmm/splits_samples_last_gen_no33.pkl'\n",
    "\n",
    "# Open and load the .pkl file\n",
    "with open(file_path, 'rb') as file:\n",
    "    samples = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "945edb99-c1ab-4358-861e-d526ecfceee1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d28920fc-4c64-4327-93ee-4fedd5b70881",
   "metadata": {},
   "outputs": [],
   "source": [
    "## recalculate GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e7c45424-7eb5-4b3f-bf43-5fbe5d04bad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_counts = pd.read_csv('../leave_1_out/genotype_counts_fullgenome231ecotypes.csv').drop('Unnamed: 0', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "da5ebd95-93c4-4c05-88d8-c367f50b47dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_dict = pd.read_csv('../key_files/var_pos_grenenet.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c2edb50f-71a6-4f77-b000-12b940c63d0f",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'split' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[21], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43msplit\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'split' is not defined"
     ]
    }
   ],
   "source": [
    "split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a671f4-271c-4f5e-a56a-dc8b4c9999cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bce724ba-28f9-49dc-acc7-2989fa4ce246",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "fa1d3ff8-714e-4de7-ada1-93d2cb14c7e8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0_1': [1048635, 1048634],\n",
       " '1_1': [1048635, 1048634],\n",
       " '2_1': [1048635, 1048634],\n",
       " '3_1': [1048635, 1048633],\n",
       " '4_1': [1048635, 1048634],\n",
       " '5_1': [1048635, 1048634],\n",
       " '6_1': [1048635, 1048634],\n",
       " '7_1': [1048635, 1048634],\n",
       " '8_1': [1048635, 1048634],\n",
       " '9_1': [1048635, 1048634],\n",
       " '10_1': [1048635, 1048634],\n",
       " '11_1': [1048635, 1048634],\n",
       " '12_1': [1048635, 1048634],\n",
       " '13_1': [1048635, 1048634],\n",
       " '14_1': [1048635, 1048634],\n",
       " '15_1': [1048635, 1048633],\n",
       " '16_1': [1048635, 1048634],\n",
       " '17_1': [1048635, 1048624],\n",
       " '18_1': [1048635, 1048634],\n",
       " '19_1': [1048635, 1048634],\n",
       " '20_1': [1048635, 1048634],\n",
       " '21_1': [1048635, 1048634],\n",
       " '22_1': [1048635, 1048631],\n",
       " '23_1': [1048635, 1048634],\n",
       " '24_1': [1048635, 1048634],\n",
       " '25_1': [1048635, 1048634],\n",
       " '26_1': [1048635, 1048634],\n",
       " '27_1': [1048635, 1048633],\n",
       " '28_1': [1048635, 1048634],\n",
       " '29_1': [1048635, 1048632],\n",
       " '0_0.8': [1048635, 838908],\n",
       " '1_0.8': [1048635, 838908],\n",
       " '2_0.8': [1048635, 838905],\n",
       " '3_0.8': [1048635, 838908],\n",
       " '4_0.8': [1048635, 838908],\n",
       " '5_0.8': [1048635, 838906],\n",
       " '6_0.8': [1048635, 838908],\n",
       " '7_0.8': [1048635, 838907],\n",
       " '8_0.8': [1048635, 838908],\n",
       " '9_0.8': [1048635, 838908],\n",
       " '10_0.8': [1048635, 838908],\n",
       " '11_0.8': [1048635, 838908],\n",
       " '12_0.8': [1048635, 838908],\n",
       " '13_0.8': [1048635, 838908],\n",
       " '14_0.8': [1048635, 838906],\n",
       " '15_0.8': [1048635, 838908],\n",
       " '16_0.8': [1048635, 838892],\n",
       " '17_0.8': [1048635, 838907],\n",
       " '18_0.8': [1048635, 838907],\n",
       " '19_0.8': [1048635, 838784],\n",
       " '20_0.8': [1048635, 838896],\n",
       " '21_0.8': [1048635, 838908],\n",
       " '22_0.8': [1048635, 838907],\n",
       " '23_0.8': [1048635, 838908],\n",
       " '24_0.8': [1048635, 838907],\n",
       " '25_0.8': [1048635, 838908],\n",
       " '26_0.8': [1048635, 838907],\n",
       " '27_0.8': [1048635, 838901],\n",
       " '28_0.8': [1048635, 838906],\n",
       " '29_0.8': [1048635, 838878],\n",
       " '0_0.5': [1048635, 524317],\n",
       " '1_0.5': [1048635, 524317],\n",
       " '2_0.5': [1048635, 524307],\n",
       " '3_0.5': [1048635, 524317],\n",
       " '4_0.5': [1048635, 524317],\n",
       " '5_0.5': [1048635, 524317],\n",
       " '6_0.5': [1048635, 524305],\n",
       " '7_0.5': [1048635, 524317],\n",
       " '8_0.5': [1048635, 524317],\n",
       " '9_0.5': [1048635, 524248],\n",
       " '10_0.5': [1048635, 524317],\n",
       " '11_0.5': [1048635, 524316],\n",
       " '12_0.5': [1048635, 524317],\n",
       " '13_0.5': [1048635, 524307],\n",
       " '14_0.5': [1048635, 524317],\n",
       " '15_0.5': [1048635, 524317],\n",
       " '16_0.5': [1048635, 524317],\n",
       " '17_0.5': [1048635, 524317],\n",
       " '18_0.5': [1048635, 524313],\n",
       " '19_0.5': [1048635, 524317],\n",
       " '20_0.5': [1048635, 524300],\n",
       " '21_0.5': [1048635, 524314],\n",
       " '22_0.5': [1048635, 524316],\n",
       " '23_0.5': [1048635, 524310],\n",
       " '24_0.5': [1048635, 524317],\n",
       " '25_0.5': [1048635, 524317],\n",
       " '26_0.5': [1048635, 524317],\n",
       " '27_0.5': [1048635, 524317],\n",
       " '28_0.5': [1048635, 524317],\n",
       " '29_0.5': [1048635, 524317],\n",
       " '0_0.3': [1048635, 314588],\n",
       " '1_0.3': [1048635, 314591],\n",
       " '2_0.3': [1048635, 314591],\n",
       " '3_0.3': [1048635, 314585],\n",
       " '4_0.3': [1048635, 314591],\n",
       " '5_0.3': [1048635, 314591],\n",
       " '6_0.3': [1048635, 314589],\n",
       " '7_0.3': [1048635, 314591],\n",
       " '8_0.3': [1048635, 314591],\n",
       " '9_0.3': [1048635, 314588],\n",
       " '10_0.3': [1048635, 314590],\n",
       " '11_0.3': [1048635, 314591],\n",
       " '12_0.3': [1048635, 314590],\n",
       " '13_0.3': [1048635, 314591],\n",
       " '14_0.3': [1048635, 314591],\n",
       " '15_0.3': [1048635, 314591],\n",
       " '16_0.3': [1048635, 314591],\n",
       " '17_0.3': [1048635, 314591],\n",
       " '18_0.3': [1048635, 314586],\n",
       " '19_0.3': [1048635, 314591],\n",
       " '20_0.3': [1048635, 314591],\n",
       " '21_0.3': [1048635, 314590],\n",
       " '22_0.3': [1048635, 314591],\n",
       " '23_0.3': [1048635, 314591],\n",
       " '24_0.3': [1048635, 314585],\n",
       " '25_0.3': [1048635, 314591],\n",
       " '26_0.3': [1048635, 314576],\n",
       " '27_0.3': [1048635, 314589],\n",
       " '28_0.3': [1048635, 314591],\n",
       " '29_0.3': [1048635, 314437],\n",
       " '0_0.1': [1048635, 104837],\n",
       " '1_0.1': [1048635, 104864],\n",
       " '2_0.1': [1048635, 104860],\n",
       " '3_0.1': [1048635, 104863],\n",
       " '4_0.1': [1048635, 104864],\n",
       " '5_0.1': [1048635, 104864],\n",
       " '6_0.1': [1048635, 104864],\n",
       " '7_0.1': [1048635, 104864],\n",
       " '8_0.1': [1048635, 104862],\n",
       " '9_0.1': [1048635, 104863],\n",
       " '10_0.1': [1048635, 104863],\n",
       " '11_0.1': [1048635, 104864],\n",
       " '12_0.1': [1048635, 104864],\n",
       " '13_0.1': [1048635, 104864],\n",
       " '14_0.1': [1048635, 104854],\n",
       " '15_0.1': [1048635, 104850],\n",
       " '16_0.1': [1048635, 104864],\n",
       " '17_0.1': [1048635, 104864],\n",
       " '18_0.1': [1048635, 104864],\n",
       " '19_0.1': [1048635, 104864],\n",
       " '20_0.1': [1048635, 104849],\n",
       " '21_0.1': [1048635, 104864],\n",
       " '22_0.1': [1048635, 104864],\n",
       " '23_0.1': [1048635, 104863],\n",
       " '24_0.1': [1048635, 104834],\n",
       " '25_0.1': [1048635, 104864],\n",
       " '26_0.1': [1048635, 104864],\n",
       " '27_0.1': [1048635, 104863],\n",
       " '28_0.1': [1048635, 104864],\n",
       " '29_0.1': [1048635, 104864],\n",
       " '0_0.01': [1048635, 10484],\n",
       " '1_0.01': [1048635, 10487],\n",
       " '2_0.01': [1048635, 10487],\n",
       " '3_0.01': [1048635, 10487],\n",
       " '4_0.01': [1048635, 10487],\n",
       " '5_0.01': [1048635, 10487],\n",
       " '6_0.01': [1048635, 10473],\n",
       " '7_0.01': [1048635, 10487],\n",
       " '8_0.01': [1048635, 10487],\n",
       " '9_0.01': [1048635, 10484],\n",
       " '10_0.01': [1048635, 10487],\n",
       " '11_0.01': [1048635, 10487],\n",
       " '12_0.01': [1048635, 10487],\n",
       " '13_0.01': [1048635, 10487],\n",
       " '14_0.01': [1048635, 10487],\n",
       " '15_0.01': [1048635, 10456],\n",
       " '16_0.01': [1048635, 10428],\n",
       " '17_0.01': [1048635, 10487],\n",
       " '18_0.01': [1048635, 10486],\n",
       " '19_0.01': [1048635, 10487],\n",
       " '20_0.01': [1048635, 10487],\n",
       " '21_0.01': [1048635, 10482],\n",
       " '22_0.01': [1048635, 10486],\n",
       " '23_0.01': [1048635, 10487],\n",
       " '24_0.01': [1048635, 10487],\n",
       " '25_0.01': [1048635, 10487],\n",
       " '26_0.01': [1048635, 10487],\n",
       " '27_0.01': [1048635, 10487],\n",
       " '28_0.01': [1048635, 10485],\n",
       " '29_0.01': [1048635, 10486],\n",
       " '0_0.001': [1048635, 1049],\n",
       " '1_0.001': [1048635, 1039],\n",
       " '2_0.001': [1048635, 1049],\n",
       " '3_0.001': [1048635, 1024],\n",
       " '4_0.001': [1048635, 1047],\n",
       " '5_0.001': [1048635, 1043],\n",
       " '6_0.001': [1048635, 1048],\n",
       " '7_0.001': [1048635, 1048],\n",
       " '8_0.001': [1048635, 1043],\n",
       " '9_0.001': [1048635, 1048],\n",
       " '10_0.001': [1048635, 1049],\n",
       " '11_0.001': [1048635, 1048],\n",
       " '12_0.001': [1048635, 1049],\n",
       " '13_0.001': [1048635, 1049],\n",
       " '14_0.001': [1048635, 1049],\n",
       " '15_0.001': [1048635, 1048],\n",
       " '16_0.001': [1048635, 1047],\n",
       " '17_0.001': [1048635, 1049],\n",
       " '18_0.001': [1048635, 1049],\n",
       " '19_0.001': [1048635, 1049],\n",
       " '20_0.001': [1048635, 1049],\n",
       " '21_0.001': [1048635, 1047],\n",
       " '22_0.001': [1048635, 1046],\n",
       " '23_0.001': [1048635, 1049],\n",
       " '24_0.001': [1048635, 1049],\n",
       " '25_0.001': [1048635, 1049],\n",
       " '26_0.001': [1048635, 1048],\n",
       " '27_0.001': [1048635, 1049],\n",
       " '28_0.001': [1048635, 1049],\n",
       " '29_0.001': [1048635, 1049]}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f7532e3-7baf-4826-abcc-fb3709223906",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_number = {}\n",
    "for i in [1,0.8,0.5,0.3,0.1,0.01, 0.001]:\n",
    "    print(i)\n",
    "    go_per_site = {}\n",
    "    for split in range(len(samples)):\n",
    "        print(split)\n",
    "        br_results = pd.read_csv(f'results/split_{split}/binomial_reg_results_last_gen.csv')\n",
    "        il = len(br_results)\n",
    "        br_results = br_results[br_results['pvalue'] < br_results['pvalue'].quantile(i)]\n",
    "        snp_number[str(split) + '_' + str(i)] = [il, len(br_results)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "cd9e1783-d0a9-4288-a848-064e28dcbff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_number = pd.DataFrame(snp_number).T.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "6476becc-eabf-474c-91f2-5f8b2e6d997e",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_number['split'] = snp_number['index'].str.split('_').str[0]\n",
    "snp_number['quantile'] = snp_number['index'].str.split('_').str[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "893dd88c-95e3-4816-abe7-ae09b1366b5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_number = snp_number.drop('index',axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "35b802c2-aab3-4f2d-ac0e-94b42180c714",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_number.columns = ['total_snps', 'snps_used', 'split', 'quantile']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d0dca725-f9df-4cfe-a9eb-fd9852b92092",
   "metadata": {},
   "outputs": [],
   "source": [
    "aprox_total_used = snp_number.groupby('quantile')['snps_used'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "a220b254-c51d-4b6f-a75a-8ca1c50326e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>quantile</th>\n",
       "      <th>snps_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.001</td>\n",
       "      <td>1.046933e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.01</td>\n",
       "      <td>1.048300e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.1</td>\n",
       "      <td>1.048604e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.3</td>\n",
       "      <td>3.145844e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.5</td>\n",
       "      <td>5.243125e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.8</td>\n",
       "      <td>8.389012e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>1.048633e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  quantile     snps_used\n",
       "0    0.001  1.046933e+03\n",
       "1     0.01  1.048300e+04\n",
       "2      0.1  1.048604e+05\n",
       "3      0.3  3.145844e+05\n",
       "4      0.5  5.243125e+05\n",
       "5      0.8  8.389012e+05\n",
       "6        1  1.048633e+06"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aprox_total_used.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "68a3c816-ba8f-42a6-a792-047d4cad1710",
   "metadata": {},
   "outputs": [],
   "source": [
    "br_results = pd.read_csv(f'results/split_1/binomial_reg_results_last_gen.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7cc7f03b-78f8-4832-a5a5-53797a439b0d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08250e22-4c5e-4754-8814-326a026df2ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3245d908-6bde-469d-99c0-af960394040a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in [1]: #,0.8,0.5,0.3,0.1,0.01, 0.001]:\n",
    "    go_per_site = {}\n",
    "    for split in range(len(samples)):\n",
    "        #print(split)\n",
    "        br_results = pd.read_csv(f'results/split_{split}/binomial_reg_results_last_gen.csv')\n",
    "        #print(len(br_results))\n",
    "        #br_results = br_results[br_results['pvalue'] < 0.05/len(br_results)]\n",
    "    \n",
    "        br_results = br_results[br_results['pvalue'] < br_results['pvalue'].quantile(i)]\n",
    "        #print(len(br_results))\n",
    "        br_results = br_results.reset_index(drop=True)\n",
    "    \n",
    "        mask_full_genotype = snp_dict['id'].isin(br_results['snp_id'])\n",
    "        \n",
    "        genotype_counts_red = genotype_counts.loc[mask_full_genotype,:]\n",
    "        \n",
    "        genotype_counts_red = genotype_counts_red.reset_index(drop=True)\n",
    "        genotype_counts_red = genotype_counts_red.replace({2: 1, 1: 0.5, 0: 0})\n",
    "        \n",
    "        site_predicted = br_results.columns[0]\n",
    "        print(site_predicted)\n",
    "        site_predict_proba = br_results[site_predicted]\n",
    "    \n",
    "        ## for each site, subrstract the actual value by the probability fo getting a success \n",
    "        genotype_counts_subs = genotype_counts_red.sub(site_predict_proba, axis=0)\n",
    "        genotype_counts_subs = genotype_counts_subs**2\n",
    "        ## ge tthe absolute value \n",
    "        \n",
    "        ## sum over ecotypes \n",
    "        go = (genotype_counts_subs.sum(axis=0)) / len(genotype_counts_subs)\n",
    "        go_per_site[site_predicted] = go\n",
    "        #print(site_predicted)\n",
    "    go_per_site_df = pd.DataFrame(go_per_site)\n",
    "    \n",
    "    go_per_site_df.index = go_per_site_df.index.astype(int)\n",
    "    \n",
    "    go_per_site_df.index = ecotypes_names\n",
    "    \n",
    "    #go_per_site_df.to_csv(f'genomic_offset_per_site_last_gen_prob_squared_l1out_quantile{i}.csv',index=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6182c6f-0b7b-497a-b52a-297202a7599d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6afcbd14-00cb-4daa-b6c7-5032e278441f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[10002,\n",
       " 10006,\n",
       " 10011,\n",
       " 10013,\n",
       " 10014,\n",
       " 159,\n",
       " 265,\n",
       " 5151,\n",
       " 5165,\n",
       " 5768,\n",
       " 5772,\n",
       " 5784,\n",
       " 5811,\n",
       " 6013,\n",
       " 6025,\n",
       " 6040,\n",
       " 6073,\n",
       " 6074,\n",
       " 6108,\n",
       " 6150,\n",
       " 6177,\n",
       " 6180,\n",
       " 6184,\n",
       " 6188,\n",
       " 6195,\n",
       " 6209,\n",
       " 6216,\n",
       " 6243,\n",
       " 6244,\n",
       " 6898,\n",
       " 6911,\n",
       " 6915,\n",
       " 6929,\n",
       " 6932,\n",
       " 6938,\n",
       " 6940,\n",
       " 6945,\n",
       " 6958,\n",
       " 6961,\n",
       " 6963,\n",
       " 6979,\n",
       " 6987,\n",
       " 7000,\n",
       " 7002,\n",
       " 7003,\n",
       " 7008,\n",
       " 7013,\n",
       " 7025,\n",
       " 7028,\n",
       " 7031,\n",
       " 7036,\n",
       " 7062,\n",
       " 7063,\n",
       " 7067,\n",
       " 7071,\n",
       " 7077,\n",
       " 7092,\n",
       " 7103,\n",
       " 7106,\n",
       " 7125,\n",
       " 7126,\n",
       " 7127,\n",
       " 7143,\n",
       " 7164,\n",
       " 7165,\n",
       " 7186,\n",
       " 7203,\n",
       " 7209,\n",
       " 7217,\n",
       " 7218,\n",
       " 7244,\n",
       " 7255,\n",
       " 7268,\n",
       " 7273,\n",
       " 7276,\n",
       " 7282,\n",
       " 7287,\n",
       " 7288,\n",
       " 7296,\n",
       " 7298,\n",
       " 7316,\n",
       " 7323,\n",
       " 7333,\n",
       " 7346,\n",
       " 7347,\n",
       " 7353,\n",
       " 7372,\n",
       " 7378,\n",
       " 7384,\n",
       " 7394,\n",
       " 7404,\n",
       " 7411,\n",
       " 7521,\n",
       " 763,\n",
       " 765,\n",
       " 766,\n",
       " 768,\n",
       " 772,\n",
       " 8214,\n",
       " 8230,\n",
       " 8231,\n",
       " 8240,\n",
       " 8247,\n",
       " 8249,\n",
       " 8311,\n",
       " 8312,\n",
       " 8351,\n",
       " 8354,\n",
       " 8357,\n",
       " 8376,\n",
       " 8387,\n",
       " 9057,\n",
       " 9058,\n",
       " 9323,\n",
       " 9371,\n",
       " 9394,\n",
       " 9416,\n",
       " 9427,\n",
       " 9470,\n",
       " 9481,\n",
       " 9506,\n",
       " 9507,\n",
       " 9510,\n",
       " 9512,\n",
       " 9517,\n",
       " 9518,\n",
       " 9521,\n",
       " 9522,\n",
       " 9523,\n",
       " 9524,\n",
       " 9526,\n",
       " 9527,\n",
       " 9528,\n",
       " 9529,\n",
       " 9534,\n",
       " 9535,\n",
       " 9537,\n",
       " 9539,\n",
       " 9542,\n",
       " 9544,\n",
       " 9547,\n",
       " 9548,\n",
       " 9549,\n",
       " 9555,\n",
       " 9557,\n",
       " 9559,\n",
       " 9560,\n",
       " 9562,\n",
       " 9564,\n",
       " 9565,\n",
       " 9574,\n",
       " 9577,\n",
       " 9579,\n",
       " 9584,\n",
       " 9586,\n",
       " 9587,\n",
       " 9591,\n",
       " 9594,\n",
       " 9595,\n",
       " 9596,\n",
       " 9597,\n",
       " 9598,\n",
       " 9600,\n",
       " 9602,\n",
       " 9606,\n",
       " 9612,\n",
       " 9619,\n",
       " 9625,\n",
       " 9629,\n",
       " 9632,\n",
       " 9634,\n",
       " 9637,\n",
       " 9640,\n",
       " 9643,\n",
       " 9649,\n",
       " 9653,\n",
       " 9657,\n",
       " 9659,\n",
       " 9697,\n",
       " 9698,\n",
       " 9699,\n",
       " 9713,\n",
       " 9716,\n",
       " 9719,\n",
       " 9723,\n",
       " 9726,\n",
       " 9736,\n",
       " 9737,\n",
       " 9739,\n",
       " 9741,\n",
       " 9743,\n",
       " 9748,\n",
       " 9749,\n",
       " 9758,\n",
       " 9759,\n",
       " 9761,\n",
       " 9764,\n",
       " 9766,\n",
       " 9769,\n",
       " 9774,\n",
       " 9775,\n",
       " 9779,\n",
       " 9781,\n",
       " 9782,\n",
       " 9784,\n",
       " 9803,\n",
       " 9804,\n",
       " 9813,\n",
       " 9814,\n",
       " 9820,\n",
       " 9837,\n",
       " 9857,\n",
       " 9879,\n",
       " 9910,\n",
       " 9920,\n",
       " 9921,\n",
       " 9925,\n",
       " 9933,\n",
       " 9941,\n",
       " 9944,\n",
       " 9947,\n",
       " 9965,\n",
       " 9966,\n",
       " 9978,\n",
       " 9985,\n",
       " 100001,\n",
       " 100002,\n",
       " 6939,\n",
       " 9940,\n",
       " 9977,\n",
       " 9992]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ecotypes_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "037624b7-828e-4820-9739-630571c33473",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0526a27b-4f83-42bb-bc11-cda4da302706",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>...</th>\n",
       "      <th>45</th>\n",
       "      <th>46</th>\n",
       "      <th>48</th>\n",
       "      <th>49</th>\n",
       "      <th>52</th>\n",
       "      <th>53</th>\n",
       "      <th>54</th>\n",
       "      <th>55</th>\n",
       "      <th>57</th>\n",
       "      <th>60</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.134432</td>\n",
       "      <td>0.133528</td>\n",
       "      <td>0.133304</td>\n",
       "      <td>0.132571</td>\n",
       "      <td>0.133631</td>\n",
       "      <td>0.133762</td>\n",
       "      <td>0.135316</td>\n",
       "      <td>0.133506</td>\n",
       "      <td>0.133577</td>\n",
       "      <td>0.135318</td>\n",
       "      <td>...</td>\n",
       "      <td>0.133495</td>\n",
       "      <td>0.134258</td>\n",
       "      <td>0.135079</td>\n",
       "      <td>0.133996</td>\n",
       "      <td>0.133767</td>\n",
       "      <td>0.134164</td>\n",
       "      <td>0.133811</td>\n",
       "      <td>0.134060</td>\n",
       "      <td>0.133765</td>\n",
       "      <td>0.134391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.135435</td>\n",
       "      <td>0.134573</td>\n",
       "      <td>0.135206</td>\n",
       "      <td>0.134445</td>\n",
       "      <td>0.134443</td>\n",
       "      <td>0.134881</td>\n",
       "      <td>0.135818</td>\n",
       "      <td>0.134616</td>\n",
       "      <td>0.134762</td>\n",
       "      <td>0.135490</td>\n",
       "      <td>...</td>\n",
       "      <td>0.134539</td>\n",
       "      <td>0.135290</td>\n",
       "      <td>0.136386</td>\n",
       "      <td>0.134973</td>\n",
       "      <td>0.135003</td>\n",
       "      <td>0.134932</td>\n",
       "      <td>0.135004</td>\n",
       "      <td>0.135215</td>\n",
       "      <td>0.134942</td>\n",
       "      <td>0.135130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.142458</td>\n",
       "      <td>0.148046</td>\n",
       "      <td>0.150108</td>\n",
       "      <td>0.150298</td>\n",
       "      <td>0.146751</td>\n",
       "      <td>0.144546</td>\n",
       "      <td>0.156080</td>\n",
       "      <td>0.145900</td>\n",
       "      <td>0.145001</td>\n",
       "      <td>0.154909</td>\n",
       "      <td>...</td>\n",
       "      <td>0.147463</td>\n",
       "      <td>0.142672</td>\n",
       "      <td>0.141004</td>\n",
       "      <td>0.143525</td>\n",
       "      <td>0.143382</td>\n",
       "      <td>0.144034</td>\n",
       "      <td>0.144504</td>\n",
       "      <td>0.143264</td>\n",
       "      <td>0.144289</td>\n",
       "      <td>0.153045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.134749</td>\n",
       "      <td>0.132613</td>\n",
       "      <td>0.132894</td>\n",
       "      <td>0.132434</td>\n",
       "      <td>0.132997</td>\n",
       "      <td>0.133663</td>\n",
       "      <td>0.132683</td>\n",
       "      <td>0.133135</td>\n",
       "      <td>0.133436</td>\n",
       "      <td>0.132619</td>\n",
       "      <td>...</td>\n",
       "      <td>0.132831</td>\n",
       "      <td>0.134176</td>\n",
       "      <td>0.135964</td>\n",
       "      <td>0.133935</td>\n",
       "      <td>0.133682</td>\n",
       "      <td>0.133643</td>\n",
       "      <td>0.133907</td>\n",
       "      <td>0.134371</td>\n",
       "      <td>0.133793</td>\n",
       "      <td>0.132400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.135577</td>\n",
       "      <td>0.133267</td>\n",
       "      <td>0.133508</td>\n",
       "      <td>0.132980</td>\n",
       "      <td>0.133748</td>\n",
       "      <td>0.134445</td>\n",
       "      <td>0.133133</td>\n",
       "      <td>0.133859</td>\n",
       "      <td>0.134203</td>\n",
       "      <td>0.133111</td>\n",
       "      <td>...</td>\n",
       "      <td>0.133518</td>\n",
       "      <td>0.134962</td>\n",
       "      <td>0.136904</td>\n",
       "      <td>0.134737</td>\n",
       "      <td>0.134518</td>\n",
       "      <td>0.134461</td>\n",
       "      <td>0.134698</td>\n",
       "      <td>0.135237</td>\n",
       "      <td>0.134580</td>\n",
       "      <td>0.132917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>0.164514</td>\n",
       "      <td>0.158256</td>\n",
       "      <td>0.160160</td>\n",
       "      <td>0.158176</td>\n",
       "      <td>0.158858</td>\n",
       "      <td>0.162157</td>\n",
       "      <td>0.152662</td>\n",
       "      <td>0.160531</td>\n",
       "      <td>0.161573</td>\n",
       "      <td>0.152570</td>\n",
       "      <td>...</td>\n",
       "      <td>0.158766</td>\n",
       "      <td>0.163028</td>\n",
       "      <td>0.167550</td>\n",
       "      <td>0.162950</td>\n",
       "      <td>0.163510</td>\n",
       "      <td>0.162211</td>\n",
       "      <td>0.162053</td>\n",
       "      <td>0.163671</td>\n",
       "      <td>0.162391</td>\n",
       "      <td>0.154364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>0.127446</td>\n",
       "      <td>0.129354</td>\n",
       "      <td>0.129863</td>\n",
       "      <td>0.129546</td>\n",
       "      <td>0.128827</td>\n",
       "      <td>0.127961</td>\n",
       "      <td>0.134109</td>\n",
       "      <td>0.128430</td>\n",
       "      <td>0.128099</td>\n",
       "      <td>0.133690</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128850</td>\n",
       "      <td>0.128248</td>\n",
       "      <td>0.127443</td>\n",
       "      <td>0.127872</td>\n",
       "      <td>0.127558</td>\n",
       "      <td>0.127896</td>\n",
       "      <td>0.127920</td>\n",
       "      <td>0.127589</td>\n",
       "      <td>0.127909</td>\n",
       "      <td>0.132056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>0.126344</td>\n",
       "      <td>0.125709</td>\n",
       "      <td>0.126777</td>\n",
       "      <td>0.125941</td>\n",
       "      <td>0.125452</td>\n",
       "      <td>0.125936</td>\n",
       "      <td>0.127299</td>\n",
       "      <td>0.125768</td>\n",
       "      <td>0.125866</td>\n",
       "      <td>0.126843</td>\n",
       "      <td>...</td>\n",
       "      <td>0.125771</td>\n",
       "      <td>0.125853</td>\n",
       "      <td>0.127261</td>\n",
       "      <td>0.126045</td>\n",
       "      <td>0.126386</td>\n",
       "      <td>0.125913</td>\n",
       "      <td>0.125850</td>\n",
       "      <td>0.126343</td>\n",
       "      <td>0.125952</td>\n",
       "      <td>0.126561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>0.130355</td>\n",
       "      <td>0.132569</td>\n",
       "      <td>0.133598</td>\n",
       "      <td>0.133129</td>\n",
       "      <td>0.131835</td>\n",
       "      <td>0.131032</td>\n",
       "      <td>0.137549</td>\n",
       "      <td>0.131575</td>\n",
       "      <td>0.131237</td>\n",
       "      <td>0.136791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.132201</td>\n",
       "      <td>0.130464</td>\n",
       "      <td>0.130250</td>\n",
       "      <td>0.130779</td>\n",
       "      <td>0.130923</td>\n",
       "      <td>0.131009</td>\n",
       "      <td>0.130973</td>\n",
       "      <td>0.130583</td>\n",
       "      <td>0.130949</td>\n",
       "      <td>0.135471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>0.121340</td>\n",
       "      <td>0.122101</td>\n",
       "      <td>0.122941</td>\n",
       "      <td>0.122571</td>\n",
       "      <td>0.121813</td>\n",
       "      <td>0.121410</td>\n",
       "      <td>0.125508</td>\n",
       "      <td>0.121593</td>\n",
       "      <td>0.121426</td>\n",
       "      <td>0.125053</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121998</td>\n",
       "      <td>0.121311</td>\n",
       "      <td>0.121639</td>\n",
       "      <td>0.121243</td>\n",
       "      <td>0.121108</td>\n",
       "      <td>0.121281</td>\n",
       "      <td>0.121517</td>\n",
       "      <td>0.121361</td>\n",
       "      <td>0.121396</td>\n",
       "      <td>0.124092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>231 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            1         2         4         5         6         9        10  \\\n",
       "0    0.134432  0.133528  0.133304  0.132571  0.133631  0.133762  0.135316   \n",
       "1    0.135435  0.134573  0.135206  0.134445  0.134443  0.134881  0.135818   \n",
       "2    0.142458  0.148046  0.150108  0.150298  0.146751  0.144546  0.156080   \n",
       "3    0.134749  0.132613  0.132894  0.132434  0.132997  0.133663  0.132683   \n",
       "4    0.135577  0.133267  0.133508  0.132980  0.133748  0.134445  0.133133   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "226  0.164514  0.158256  0.160160  0.158176  0.158858  0.162157  0.152662   \n",
       "227  0.127446  0.129354  0.129863  0.129546  0.128827  0.127961  0.134109   \n",
       "228  0.126344  0.125709  0.126777  0.125941  0.125452  0.125936  0.127299   \n",
       "229  0.130355  0.132569  0.133598  0.133129  0.131835  0.131032  0.137549   \n",
       "230  0.121340  0.122101  0.122941  0.122571  0.121813  0.121410  0.125508   \n",
       "\n",
       "           11        12        13  ...        45        46        48  \\\n",
       "0    0.133506  0.133577  0.135318  ...  0.133495  0.134258  0.135079   \n",
       "1    0.134616  0.134762  0.135490  ...  0.134539  0.135290  0.136386   \n",
       "2    0.145900  0.145001  0.154909  ...  0.147463  0.142672  0.141004   \n",
       "3    0.133135  0.133436  0.132619  ...  0.132831  0.134176  0.135964   \n",
       "4    0.133859  0.134203  0.133111  ...  0.133518  0.134962  0.136904   \n",
       "..        ...       ...       ...  ...       ...       ...       ...   \n",
       "226  0.160531  0.161573  0.152570  ...  0.158766  0.163028  0.167550   \n",
       "227  0.128430  0.128099  0.133690  ...  0.128850  0.128248  0.127443   \n",
       "228  0.125768  0.125866  0.126843  ...  0.125771  0.125853  0.127261   \n",
       "229  0.131575  0.131237  0.136791  ...  0.132201  0.130464  0.130250   \n",
       "230  0.121593  0.121426  0.125053  ...  0.121998  0.121311  0.121639   \n",
       "\n",
       "           49        52        53        54        55        57        60  \n",
       "0    0.133996  0.133767  0.134164  0.133811  0.134060  0.133765  0.134391  \n",
       "1    0.134973  0.135003  0.134932  0.135004  0.135215  0.134942  0.135130  \n",
       "2    0.143525  0.143382  0.144034  0.144504  0.143264  0.144289  0.153045  \n",
       "3    0.133935  0.133682  0.133643  0.133907  0.134371  0.133793  0.132400  \n",
       "4    0.134737  0.134518  0.134461  0.134698  0.135237  0.134580  0.132917  \n",
       "..        ...       ...       ...       ...       ...       ...       ...  \n",
       "226  0.162950  0.163510  0.162211  0.162053  0.163671  0.162391  0.154364  \n",
       "227  0.127872  0.127558  0.127896  0.127920  0.127589  0.127909  0.132056  \n",
       "228  0.126045  0.126386  0.125913  0.125850  0.126343  0.125952  0.126561  \n",
       "229  0.130779  0.130923  0.131009  0.130973  0.130583  0.130949  0.135471  \n",
       "230  0.121243  0.121108  0.121281  0.121517  0.121361  0.121396  0.124092  \n",
       "\n",
       "[231 rows x 30 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('genomic_offset_per_site_last_gen_prob_squared_l1out_quantile1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8646e8bf-3238-4f79-90b9-0749043563fb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ccc3106-980a-4339-9bd0-f22c84be2231",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genomic_offset_per_site_last_gen_prob_squared_l1out\n",
    "# combined_results_onlysign_p_top_ecotypes.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "821a2cac-94e9-46ee-a50c-ff0f769d23a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## start here  # genomic_offset_per_site_last_gen_prob_squared_l1out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a875befb-281d-4d85-a245-ad0a28aed05e",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_per_site_df = pd.read_csv('genomic_offset_per_site_last_gen_prob_squared_l1out.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3b1afc-d2fe-43a8-81b0-bb34825c5b4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b5578c3-3552-49ef-833a-b94caa9bcb0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "571e7b10-5ea1-41d5-9770-94582a1bbb19",
   "metadata": {},
   "outputs": [],
   "source": [
    "for quantile in [1,0.8,0.5,0.3,0.1,0.01, 0.001]:\n",
    "    \n",
    "    go_per_site_df = pd.read_csv(f'genomic_offset_per_site_last_gen_prob_squared_l1out_quantile{quantile}.csv')\n",
    "    \n",
    "    go_per_site_df.index = ecotypes_names\n",
    "    \n",
    "    results = {}\n",
    "    for site in unique_sites:\n",
    "        if site != 33:\n",
    "            ## one site \n",
    "            ## retreive ecotype frequecnies \n",
    "            ef1 = ef[[col for col in ef.columns if col.startswith(f'{site}_')]]\n",
    "            cd = go_per_site_df[str(site)]\n",
    "            \n",
    "            for sample in ef1.columns:\n",
    "                ef_f = ef1[sample]\n",
    "                conc = pd.concat([ef_f, cd],axis=1)\n",
    "                conc = conc.sort_values(sample)\n",
    "                ## becasue some ecotpyes are nan \n",
    "                conc = conc.dropna()\n",
    "                X_ranked = conc[sample]\n",
    "                y_ranked = conc[str(site)]\n",
    "        \n",
    "                sp_correlation, _ = spearmanr(X_ranked, y_ranked)\n",
    "                pearsonr_value = pearsonr(X_ranked, y_ranked)[0]\n",
    "                kendall_tau, _ = kendalltau(X_ranked, y_ranked)\n",
    "    \n",
    "                X = conc[str(site)].values.reshape(-1, 1)  # Independent variable\n",
    "                y = conc[sample].values # Dependent variable\n",
    "        \n",
    "                # Fit the linear model\n",
    "                model = LinearRegression().fit(X, y)\n",
    "                # Predict the valu.es\n",
    "                y_pred = model.predict(X)\n",
    "                \n",
    "                r_squared = r2_score(y, y_pred)\n",
    "        \n",
    "                res_list = [sp_correlation, pearsonr_value, r_squared, kendall_tau]\n",
    "        \n",
    "                for i in [10,20,30,40,50, 60, 100]:\n",
    "                    top_10_observed_indices = conc[sample].sort_values().tail(i).index\n",
    "                    top_10_predicted_indices = conc[str(site)].sort_values().head(i).index\n",
    "            \n",
    "                    # Compare top 10 indices\n",
    "                    common_top_indices = len(set(top_10_observed_indices).intersection(set(top_10_predicted_indices)))\n",
    "            \n",
    "                    # Calculate the percentage of common top 10 indices\n",
    "                    top_10_match_percentage = (common_top_indices / i) * 100\n",
    "                    res_list.append(top_10_match_percentage)\n",
    "                results[sample] = res_list\n",
    "    \n",
    "    results = pd.DataFrame(results).T\n",
    "    \n",
    "    results.columns = ['sp_correlation', 'pearsonr', 'r_squared','kendall_tau', 'top_10','top_20','top_30','top_40','top_50', 'top_60', 'top_100']\n",
    "    \n",
    "    results['sp_correlation'] = results['sp_correlation'] *-1\n",
    "    \n",
    "    results['pearsonr'] = results['pearsonr']  *-1\n",
    "    results['kendall_tau'] = results['kendall_tau']  *-1\n",
    "    results = results.reset_index()\n",
    "    \n",
    "    results['site'] = results['index'].str.split('_').str[0]\n",
    "    \n",
    "    results['plot'] = results['index'].str.split('_').str[2]\n",
    "    print(quantile)\n",
    "    results.to_csv(f'results_{quantile}.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "977b4e67-0ddb-4837-87ae-94be5784f9e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05192f3c-397d-4d88-a135-993cffc61c5d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5baf79-b7d3-488d-b599-140901bf2cea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f245f6e8-9af5-4f43-a79a-b0ba6fb30967",
   "metadata": {},
   "outputs": [],
   "source": [
    "all = {}\n",
    "for quantile in [1,0.8,0.5,0.3,0.1,0.01, 0.001]:\n",
    "    data = pd.read_csv(f'results_{quantile}.csv',)\n",
    "    all[quantile] = data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2571154a-6ad8-43f3-b6aa-d1307b0015df",
   "metadata": {},
   "outputs": [],
   "source": [
    "all = pd.concat(all).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfd1b779-bddc-43c1-893c-f8579db835d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "all.columns = ['quantile_snps_used', 'level_1', 'sample', 'sp_correlation', 'pearsonr',\n",
    "       'r_squared', 'kendall_tau', 'top_10', 'top_20', 'top_30', 'top_40',\n",
    "       'top_50', 'top_60', 'top_100', 'site', 'plot']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd4a1f34-c909-46ac-88fb-abeceddc25a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "all = all.drop(['level_1', 'top_10', 'top_20', 'top_30', 'top_40',\n",
    "       'top_50', 'top_60', 'top_100'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02deafab-3eb0-4cac-86ba-a48493886cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1f80e7b-adf0-4347-a75c-cf17d3fec655",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create a figure and a single axis\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Create the strip plot on the axis\n",
    "sns.stripplot(data=all, y='sp_correlation', x='quantile_snps_used', ax=ax, color='grey', alpha=0.3, s=7.5, zorder=1)\n",
    "\n",
    "# Overlay the box plot on the same axis\n",
    "sns.boxplot(data=all, y='sp_correlation', x='quantile_snps_used', fill=False,showfliers=False, color='black', width=0.5, zorder=2, ax=ax)\n",
    "\n",
    "# Add a horizontal line at y=0\n",
    "ax.axhline(0, color='grey', linestyle='--')\n",
    "#ax.set_title('Climatic Distance and Genomic Offset')\n",
    "\n",
    "# Set properties for the axis\n",
    "ax.grid(axis='y')\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "plt.xlabel('Quantile of SNPs included in GO calculation')\n",
    "plt.ylabel('Spearman Correlation')\n",
    "\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "plt.savefig('Quantile_snps_corr_go.png')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "218ffb14-0ed5-43df-a58e-55a9766e01b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "469b302c-7bdd-45b0-aa9f-64d110cf4da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Add grid to the background for y-axis\n",
    "plt.grid(True, which='major', axis='y', linestyle='--', linewidth=0.7)\n",
    "\n",
    "# Remove all spines\n",
    "sns.despine(left=True, right=True, top=True, bottom=True)\n",
    "\n",
    "ax = sns.stripplot(data=combined_results, y='sp_correlation', x='dataset', order=order, color='grey', alpha=0.3, s=7.5, zorder=1)\n",
    "sns.boxplot(data=combined_results, y='sp_correlation', x='dataset', order=order, showfliers=False, color='black', width=0.5, zorder=2, fill=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d3568d9-09e0-4fff-b0e2-07b9e28d65b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25bcae4f-5c82-4721-a58b-ee1aac8b4d61",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f54b2f4c-d4cb-4536-8ad3-76e336f5a66b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37931c3e-faa0-4e44-8fe9-873ebecf472e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import spearmanr\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.metrics import r2_score\n",
    "from scipy.stats import pearsonr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02119c30-63dc-4f7c-b004-614e978025ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_mean = {}\n",
    "for site in unique_sites:\n",
    "    if site != 33:\n",
    "        ## one site \n",
    "        ## retreive ecotype frequecnies \n",
    "        ef1 = ef[[col for col in ef.columns if col.startswith(f'{site}_')]]\n",
    "        ef_mean = ef1.mean(axis=1)\n",
    "        ef_mean.name = 'mean'\n",
    "        cd = go_per_site_df[str(site)]\n",
    "        \n",
    "        conc = pd.concat([ef_mean, cd],axis=1)\n",
    "        #conc = conc.sort_values('mean')\n",
    "        ## becasue some ecotpyes are nan \n",
    "        conc = conc.dropna()\n",
    "        X_ranked = conc['mean']\n",
    "        y_ranked = conc[str(site)]\n",
    "\n",
    "        sp_correlation, _ = spearmanr(X_ranked, y_ranked)\n",
    "        pearsonr_value = pearsonr(X_ranked, y_ranked)[0]\n",
    "        kendall_tau, _ = kendalltau(X_ranked, y_ranked)\n",
    "\n",
    "        X = conc[str(site)].values.reshape(-1, 1)  # Independent variable\n",
    "        y = conc['mean'].values # Dependent variable\n",
    "\n",
    "        # Fit the linear model\n",
    "        model = LinearRegression().fit(X, y)\n",
    "        # Predict the valu.es\n",
    "        y_pred = model.predict(X)\n",
    "        \n",
    "        r_squared = r2_score(y, y_pred)\n",
    "\n",
    "        res_list = [sp_correlation, pearsonr_value, r_squared, kendall_tau]\n",
    "        for i in [10,20,30,40,50]:\n",
    "            ## tail on ecotypes because the largest go on bottom and head for go because the smalles go teh better/\n",
    "            top_10_observed_indices =  conc['mean'].sort_values().tail(i).index\n",
    "            top_10_predicted_indices = conc[str(site)].sort_values().head(i).index\n",
    "    \n",
    "            # Compare top 10 indices\n",
    "            common_top_indices = len(set(top_10_observed_indices).intersection(set(top_10_predicted_indices)))\n",
    "    \n",
    "            # Calculate the percentage of common top 10 indices\n",
    "            top_10_match_percentage = (common_top_indices / i) * 100\n",
    "            res_list.append(top_10_match_percentage)\n",
    "        results_mean[str(site)] = res_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db71fcee-5a90-4453-95c2-05ee88edfbe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_mean = pd.DataFrame(results_mean).T\n",
    "\n",
    "results_mean.columns = ['sp_correlation', 'pearsonr', 'r_squared','kendall_tau', 'top_10','top_20','top_30','top_40','top_50']\n",
    "\n",
    "results_mean['sp_correlation'] = results_mean['sp_correlation'] * -1\n",
    "\n",
    "results_mean['pearsonr'] = results_mean['pearsonr']  * -1\n",
    "results_mean['kendall_tau'] = results_mean['kendall_tau']  * -1\n",
    "\n",
    "results_mean = results_mean.reset_index()\n",
    "\n",
    "results_mean = results_mean.rename(columns = {'index':'site'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "739134f5-b72a-4dbe-bd40-c0b979ab4559",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = {}\n",
    "for site in unique_sites:\n",
    "    if site != 33:\n",
    "        ## one site \n",
    "        ## retreive ecotype frequecnies \n",
    "        ef1 = ef[[col for col in ef.columns if col.startswith(f'{site}_')]]\n",
    "        cd = go_per_site_df[str(site)]\n",
    "        \n",
    "        for sample in ef1.columns:\n",
    "            ef_f = ef1[sample]\n",
    "            conc = pd.concat([ef_f, cd],axis=1)\n",
    "            conc = conc.sort_values(sample)\n",
    "            ## becasue some ecotpyes are nan \n",
    "            conc = conc.dropna()\n",
    "            X_ranked = conc[sample]\n",
    "            y_ranked = conc[str(site)]\n",
    "    \n",
    "            sp_correlation, _ = spearmanr(X_ranked, y_ranked)\n",
    "            pearsonr_value = pearsonr(X_ranked, y_ranked)[0]\n",
    "            kendall_tau, _ = kendalltau(X_ranked, y_ranked)\n",
    "\n",
    "            X = conc[str(site)].values.reshape(-1, 1)  # Independent variable\n",
    "            y = conc[sample].values # Dependent variable\n",
    "    \n",
    "            # Fit the linear model\n",
    "            model = LinearRegression().fit(X, y)\n",
    "            # Predict the valu.es\n",
    "            y_pred = model.predict(X)\n",
    "            \n",
    "            r_squared = r2_score(y, y_pred)\n",
    "    \n",
    "            res_list = [sp_correlation, pearsonr_value, r_squared, kendall_tau]\n",
    "    \n",
    "            for i in [10,20,30,40,50, 60, 100]:\n",
    "                top_10_observed_indices = conc[sample].sort_values().tail(i).index\n",
    "                top_10_predicted_indices = conc[str(site)].sort_values().head(i).index\n",
    "        \n",
    "                # Compare top 10 indices\n",
    "                common_top_indices = len(set(top_10_observed_indices).intersection(set(top_10_predicted_indices)))\n",
    "        \n",
    "                # Calculate the percentage of common top 10 indices\n",
    "                top_10_match_percentage = (common_top_indices / i) * 100\n",
    "                res_list.append(top_10_match_percentage)\n",
    "            results[sample] = res_list\n",
    "\n",
    "results = pd.DataFrame(results).T\n",
    "\n",
    "results.columns = ['sp_correlation', 'pearsonr', 'r_squared','kendall_tau', 'top_10','top_20','top_30','top_40','top_50', 'top_60', 'top_100']\n",
    "\n",
    "results['sp_correlation'] = results['sp_correlation'] *-1\n",
    "\n",
    "results['pearsonr'] = results['pearsonr']  *-1\n",
    "results['kendall_tau'] = results['kendall_tau']  *-1\n",
    "results = results.reset_index()\n",
    "\n",
    "results['site'] = results['index'].str.split('_').str[0]\n",
    "\n",
    "results['plot'] = results['index'].str.split('_').str[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4974af04-23b3-4bff-88a6-6bfbaa3491a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "results['site'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e25202-c735-478e-bca8-afe7b287ecd4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3f0b54-c27c-4e78-83ee-5bdbfac38e52",
   "metadata": {},
   "outputs": [],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8e59c1a-fa93-4176-8e8e-94dfc7543e75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = results.groupby('site')['sp_correlation'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(16, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=results, y='sp_correlation', x='site', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51ad0b30-9445-4f98-a78e-c81d4bab59fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "382023a7-e712-4d1c-b8da-96ec0b2cac3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd_results_mean = pd.read_csv('../climate_distance/naive_climate_distance_top_ecotypes_mean.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "717ab507-6ff9-4b94-abc2-639d3c3943c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_mean['site'] = results_mean['site'].astype(int)\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results_mean['dataset'] = 'genomic_offset'\n",
    "cd_results_mean['dataset'] = 'climatic_distance'\n",
    "# Concatenate the two DataFrames\n",
    "combined_results_mean = pd.concat([results_mean, cd_results_mean])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27e1d00f-6abf-4b50-b3ff-fc4511c4c175",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c416f2a4-f9a9-406e-bc68-50aebfc16e1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd_results = pd.read_csv('../climate_distance/naive_climate_distance_top_ecotypes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21f8b06a-7478-4eb8-9624-5ab055922da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cd_results_mean = pd.read_csv('../climate_distance/climate_distance_results_mean_leave_1_out.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14acba5c-804c-4d6e-a2f6-71bc1762f94d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cd_results = pd.read_csv('../climate_distance/naive_climate_distance_top_ecotypes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b30280d9-17ac-4b00-9eb4-4bd1c2cf73d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd_model = pd.read_csv('../climate_distance/climate_distance_model_results_leave_1_out_ecotype_freq_top.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b2546b8-66d7-4a1f-b30d-40385cd71417",
   "metadata": {},
   "outputs": [],
   "source": [
    "baseline_pairwise_topecotypes_samples = pd.read_csv('baseline_pairwise_topecotypes_samples_top.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80e87912-51a1-4da4-8491-5144fd3f683c",
   "metadata": {},
   "outputs": [],
   "source": [
    "results['site'] = results['site'].astype(int)\n",
    "baseline_pairwise_topecotypes_samples['site'] = baseline_pairwise_topecotypes_samples['site'].astype(int)\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results['dataset'] = 'genomic_offset'\n",
    "cd_results['dataset'] = 'climatic_distance'\n",
    "\n",
    "cd_model['dataset'] = 'climatic_distance_model'\n",
    "baseline_pairwise_topecotypes_samples['dataset'] = 'baseline_pairwise_samples'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8080e376-3fb4-4f90-8eaa-4d7d50791ed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_results = pd.concat([results, cd_results, baseline_pairwise_topecotypes_samples, cd_model],axis=0)  # ,baseline_pairwise_topecotypes_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5869e77e-97ca-4a50-ba24-0fcac7ce0222",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "combined_results = combined_results.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8aa055-fe78-4294-a8af-8719b8c82754",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['kendall_tau'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(16, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='kendall_tau', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb942bfc-1059-45d8-9aaf-2ad3f343366a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['kendall_tau'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(16, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='kendall_tau', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91fc6e68-2b5b-4cf9-88ff-93f42d25fe0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aee72e10-3265-47bd-a99d-f2810fa3372b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combined_results.to_csv('combined_results_onlysign_p_top_ecotypes.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60bbbab-808f-4baa-8dfe-d7c60b545bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_results.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b56f45-a867-4b86-bfba-7eb7afed292a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "variable = 'kendall_tau'\n",
    "order_touse = sort_bio1_int\n",
    "\n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette = {\n",
    "    'climatic_distance': 'white',  # Light green\n",
    "    'genomic_offset': '#32CD32',     # Dark green\n",
    "    'climatic_distance_model': 'grey',  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "}\n",
    "\n",
    "\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, axes = plt.subplots(6, 1, figsize=(12, 26), sharex=True, sharey=True)\n",
    "\n",
    "# Plot 1: Only 'climatic distance'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'] == 'climatic_distance'], order=order_touse,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[0], palette=palette)\n",
    "axes[0].axhline(0, color='grey', linestyle='--')\n",
    "axes[0].set_title('Climatic Distance')\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], order=order_touse,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[1], palette=palette)\n",
    "axes[1].axhline(0, color='grey', linestyle='--')\n",
    "axes[1].set_title('Climatic Distance and Genomi Offset')\n",
    "\n",
    "# Plot 3: 'climatic distance', 'genomi_offset', and 'climatic_dsitance model'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'climatic_distance_model'])], order=order_touse,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[2], palette=palette)\n",
    "axes[2].axhline(0, color='grey', linestyle='--')\n",
    "axes[2].set_title('Climatic Distance, Genomi Offset, and Climatic Distance Model')\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'climatic_distance_model', 'baseline_pairwise_samples'])], \n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[3], palette=palette)\n",
    "axes[3].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['baseline_pairwise_samples', 'genomic_offset'])], \n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[4], palette=palette)\n",
    "axes[4].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'climatic_distance_model'])], \n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[5], palette=palette)\n",
    "axes[5].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Set common properties\n",
    "for ax in axes.flat:\n",
    "    ax.grid(axis='y')\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1522358c-70ad-4881-a924-7cdd4f36df5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52af5e1f-5e9e-4a20-a07c-2ab861a78d98",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52a5c523-d1bb-476c-8505-3c712d06e68f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "variable = 'sp_correlation'\n",
    "order_touse = sort_bio1_int\n",
    "\n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette = {\n",
    "    'climatic_distance': 'white',  # Light green\n",
    "    'genomic_offset': '#32CD32',     # Dark green\n",
    "    'climatic_distance_model': 'grey',  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "}\n",
    "\n",
    "\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, axes = plt.subplots(6, 1, figsize=(12, 26), sharex=True, sharey=True)\n",
    "\n",
    "# Plot 1: Only 'climatic distance'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'] == 'climatic_distance'], order=order_touse,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[0], palette=palette)\n",
    "axes[0].axhline(0, color='grey', linestyle='--')\n",
    "axes[0].set_title('Climatic Distance')\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], order=order_touse,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[1], palette=palette)\n",
    "axes[1].axhline(0, color='grey', linestyle='--')\n",
    "axes[1].set_title('Climatic Distance and Genomi Offset')\n",
    "\n",
    "# Plot 3: 'climatic distance', 'genomi_offset', and 'climatic_dsitance model'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'climatic_distance_model'])], order=order_touse,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[2], palette=palette)\n",
    "axes[2].axhline(0, color='grey', linestyle='--')\n",
    "axes[2].set_title('Climatic Distance, Genomi Offset, and Climatic Distance Model')\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'climatic_distance_model', 'baseline_pairwise_samples'])], \n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[3], palette=palette)\n",
    "axes[3].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['baseline_pairwise_samples', 'genomic_offset'])], \n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[4], palette=palette)\n",
    "axes[4].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'climatic_distance_model'])], \n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[5], palette=palette)\n",
    "axes[5].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Set common properties\n",
    "for ax in axes.flat:\n",
    "    ax.grid(axis='y')\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c06379-55d4-44f2-8061-f6b455eade3a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "293d7572-99dc-4161-b93b-0923b2427d6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0733d62-75b4-4785-974b-3575b59c5c7b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "variable = 'sp_correlation'\n",
    "order_touse = sort_bio1\n",
    "s_value = 7\n",
    "alpha_value = 0.6 \n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette_stripplot = {\n",
    "    'climatic_distance': 'darkgrey',  # Light green\n",
    "    'genomic_offset': '#32CD32',     # Dark green\n",
    "    'climatic_distance_model': 'grey',  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "}\n",
    "\n",
    "palette_boxplot = {\n",
    "    'climatic_distance': 'lightgrey',  # Light green\n",
    "    'genomic_offset': '#32CD32',     # Dark green\n",
    "    'climatic_distance_model': 'grey',  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "\n",
    "}\n",
    "\n",
    "#combined_results.\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, axes = plt.subplots(2, 1, figsize=(15, 10), sharex=True, sharey=True)\n",
    "\n",
    "# Plot 1: Only 'climatic distance'\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'] == 'climatic_distance'], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[0], palette=palette_stripplot)\n",
    "\n",
    "sns.boxplot(data=combined_results[combined_results['dataset'] == 'climatic_distance'], order = sort_bio1.astype(int),\n",
    "            y=variable, x='site', hue='dataset', ax=axes[0],fill=False,\n",
    "            showfliers=False, color='black',palette=palette_boxplot)\n",
    "\n",
    "axes[0].axhline(0, color='grey', linestyle='--')\n",
    "axes[0].set_title('Climatic Distance')\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[1], palette=palette_stripplot, dodge=True)\n",
    "\n",
    "box_plot = sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], \n",
    "                       order=sort_bio1.astype(int), y=variable, x='site', \n",
    "                       hue='dataset', ax=axes[1],fill=False,\n",
    "                       showfliers=False, palette=palette_boxplot)\n",
    "\n",
    "# Create the boxplot\n",
    "axes[1].axhline(0, color='grey', linestyle='--')\n",
    "axes[1].set_title('Climatic Distance and Genomi Offset')\n",
    "\n",
    "# Set common properties\n",
    "for ax in axes.flat:\n",
    "    ax.grid(axis='y')\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "\n",
    "axes[0].grid(True, which='major', axis='both', linestyle='--', linewidth=0.7)\n",
    "axes[1].grid(True, which='major', axis='both', linestyle='--', linewidth=0.7)\n",
    "\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfce7e41-9272-463a-b104-8ad47531b308",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e04a4663-fb32-4f01-bbac-738e46c7ebb6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9297abb5-0460-45c0-a626-6f4ebb510b47",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "670dc0fe-af8e-4fbc-ba09-64a6fcac75b6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13fe3cd6-9270-4436-a844-4501dc6ca8c5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adb2508a-4156-49c3-a7ac-2467c2860be4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f84efc88-c0cc-4605-ac05-94b458a1eed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "order_touse = sort_bio1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bfab733-f5f5-46b3-ab11-2093d410a7ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable = 'sp_correlation'\n",
    "order_touse = sort_bio1\n",
    "s_value = 9\n",
    "alpha_value = 0.6 \n",
    "font_size = 12\n",
    "\n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "green = '#2C5530'\n",
    "grey = '#969C98'\n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette_stripplot = {\n",
    "    'climatic_distance': 'darkgrey',  # Light green\n",
    "    'genomic_offset': green,     # Dark green\n",
    "    'climatic_distance_model': grey,  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "}\n",
    "\n",
    "palette_boxplot = {\n",
    "    'climatic_distance': 'lightgrey',  # Light green\n",
    "    'genomic_offset': green,     # Dark green\n",
    "    'climatic_distance_model': grey,  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "\n",
    "}\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = results.groupby('site')[variable].mean().sort_values()\n",
    "order_touse = medians_results.index\n",
    "#combined_results.\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, ax = plt.subplots(figsize=(18, 6))\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "stripplot = sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', palette=palette_stripplot, dodge=True)\n",
    "\n",
    "box_plot = sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], \n",
    "                       order=sort_bio1.astype(int), y=variable, x='site', \n",
    "                       hue='dataset',fill=False,\n",
    "                       showfliers=False, palette=palette_boxplot,\n",
    "                      legend= False)\n",
    "\n",
    "\n",
    "# Add stars for statistical significance and color them based on which one is higher\n",
    "for site in test_results:\n",
    "    p_value = test_results[site].pvalue\n",
    "    if p_value < 0.05:\n",
    "        x_position = list(order_touse.astype(int)).index(site)\n",
    "        y_max = combined_results[combined_results['site'] == site][variable].max()\n",
    "        site_data = merged_data[merged_data['site'] == site]\n",
    "        genomic_mean = site_data[variable + '_geno'].mean()\n",
    "        climatic_mean = site_data[variable + '_clim'].mean()\n",
    "        \n",
    "        if genomic_mean > climatic_mean:\n",
    "            star_color = green  # Green if genomic offset is higher\n",
    "        else:\n",
    "            star_color = grey  # Grey if climatic distance is higher\n",
    "\n",
    "        ymax = 0.95\n",
    "        # Add stars for significance with the appropriate color\n",
    "        ax.text(x_position, ymax + 0.05, \"*\", ha='center', va='bottom', color=star_color, fontsize=23)\n",
    "\n",
    "        # Add horizontal bracket (bracket over the comparison)\n",
    "        ax.annotate('', xy=(x_position,ymax + 0.06), xycoords='data',\n",
    "                    xytext=(x_position, ymax - 0.06), textcoords='data',\n",
    "                    arrowprops={'arrowstyle': ']-', 'color': 'grey', 'linewidth': 1.5})\n",
    "\n",
    "ax.grid(axis='y')\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "\n",
    "ax.grid(True, which='major', axis='y', linestyle='--', linewidth=0.7)\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "legend_handles = handles[:len(stripplot.get_legend_handles_labels()[0])]  # Keep only stripplot legend handles\n",
    "legend_labels = ['Genomic Offset', 'Climatic Distance']  # Custom legend labels\n",
    "\n",
    "# Adding the filtered legend back to the plot\n",
    "ax.legend(legend_handles, legend_labels, bbox_to_anchor=(1.01, 0.97), loc='upper left', borderaxespad=0.)\n",
    "ax.set_ylim(-0.5, 1.1)  \n",
    "\n",
    "ax.set_xlabel('Experimental Sites')\n",
    "ax.set_ylabel('Spearman Correlation')\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.rc('font', family='sans-serif', size=font_size, weight='normal')  # Regular font size\n",
    "plt.rc('axes', titlesize=font_size, labelsize=font_size)  # Title and label size for axes\n",
    "plt.rc('xtick', labelsize=font_size)  # X tick labels size\n",
    "plt.rc('ytick', labelsize=font_size)  # Y tick labels size\n",
    "dark_grey = '#4D4D4D' \n",
    "plt.tick_params(axis='both', colors=dark_grey)  # Set tick colors\n",
    "\n",
    "plt.savefig('spearman_go_vs_climatic_distance_ordergo.png')\n",
    "plt.savefig('spearman_go_vs_climatic_distance_ordergo.svg')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71382b5-cee8-4fd7-b61e-30a72fcb197d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e308bfcd-6a4c-4983-a207-dfb2735d0934",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5efdbf50-351a-4e55-932d-d07e9d534dd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations\n",
    "from scipy.stats import mannwhitneyu\n",
    "import pandas as pd\n",
    "\n",
    "# Define the order of your groups\n",
    "groups = ['climatic_distance',  'climatic_distance_model', 'genomic_offset', 'baseline_pairwise_samples']  # ['climatic_distance',  'climatic_distance_model', 'genomic_offset', 'baseline_pairwise_samples']\n",
    "\n",
    "# Create an empty DataFrame to store pairwise p-values\n",
    "p_values = pd.DataFrame(index=groups, columns=groups)\n",
    "\n",
    "# Loop over each pair of groups and perform Wilcoxon rank-sum test\n",
    "for group1, group2 in combinations(groups, 2):\n",
    "    group1_data = combined_results[combined_results['dataset'] == group1]['sp_correlation']\n",
    "    group2_data = combined_results[combined_results['dataset'] == group2]['sp_correlation']\n",
    "    \n",
    "    # Perform the Wilcoxon rank-sum test\n",
    "    stat, p_value = mannwhitneyu(group1_data, group2_data, alternative='two-sided')\n",
    "    \n",
    "    # Store the p-value\n",
    "    p_values.loc[group1, group2] = p_value\n",
    "    p_values.loc[group2, group1] = p_value  # fill in symmetric value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a87bff9-4155-4810-ab1d-b736df92f50f",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5baaa13a-138b-4fa0-a5ec-74b48a7cf2b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44fa15ed-47b3-4c22-a131-6a56800a16f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "# Define the order of your groups\n",
    "order = ['climatic_distance', 'climatic_distance_model', 'genomic_offset', 'baseline_pairwise_samples']\n",
    "ticks = ['Climatic\\nDistance', 'Climatic\\nDistance Model', 'Genomic\\nOffset', 'Within Site\\nSamples']\n",
    "\n",
    "# Initialize the plot\n",
    "plt.figure(figsize=(8, 8)) \n",
    "ax = sns.stripplot(data=combined_results, y='sp_correlation', x='dataset', order=order, color='grey', alpha=0.3, s=7.5, zorder=1)\n",
    "sns.boxplot(data=combined_results, y='sp_correlation', x='dataset', order=order, showfliers=False, color='black', width=0.5, zorder=2, fill=False)\n",
    "\n",
    "# Perform pairwise Mann-Whitney U tests between the relevant groups\n",
    "group_genomic_offset = combined_results[combined_results['dataset'] == 'genomic_offset']['sp_correlation']\n",
    "group_climatic_distance = combined_results[combined_results['dataset'] == 'climatic_distance']['sp_correlation']\n",
    "group_climatic_distance_model = combined_results[combined_results['dataset'] == 'climatic_distance_model']['sp_correlation']\n",
    "\n",
    "# Mann-Whitney U test between 'genomic_offset' and 'climatic_distance'\n",
    "stat, p_value_1 = mannwhitneyu(group_genomic_offset, group_climatic_distance, alternative='two-sided')\n",
    "\n",
    "# Mann-Whitney U test between 'genomic_offset' and 'climatic_distance_model'\n",
    "stat, p_value_2 = mannwhitneyu(group_genomic_offset, group_climatic_distance_model, alternative='two-sided')\n",
    "\n",
    "# Define significance threshold\n",
    "significance_threshold = 0.05\n",
    "\n",
    "# Add stars for significant comparisons\n",
    "ymax = combined_results['sp_correlation'].max() + 0.1\n",
    "height_increment = 0.1\n",
    "\n",
    "# Add stars and brackets if significant for 'genomic_offset' vs 'climatic_distance'\n",
    "if p_value_1 < significance_threshold:\n",
    "    x1, x2 = order.index('genomic_offset'), order.index('climatic_distance')  # positions of the groups\n",
    "    y, h, col = ymax, height_increment, 'grey'  # height and color of the annotation\n",
    "    ax.plot([x1, x1, x2, x2], [y, y + h, y + h, y], lw=1.5, c=col)\n",
    "    ax.text((x1 + x2) * 0.5, y + h, \"*\", ha='center', va='bottom', color='grey', fontsize=23)\n",
    "\n",
    "# Add stars and brackets if significant for 'genomic_offset' vs 'climatic_distance_model'\n",
    "if p_value_2 < significance_threshold:\n",
    "    x1, x2 = order.index('genomic_offset'), order.index('climatic_distance_model')  # positions of the groups\n",
    "    y, h, col = ymax + 0.3, height_increment, 'grey'  # increment height slightly to avoid overlap\n",
    "    ax.plot([x1, x1, x2, x2], [y, y + h, y + h, y], lw=1.5, c=col)\n",
    "    ax.text((x1 + x2) * 0.5, y + h, \"*\", ha='center', va='bottom', color='grey', fontsize=23)\n",
    "\n",
    "# Customize axis labels and ticks\n",
    "plt.xticks([0, 1, 2, 3], ticks)\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Spearman Correlation')\n",
    "\n",
    "# Add grid to the background for y-axis\n",
    "plt.grid(True, which='major', axis='y', linestyle='--', linewidth=0.7)\n",
    "\n",
    "# Remove all spines\n",
    "sns.despine(left=True, right=True, top=True, bottom=True)\n",
    "\n",
    "\n",
    "plt.savefig('go_cm_pairwise_samples_overall.png', )\n",
    "plt.savefig('go_cm_pairwise_samples_overall.svg')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52db3a3b-aad5-4224-aad7-6e92ac5e096c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9207b6a-e130-4764-bb6e-e0891da7d039",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73ed9689-8070-407d-b8b5-d6803ecdc2c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable = 'sp_correlation'\n",
    "order_touse = sort_bio1\n",
    "s_value = 9\n",
    "alpha_value = 0.6 \n",
    "\n",
    "green = '#2C5530'\n",
    "grey = '#969C98'\n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette_stripplot = {\n",
    "    'climatic_distance': 'darkgrey',  # Light green\n",
    "    'genomic_offset': green,     # Dark green\n",
    "    'climatic_distance_model': grey,  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "}\n",
    "\n",
    "palette_boxplot = {\n",
    "    'climatic_distance': 'lightgrey',  # Light green\n",
    "    'genomic_offset': green,     # Dark green\n",
    "    'climatic_distance_model': grey,  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "\n",
    "}\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = results.groupby('site')[variable].mean().sort_values()\n",
    "#combined_results.\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, ax = plt.subplots(figsize=(18, 6))\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "stripplot = sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', palette=palette_stripplot, dodge=True)\n",
    "\n",
    "box_plot = sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], \n",
    "                       order=sort_bio1.astype(int), y=variable, x='site', \n",
    "                       hue='dataset',fill=False,\n",
    "                       showfliers=False, palette=palette_boxplot,\n",
    "                      legend= False)\n",
    "\n",
    "\n",
    "# Add stars for statistical significance and color them based on which one is higher\n",
    "for site in test_results:\n",
    "    p_value = test_results[site].pvalue\n",
    "    if p_value < 0.05:\n",
    "        x_position = list(order_touse.astype(int)).index(site)\n",
    "        y_max = combined_results[combined_results['site'] == site][variable].max()\n",
    "        site_data = merged_data[merged_data['site'] == site]\n",
    "        genomic_mean = site_data[variable + '_geno'].mean()\n",
    "        climatic_mean = site_data[variable + '_clim'].mean()\n",
    "        \n",
    "        if genomic_mean > climatic_mean:\n",
    "            star_color = green  # Green if genomic offset is higher\n",
    "        else:\n",
    "            star_color = grey  # Grey if climatic distance is higher\n",
    "\n",
    "        ymax = 0.95\n",
    "        # Add stars for significance with the appropriate color\n",
    "        ax.text(x_position, ymax + 0.05, \"*\", ha='center', va='bottom', color=star_color, fontsize=23)\n",
    "\n",
    "        # Add horizontal bracket (bracket over the comparison)\n",
    "        ax.annotate('', xy=(x_position,ymax + 0.06), xycoords='data',\n",
    "                    xytext=(x_position, ymax - 0.06), textcoords='data',\n",
    "                    arrowprops={'arrowstyle': ']-', 'color': 'grey', 'linewidth': 1.5})\n",
    "\n",
    "ax.grid(axis='y')\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "\n",
    "ax.grid(True, which='major', axis='y', linestyle='--', linewidth=0.7)\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "legend_handles = handles[:len(stripplot.get_legend_handles_labels()[0])]  # Keep only stripplot legend handles\n",
    "legend_labels = ['Genomic Offset', 'Climatic Distance']  # Custom legend labels\n",
    "\n",
    "# Adding the filtered legend back to the plot\n",
    "ax.legend(legend_handles, legend_labels, bbox_to_anchor=(1.01, 0.97), loc='upper left', borderaxespad=0.)\n",
    "ax.set_ylim(-0.5, 1.1)  \n",
    "\n",
    "ax.set_xlabel('Experimental Sites')\n",
    "ax.set_ylabel('Spearman Correlation')\n",
    "# Adjust layout\n",
    "\n",
    "\n",
    "plt.rc('font', family='sans-serif', size=font_size, weight='normal')  # Regular font size\n",
    "plt.rc('axes', titlesize=font_size, labelsize=font_size)  # Title and label size for axes\n",
    "plt.rc('xtick', labelsize=font_size)  # X tick labels size\n",
    "plt.rc('ytick', labelsize=font_size)  # Y tick labels size\n",
    "dark_grey = '#4D4D4D' \n",
    "plt.tick_params(axis='both', colors=dark_grey)  # Set tick colors\n",
    "\n",
    "plt.savefig('spearman_go_vs_climatic_distance_orderbio1.png')\n",
    "plt.savefig('spearman_go_vs_climatic_distance_orderbio1.svg')\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "598d57f2-7c4b-4a5d-969b-7f993e486fb2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a83fdabd-57ab-45dd-b6e9-2707a24c54bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable = 'sp_correlation'\n",
    "order_touse = sort_bio1\n",
    "s_value = 9\n",
    "alpha_value = 0.5\n",
    "\n",
    "green = '#2C5530'\n",
    "grey = '#969C98'\n",
    "black = '#485696'\n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette_stripplot = {\n",
    "    'climatic_distance': 'darkgrey',  # Light green\n",
    "    'genomic_offset': green,     # Dark green\n",
    "    'climatic_distance_model': grey,  # Lime green\n",
    "    'baseline_pairwise_samples': black # Forest green\n",
    "}\n",
    "\n",
    "palette_boxplot = {\n",
    "    'climatic_distance': 'lightgrey',  # Light green\n",
    "    'genomic_offset': green,     # Dark green\n",
    "    'climatic_distance_model': grey,  # Lime green\n",
    "    'baseline_pairwise_samples': black # Forest green\n",
    "\n",
    "}\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = results.groupby('site')[variable].mean().sort_values()\n",
    "#combined_results.\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, ax = plt.subplots(figsize=(18, 6))\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "stripplot = sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'baseline_pairwise_samples'])], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', palette=palette_stripplot, dodge=True, zorder= 1)\n",
    "\n",
    "box_plot = sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'baseline_pairwise_samples'])], \n",
    "                       order=sort_bio1.astype(int), y=variable, x='site', \n",
    "                       hue='dataset',fill=False,\n",
    "                       showfliers=False, palette=palette_boxplot,\n",
    "                      legend= False,\n",
    "                      zorder= 2)\n",
    "\n",
    "\n",
    "\n",
    "ax.grid(axis='y')\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "\n",
    "ax.grid(True, which='major', axis='y', linestyle='--', linewidth=0.7)\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "legend_handles = handles[:len(stripplot.get_legend_handles_labels()[0])]  # Keep only stripplot legend handles\n",
    "legend_labels = ['Genomic Offset', 'Climatic Distance']  # Custom legend labels\n",
    "\n",
    "# Adding the filtered legend back to the plot\n",
    "ax.legend(legend_handles, legend_labels, bbox_to_anchor=(1.01, 0.97), loc='upper left', borderaxespad=0.)\n",
    "ax.set_ylim(-0.5, 1.1)  \n",
    "\n",
    "ax.set_xlabel('Experimental Sites')\n",
    "ax.set_ylabel('Spearman Correlation')\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5769e8c6-5654-4551-9755-5e7188532cdf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35c86379-730c-4dfb-9a01-7392245c3bbd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df2ecc31-51dc-4312-a6cb-273937b5e7d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d907d46c-3a51-4a32-930a-52695a6bf104",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable = 'kendall_tau'\n",
    "order_touse = sort_bio1\n",
    "s_value = 7\n",
    "alpha_value = 0.6 \n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette_stripplot = {\n",
    "    'climatic_distance': 'darkgrey',  # Light green\n",
    "    'genomic_offset': 'green',     # Dark green\n",
    "    'climatic_distance_model': 'grey',  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "}\n",
    "\n",
    "palette_boxplot = {\n",
    "    'climatic_distance': 'lightgrey',  # Light green\n",
    "    'genomic_offset': '#32CD32',     # Dark green\n",
    "    'climatic_distance_model': 'grey',  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = results.groupby('site')[variable].mean().sort_values()\n",
    "\n",
    "#combined_results.\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, ax = plt.subplots(figsize=(18, 6))\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "stripplot = sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', palette=palette_stripplot, dodge=True)\n",
    "\n",
    "box_plot = sns.boxplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], \n",
    "                       order=sort_bio1.astype(int), y=variable, x='site', \n",
    "                       hue='dataset',fill=False,\n",
    "                       showfliers=False, palette=palette_boxplot,\n",
    "                      legend= False)\n",
    "\n",
    "\n",
    "# Add stars for statistical significance and color them based on which one is higher\n",
    "for site in test_results:\n",
    "    p_value = test_results[site].pvalue\n",
    "    if p_value < 0.05:\n",
    "        x_position = list(order_touse.astype(int)).index(site)\n",
    "        y_max = combined_results[combined_results['site'] == site][variable].max()\n",
    "        site_data = merged_data[merged_data['site'] == site]\n",
    "        genomic_mean = site_data[variable + '_geno'].mean()\n",
    "        climatic_mean = site_data[variable + '_clim'].mean()\n",
    "        \n",
    "        if genomic_mean > climatic_mean:\n",
    "            star_color = 'green'  # Green if genomic offset is higher\n",
    "        else:\n",
    "            star_color = 'darkgrey'  # Grey if climatic distance is higher\n",
    "\n",
    "        ymax = 0.7\n",
    "        # Add stars for significance with the appropriate color\n",
    "        ax.text(x_position, ymax + 0.05, \"*\", ha='center', va='bottom', color=star_color, fontsize=16)\n",
    "\n",
    "        # Add horizontal bracket (bracket over the comparison)\n",
    "        ax.annotate('', xy=(x_position,ymax + 0.06), xycoords='data',\n",
    "                    xytext=(x_position, ymax - 0.06), textcoords='data',\n",
    "                    arrowprops={'arrowstyle': ']-', 'color': 'grey', 'linewidth': 1.5})\n",
    "\n",
    "ax.grid(axis='y')\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "\n",
    "ax.grid(True, which='major', axis='y', linestyle='--', linewidth=0.7)\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "# Customizing the legend to show only stripplot legend items\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "legend_handles = handles[:len(stripplot.get_legend_handles_labels()[0])]  # Keep only stripplot legend handles\n",
    "legend_labels = ['Genomic Offset', 'Climatic Distance']  # Custom legend labels\n",
    "\n",
    "# Adding the filtered legend back to the plot\n",
    "ax.legend(legend_handles, legend_labels, bbox_to_anchor=(1.01, 0.97), loc='upper left', borderaxespad=0.)\n",
    "ax.set_ylim(-0.4, 0.9)  \n",
    "\n",
    "ax.set_xlabel('Experimental Sites')\n",
    "ax.set_ylabel('Spearman Correlation')\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e996bace-f252-4621-8b8a-273f85a623fc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c475191-dbcf-4aa9-a79c-b8adbae8e6b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "743fba0c-58e6-455b-81e0-9ee55169210b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ba855cb-3346-4952-92e1-82afaede1dd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f27e4ad0-0cd3-4231-b562-4173686b5520",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a59392dc-062c-41d5-a0fa-95c3c6971c2e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f927c0-158a-4a06-ac6b-cb1959be7495",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13c57540-f435-4b19-b1e7-7b8efb573243",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable = 'top_30'\n",
    "from scipy.stats import ttest_rel, wilcoxon\n",
    "\n",
    "# To store the p-values\n",
    "test_results = {}\n",
    "\n",
    "for site in combined_results['site'].unique():\n",
    "    \n",
    "    site_data = combined_results[combined_results['site'] == site]    \n",
    "    clim = site_data[site_data['dataset'] == 'climatic_distance'][variable] \n",
    "    geno = site_data[site_data['dataset'] == 'genomic_offset'][variable]\n",
    "    # If not normally distributed, use Wilcoxon signed-rank test\n",
    "    test_results[site] = wilcoxon(geno, clim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f50e230c-78db-4ecd-82ed-b3593fd8b257",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a693965a-6326-40fb-b114-5c83d6ff08cc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "  # This will give you the p-values for each sit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4072c94-5e36-4267-9a93-99d8378e9aa7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb2ba90-a214-41ce-abaa-5d7f5d31c82a",
   "metadata": {},
   "outputs": [],
   "source": [
    "clim['sp_correlation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "009c0aad-7601-4a96-b0a5-025df7a7dd53",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a61c3302-fd36-4205-b337-78ee545854bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "781ffd73-fe42-41c1-8f6d-bebc13fb6577",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a4bf0e-bcd8-4753-b68c-cb28768468ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable = 'sp_correlation'\n",
    "order_touse = sort_bio1\n",
    "s_value = 7\n",
    "alpha_value = 0.6 \n",
    "# Define a custom color palette mapping each dataset to a specific color\n",
    "palette = {\n",
    "    'climatic_distance': 'lightgrey',  # Light green\n",
    "    'genomic_offset': '#32CD32',     # Dark green\n",
    "    'climatic_distance_model': 'grey',  # Lime green\n",
    "    'baseline_pairwise_samples': 'black' # Forest green\n",
    "}\n",
    "\n",
    "#combined_results.\n",
    "# Create 4 subplots (4 rows, 1 column)\n",
    "fig, axes = plt.subplots(6, 1, figsize=(12, 26), sharex=True, sharey=True)\n",
    "\n",
    "# Plot 1: Only 'climatic distance'\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'] == 'climatic_distance'], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[0], palette=palette)\n",
    "axes[0].axhline(0, color='grey', linestyle='--')\n",
    "axes[0].set_title('Climatic Distance')\n",
    "\n",
    "# Plot 2: 'climatic distance' and 'genomi_offset'\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset'])], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[1], palette=palette)\n",
    "axes[1].axhline(0, color='grey', linestyle='--')\n",
    "axes[1].set_title('Climatic Distance and Genomi Offset')\n",
    "\n",
    "# Plot 3: 'climatic distance', 'genomi_offset', and 'climatic_dsitance model'\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'climatic_distance_model'])], order=order_touse,s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', ax=axes[2], palette=palette)\n",
    "axes[2].axhline(0, color='grey', linestyle='--')\n",
    "axes[2].set_title('Climatic Distance, Genomi Offset, and Climatic Distance Model')\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'genomic_offset', 'climatic_distance_model', 'baseline_pairwise_samples'])],s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[3], palette=palette)\n",
    "axes[3].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'].isin(['baseline_pairwise_samples', 'genomic_offset'])],s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[4], palette=palette)\n",
    "axes[4].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Plot 4: All datasets\n",
    "sns.stripplot(data=combined_results[combined_results['dataset'].isin(['climatic_distance', 'climatic_distance_model'])],s=s_value,alpha=alpha_value,\n",
    "            y=variable, x='site', hue='dataset', order=order_touse, ax=axes[5], palette=palette)\n",
    "axes[5].axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "\n",
    "# Set common properties\n",
    "for ax in axes.flat:\n",
    "    ax.grid(axis='y')\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0e23393-ce54-49f0-8660-df7607aa335e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4382bc-c04b-488c-af7f-6209b33783e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('Top 10 ecotypes')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57ee6950-77de-4080-9839-78b606e251d1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62d465e2-1ab3-42f7-bb1b-0949a849f706",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['kendall_tau'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.stripplot(data=combined_results, y='kendall_tau', x='site', hue='dataset', order=medians_results.index.astype(str), s=8, alpha = 0.5)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73b6c96c-288e-42ad-ae59-945481edf64e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "969ff03c-d982-4163-9d93-c66ff238a1ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "combined_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "531362a6-bb3f-4821-a7f3-1a5bb81ecb62",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32d7eef1-a746-4d6d-8108-b31c71803816",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['r_squared'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset']!= 'baseline_pairwise_samples'], y='r_squared', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('Top 10 ecotypes')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fb1feb7-a72f-44a0-a039-54d3d8ed9904",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['pearsonr'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results[combined_results['dataset']!= 'baseline_pairwise_samples'], y='pearsonr', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('Top 10 ecotypes')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03e5a56e-c860-4abf-8422-67ef23e0a0d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84a41cf0-9c49-486f-913f-e09f0bdf819a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=sort_bio1)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('Top 10 ecotypes')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90017783-c520-4b00-a281-86575fda46bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index.astype(str))\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('Top 10 ecotypes')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5676688-2d21-4699-88d3-b2993388dd8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8a10ef4-dcdd-40e9-a3c7-f067cd2bb54a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.stripplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index.astype(str), s=8, alpha = 0.5)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e32b9c90-6525-43fc-8171-980f5a06b88b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a102426-a916-4fd7-8f5b-fdbebcc8c389",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = results.groupby('site')['top_20'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='top_20', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('Top 10 ecotypes')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63204456-d5b9-4a29-921a-70c9571ea6af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20642341-19bf-461b-9817-5fc48a95d970",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['top_30'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='top_30', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "plt.title('Top 20 ecotypes')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94e7cba8-7cbc-446b-8d06-5f8eaeab2bd3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d6a095e-784e-4301-9a6c-e848998ebc8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['top_50'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='top_50', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "plt.title('Top 20 ecotypes')\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8682123-bb73-428f-ae6c-290118715797",
   "metadata": {},
   "outputs": [],
   "source": [
    "medians_results.index.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aab6dd1-f755-4012-9605-b08914a2165b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['top_50'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='top_50', x='site', hue='dataset', order=medians_results.index.astype(str))\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b6ce31a-459f-419e-87a4-bd3515090655",
   "metadata": {},
   "outputs": [],
   "source": [
    "medians_results = cd_results.groupby('site')['top_50'].mean().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31d506c8-f05d-4297-b1c3-93ba18272fe5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "order_str = [str(i) for i in medians_results.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8831a4bd-823e-46d1-b6db-7468d423a861",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fec8ee25-80e2-4a48-adcf-1536cdc8973a",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_results['site'] = combined_results['site'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c2fa8df-5cf6-4336-a857-fca33ccfe0c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9f9e66f-e1df-4bd8-b5a9-5e666ca691dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the order as string for the 'site' column\n",
    "order_str = ['12', '11', '9', '10', '28', '49', '57', '13', '23', '25', '53', '24', '5', '42', '46', '2', \n",
    "             '52', '54', '48', '1', '27', '55', '26', '37', '60', '32', '6', '43', '45', '4']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567d4d69-4e4e-4b05-aa6b-231f3ad2340f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbf09173-08bb-49f1-abad-c8e6377a958a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the size of the plot\n",
    "plt.figure(figsize=(25, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.stripplot(data=combined_results, y='top_50', x='site', hue='dataset', alpha = .5, order=order_str)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9532a2e-f1a0-49a4-93bd-6d1d63a9c31a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffac7bcf-6287-481c-84d4-bdab0d989315",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ef86f9a-ef71-42d9-abb2-e883dcbf5ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the size of the plot\n",
    "plt.figure(figsize=(25, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.stripplot(data=combined_results, y='top_50', x='site', hue='dataset', alpha = .5, order=sort_bio1, s= 7)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe863e3b-8382-4478-b73e-bc71ee1727fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the size of the plot\n",
    "plt.figure(figsize=(25, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.swarmplot(data=combined_results, y='top_50', x='site', hue='dataset', alpha = .5, order=order_str)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4a9521a-09c0-49b0-acda-51de20825f93",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a3e209c-85b5-4431-922c-a35bb7d57715",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_results_new = pd.read_csv('chat.csv')\n",
    "\n",
    "# Convert the 'site' column to string and apply the order again for the new dataset\n",
    "combined_results_new['site'] = combined_results_new['site'].astype(str)\n",
    "\n",
    "# Define the order as string for the 'site' column\n",
    "order_str = ['12', '11', '9', '10', '28', '49', '57', '13', '23', '25', '53', '24', '5', '42', '46', '2', \n",
    "             '52', '54', '48', '1', '27', '55', '26', '37', '60', '32', '6', '43', '45', '4']\n",
    "\n",
    "# Create the swarm plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "sns.swarmplot(data=combined_results_new, y='top_50', x='site', hue='dataset', alpha=0.5, order=order_str)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "plt.title('Top 50 Ecotypes by Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e6aa239-d10a-410d-8df2-a0c4ab300e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "results.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe2d1e91-9ba7-4abe-b85c-7ac8d3faa020",
   "metadata": {},
   "outputs": [],
   "source": [
    "results['site'] = results['site'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56067167-2c6b-4987-8051-575bc6da9c6c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0655d81b-2b0d-4d64-a187-abbafa85ee94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2c46a9a-096d-4c3c-b971-6286f75900b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_results.groupby('dataset')['sp_correlation'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3123495e-7276-42d9-a892-2f5e631812c2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb383847-701e-4561-848a-f92b603f121d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results['dataset'] = 'results'\n",
    "cd_results['dataset'] = 'results_cd'\n",
    "\n",
    "# Concatenate the two DataFrames\n",
    "combined_results = pd.concat([results, cd_results])\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['r_squared'].mean().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.violinplot(data=combined_results, y='r_squared', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90416d1f-1053-4caa-b03e-79812cc65692",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2de2fe2-b290-40da-a67f-86830a86b277",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_results.groupby('dataset')['sp_correlation'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab4ef9bb-2b96-4753-9320-a9493bcb927f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results['dataset'] = 'results'\n",
    "cd_results['dataset'] = 'results_cd'\n",
    "\n",
    "# Concatenate the two DataFrames\n",
    "combined_results = pd.concat([results, cd_results])\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].median().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(27, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.violinplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45e978e0-81ce-4bb7-b393-4b38a6f5a43e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204751c7-aae0-419f-bd22-31fb45dbac5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results['dataset'] = 'results'\n",
    "cd_results['dataset'] = 'results_cd'\n",
    "\n",
    "# Concatenate the two DataFrames\n",
    "combined_results = pd.concat([results, cd_results])\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].median().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.violinplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04c17eba-759b-485a-9ae7-ac7ac6dd8793",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results['dataset'] = 'results'\n",
    "cd_results['dataset'] = 'results_cd'\n",
    "\n",
    "# Concatenate the two DataFrames\n",
    "combined_results = pd.concat([results, cd_results])\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].median().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Site')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "047294e8-3892-4c9d-bcc4-59d709937110",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results['dataset'] = 'results'\n",
    "cd_results['dataset'] = 'results_cd'\n",
    "\n",
    "# Concatenate the two DataFrames\n",
    "combined_results = pd.concat([results, cd_results])\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = results.groupby('site')['sp_correlation'].median().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.boxplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('sp_correlation')\n",
    "plt.xlabel('Site')\n",
    "plt.ylabel('sp_correlation')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11d15e07-781b-44e3-b692-7400d365bf38",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Add a new column to differentiate between the two DataFrames\n",
    "results['dataset'] = 'results'\n",
    "cd_results['dataset'] = 'results_cd'\n",
    "\n",
    "# Concatenate the two DataFrames\n",
    "combined_results = pd.concat([results, cd_results])\n",
    "\n",
    "# Calculate medians of 'sp_correlation' for 'results' only to determine the order\n",
    "medians_results = cd_results.groupby('site')['sp_correlation'].median().sort_values()\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(16, 6))\n",
    "\n",
    "# Create the violin plot, using 'hue' to distinguish between 'results' and 'results_cd'\n",
    "sns.swarmplot(data=combined_results, y='sp_correlation', x='site', hue='dataset', order=medians_results.index)\n",
    "\n",
    "# Draw a horizontal line at y=0\n",
    "plt.axhline(0, color='grey', linestyle='--')\n",
    "\n",
    "# Grid and spine settings\n",
    "plt.grid(axis='y')\n",
    "ax = plt.gca()  # Get current axes\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# Add labels and title\n",
    "plt.title('sp_correlation')\n",
    "plt.xlabel('Site')\n",
    "plt.ylabel('sp_correlation')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfd7ad06-923b-4153-9997-5f434a69e3a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "conc = pd.concat([ef_f, cd],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8781fcc-4a6e-4ac3-a9a0-ddb45c138b25",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "conc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dca33bc-ac69-481d-8104-1b387c89f267",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0533fae0-5d4a-4051-863b-7008a8640a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = os.listdir('../baypass_lastgen/individual_gfiles_last_gen/')\n",
    "\n",
    "partitions = [int(file.split('_')[1].replace('.txt', '')) for file in files if '.txt' in file]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69dc973c-506b-4262-8b51-5d5942175a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "partitions_r = {}\n",
    "for i in range(len(partitions)):\n",
    "    pickle_file_path = f'../baypass_lastgen/individual_gfiles_last_gen/loci_partition_{i}'\n",
    "    with open(pickle_file_path, 'rb') as file:\n",
    "        loci_f = pickle.load(file)\n",
    "    results = pd.read_csv(f'results/partition{i}.csv')\n",
    "    results['snp_id'] = loci_f\n",
    "    partitions_r[i] = results\n",
    "results = pd.concat(partitions_r).reset_index(drop=True)\n",
    "#results.to_csv(f'results/split_{split}/binomial_reg_results_last_gen.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e726bba7-6b4f-4043-8464-6fc55b250fec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ca0a5c7-456b-47f4-939f-f85f80f73d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_origin_bio1 = pd.read_csv('../key_files/snp_origin_bio1_grenenetvcf.csv')\n",
    "\n",
    "dict_snps = pd.read_csv('../key_files/var_pos_grenenet.csv')\n",
    "\n",
    "snp_origin_bio1 = pd.concat([dict_snps['id'], snp_origin_bio1],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd4090b0-d885-425e-bcfa-84030b6cdf59",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_origin_bio1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83d8945-785d-4f1b-b2a7-986958e46649",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08589616-ef87-49d0-832e-05dd8f5e7633",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = results.merge(snp_origin_bio1, left_on = 'snp_id', right_on = 'id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf1b8a51-4f4c-4eda-a81c-1eac932109df",
   "metadata": {},
   "outputs": [],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff25e459-fc5d-4856-8d5e-6ef356733237",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_sign = results[results['pvalue'] < 0.05/len(results)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "def447f5-e880-4888-8ba4-9bd7c5f0317a",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_sign['slope'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "367340f9-a7ef-46b0-8876-52e9bfcfe837",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(results_sign['slope'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e60763e-f036-4ebc-a2ec-9ee92d335def",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_sign.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3546fde-faed-43b1-8194-b6e5c04c4e5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "\n",
    "# Assuming snp_origin_beta_binomial_reg is your DataFrame with SNP data\n",
    "\n",
    "# Create a scatter plot using density for color\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "sns.scatterplot(\n",
    "    data=results_sign,\n",
    "    x='snp_origin_bio1',\n",
    "    y='slope',\n",
    "    hue='pvalue',  # Use the density as color\n",
    "    palette=\"viridis\",  # Color mapping\n",
    "    edgecolor=\"none\",  # Remove white edge around the dots\n",
    "    alpha=0.3,  # Set transparency\n",
    "    ax=ax\n",
    ")\n",
    "ax.axhline(0, color='grey', linestyle='--', linewidth=1)\n",
    "ax.axvline(0, color='grey', linestyle='--', linewidth=1)\n",
    "\n",
    "# Add a color bar to show the density scale\n",
    "\n",
    "plt.ylabel('Slope binomial Reg')\n",
    "# Display the plot\n",
    "#plt.savefig('snp_origin_across_space.png')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92a105f7-a74c-4592-b8cf-325adc02438a",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_sign.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b74912b8-3eea-4f22-8b0e-8256051355c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "\n",
    "# Assuming snp_origin_beta_binomial_reg is your DataFrame with SNP data\n",
    "\n",
    "# Create a scatter plot using density for color\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "sns.scatterplot(\n",
    "    data=results_sign,\n",
    "    x='snp_origin_bio1',\n",
    "    y='2',\n",
    "    hue='pvalue',  # Use the density as color\n",
    "    palette=\"viridis\",  # Color mapping\n",
    "    edgecolor=\"none\",  # Remove white edge around the dots\n",
    "    alpha=0.3,  # Set transparency\n",
    "    ax=ax\n",
    ")\n",
    "ax.axhline(0, color='grey', linestyle='--', linewidth=1)\n",
    "ax.axvline(0, color='grey', linestyle='--', linewidth=1)\n",
    "\n",
    "# Add a color bar to show the density scale\n",
    "\n",
    "plt.ylabel('Slope binomial Reg')\n",
    "# Display the plot\n",
    "#plt.savefig('snp_origin_across_space.png')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb99386a-cdc2-4c33-badc-0e8c460b3dbf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9b73f5-2fb1-4ef2-a688-8e430d5f8408",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_sign_most = results_sign.sort_values('pvalue').head(10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87dea1d5-bf34-4d4f-988e-d24b2410c290",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "\n",
    "# Assuming snp_origin_beta_binomial_reg is your DataFrame with SNP data\n",
    "\n",
    "# Create a scatter plot using density for color\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "sns.scatterplot(\n",
    "    data=results_sign_most,\n",
    "    x='snp_origin_bio1',\n",
    "    y='45',\n",
    "    hue='pvalue',  # Use the density as color\n",
    "    palette=\"viridis\",  # Color mapping\n",
    "    edgecolor=\"none\",  # Remove white edge around the dots\n",
    "    alpha=0.3,  # Set transparency\n",
    "    ax=ax,\n",
    "    legend = False\n",
    ")\n",
    "ax.axhline(0, color='grey', linestyle='--', linewidth=1)\n",
    "ax.axvline(0, color='grey', linestyle='--', linewidth=1)\n",
    "\n",
    "# Add a color bar to show the density scale\n",
    "\n",
    "plt.ylabel('Slope binomial Reg')\n",
    "# Display the plot\n",
    "#plt.savefig('snp_origin_across_space.png')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6700c569-774f-4eed-a89f-d98b21dde5e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f3a4bd-fd94-4341-9365-757419201078",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_site_scaled = pd.read_csv('env_site_scaled.csv').drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2e6194c-d157-4894-b0d8-561629673bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "est = results_sign[['slope', 'pvalue', 'intercept']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a232358d-9d0d-4037-9484-783b492a6636",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_sign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d167ee5-a5f0-498d-98c4-b300a9427f95",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a0feea-cc3e-49a8-a443-3e095485e4f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "prob_success_all = {}\n",
    "for site,env_value in zip(env_site_scaled['site'], env_site_scaled['env_scaled']):\n",
    "    linear_predictor = env_value * est['slope'] + est['intercept']\n",
    "    prob_success = 1 / (1 + np.exp(-linear_predictor))\n",
    "    prob_success_all[site] = prob_success\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81671fda-0fbe-4d5c-890b-f7a892905936",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(prob_success_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06229b7c-2db7-464f-8a1f-6fd90c8a67aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "prob_success_all = prob_success_all.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ec9f69-64c2-4c6c-905d-777b9f842017",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask_full_genotype = snp_dict['id'].isin(results_sign['snp_id'])\n",
    "genotype_counts_red = genotype_counts.loc[mask_full_genotype,:]\n",
    "genotype_counts_red = genotype_counts_red.reset_index(drop=True)\n",
    "genotype_counts_red = genotype_counts_red.replace({2: 1, 1: 0.5, 0: 0})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e463d298-286d-4bec-990a-8ad0a6aa2f29",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genotype_counts_red.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c61833-47e2-490d-8206-2044afd55b93",
   "metadata": {},
   "outputs": [],
   "source": [
    "prob_success_1_site.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8472d0d-65c3-4dd4-9158-8f55d6ade301",
   "metadata": {},
   "outputs": [],
   "source": [
    "111454048"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4854b990-fd1e-4510-b2be-5aaffa920f34",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d96307b-bdac-46ec-a7c7-03ec6ac376c4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mask_full_genotype = snp_dict['id'].isin(results_sign['snp_id'])\n",
    "genotype_counts_red = genotype_counts.loc[mask_full_genotype,:]\n",
    "genotype_counts_red = genotype_counts_red.reset_index(drop=True)\n",
    "genotype_counts_red = genotype_counts_red.replace({2: 1, 1: 0.5, 0: 0})\n",
    "\n",
    "\n",
    "go_per_site = {}\n",
    "for site in unique_sites:\n",
    "    if site != 33:\n",
    "        print(site)\n",
    "        prob_success_1_site = prob_success_all[site]\n",
    "        genotype_counts_mul = genotype_counts_red.sub(prob_success_1_site, axis=0)   \n",
    "        genotype_counts_mul = genotype_counts_mul**2\n",
    "        ## sum over ecotypes \n",
    "        go = (genotype_counts_mul.sum(axis=0)) / len(genotype_counts_mul)\n",
    "        go_per_site[site] = go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8f31518-d55e-49fd-837d-1a319c05975e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3de6651b-3b61-4b40-b0ad-869ff377feb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_per_site_df = pd.DataFrame(go_per_site)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "427b2e4e-cd46-4378-b03d-94c5676bc6d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_per_site_df.index = go_per_site_df.index.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6ec4f59-dbf0-4866-b34a-b2851587f2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_per_site_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa9744cd-20a9-4fc2-a77c-755bbb051b4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_per_site_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cf648d6-43ba-4f52-ab2d-182c3717dda4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
