{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee56f07c-3e89-4b76-b090-c3d52e506287",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63fd1b5c-c566-4285-87e4-b0feeb295d50",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e2ec606e-7dbe-47f9-a08a-d0bec9189329",
   "metadata": {},
   "outputs": [],
   "source": [
    "# wza_kendalltau_results_bio1.csv'\n",
    "# wza_results_lfmm_bio1.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "df8dcc3d-6b5c-4ffd-af2e-b44847d54b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "tair = pd.read_csv('../key_files/TAIR10_GFF3_genes_transposons_formatted4topr_w_blocks.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0824708d-cf62-4b43-aee3-be0b599c1ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Define a function to match the block from df_genes to the blocks list in tair and retrieve the gene_id(s) from tair\n",
    "def match_gene_id(row, tair):\n",
    "    gene_block = row['block']  # Get the gene block value from df_genes\n",
    "    tair_rows = tair[tair['blocks'].apply(lambda blocks: gene_block in blocks)]  # Check if the gene block exists in any tair block list\n",
    "    if not tair_rows.empty:\n",
    "        return ','.join(tair_rows['gene_symbol'].values)  # Return all gene_ids as a comma-separated string if multiple matches are found\n",
    "    return None  # Return None if no match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "22ea4254-5e53-4b47-9f02-b77da71adab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "dict_blocks = '../key_files/blocks_snpsid_dict.pkl'\n",
    "\n",
    "with open(dict_blocks, 'rb') as file:\n",
    "    dict_blocks = pickle.load(file)\n",
    "\n",
    "reverse_mapping = {item: key for key, values in dict_blocks.items() for item in values}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "83edd88c-b131-4f7f-9cee-b2a3d4586701",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lfmm_wza_block_gene_info_ncbi = pd.read_csv('df_lfmm_wza_block_gene_info_ncbi.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "17a5ca58-c7f9-415f-a88a-205fb60d3729",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>AT1G55670</td>\n",
       "      <td>PSAG</td>\n",
       "      <td>Encodes subunit G of photosystem I, an 11-kDa ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G11840</td>\n",
       "      <td>GLX1</td>\n",
       "      <td>Encodes a glyoxalase I homolog ATGLX1.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70580</td>\n",
       "      <td>AOAT2</td>\n",
       "      <td>Encodes a protein with glyoxylate aminotransfe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70610</td>\n",
       "      <td>ABCB26</td>\n",
       "      <td>member of TAP subfamily</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70660</td>\n",
       "      <td>MMZ2</td>\n",
       "      <td>MMZ2/UEV1B encodes a protein that may play a r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60220</td>\n",
       "      <td>TET4</td>\n",
       "      <td>Member of TETRASPANIN family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60230</td>\n",
       "      <td>SEN2</td>\n",
       "      <td>putative subunit of tRNA splicing endonuclease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60300</td>\n",
       "      <td>LecRK-I.9</td>\n",
       "      <td>Lectin Receptor Kinase involved in protein-pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60340</td>\n",
       "      <td>AAK6</td>\n",
       "      <td>Encodes a nuclear adenylate kinase that intera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60360</td>\n",
       "      <td>ALP</td>\n",
       "      <td>Encodes a senescence-associated thiol protease.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     block    gene_id  gene_name  \\\n",
       "0   1_2867  AT1G55670       PSAG   \n",
       "1    1_485  AT1G11840       GLX1   \n",
       "2    1_485  AT1G70580      AOAT2   \n",
       "3    1_485  AT1G70610     ABCB26   \n",
       "4    1_485  AT1G70660       MMZ2   \n",
       "..     ...        ...        ...   \n",
       "59  5_2922  AT5G60220       TET4   \n",
       "60  5_2922  AT5G60230       SEN2   \n",
       "61  5_2922  AT5G60300  LecRK-I.9   \n",
       "62  5_2922  AT5G60340       AAK6   \n",
       "63  5_2922  AT5G60360        ALP   \n",
       "\n",
       "                                          description  \n",
       "0   Encodes subunit G of photosystem I, an 11-kDa ...  \n",
       "1              Encodes a glyoxalase I homolog ATGLX1.  \n",
       "2   Encodes a protein with glyoxylate aminotransfe...  \n",
       "3                             member of TAP subfamily  \n",
       "4   MMZ2/UEV1B encodes a protein that may play a r...  \n",
       "..                                                ...  \n",
       "59                       Member of TETRASPANIN family  \n",
       "60     putative subunit of tRNA splicing endonuclease  \n",
       "61  Lectin Receptor Kinase involved in protein-pro...  \n",
       "62  Encodes a nuclear adenylate kinase that intera...  \n",
       "63    Encodes a senescence-associated thiol protease.  \n",
       "\n",
       "[64 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_lfmm_wza_block_gene_info_ncbi['']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "72fbb5ed-ec77-473d-8520-cc7633cb6cda",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#dict_blocks['2_1265']  11533904 11534263"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "29a46f19-ac66-4e43-9a4f-37972230c928",
   "metadata": {},
   "outputs": [],
   "source": [
    "## all samples wza on lfmm\n",
    "wza_lfmm = pd.read_csv('../wza/wza_results_lfmm_bio1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a35b6a39-a86c-4c64-bd09-db78737187f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "wza_lfmm = wza_lfmm.rename(columns = {'gene':'block'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6063ae39-5781-4f6d-a398-41ca36bea4fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_wza_lfmm = wza_lfmm[wza_lfmm['Z_pVal'] <= 0.05/len(wza_lfmm)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dd9987a8-4be0-4d35-b5e5-d72fdc720cb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4071580/2714146059.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sign_wza_lfmm['matching_gene_id'] = sign_wza_lfmm.apply(lambda row: match_gene_id(row, tair), axis=1)\n"
     ]
    }
   ],
   "source": [
    "sign_wza_lfmm['matching_gene_id'] = sign_wza_lfmm.apply(lambda row: match_gene_id(row, tair), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "30cd4b03-83aa-453e-b4ea-d2f351114ab7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28918,AT3G28920,AT3G28925,AT3G28930,AT3G28940,AT3G28950,AT3G28956'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm[sign_wza_lfmm['block']=='3_137']['matching_gene_id'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bb5e715-84a9-4e42-bc93-9f5261d23496",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "95013264-7e3e-49c1-b421-3d8e048e569b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7214</th>\n",
       "      <td>2_970</td>\n",
       "      <td>1716</td>\n",
       "      <td>55</td>\n",
       "      <td>12.847247</td>\n",
       "      <td>2.080748e-13</td>\n",
       "      <td>2.243761e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4789</th>\n",
       "      <td>1_702</td>\n",
       "      <td>1699</td>\n",
       "      <td>267</td>\n",
       "      <td>12.236921</td>\n",
       "      <td>1.282245e-221</td>\n",
       "      <td>8.324132e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14961</th>\n",
       "      <td>5_2244</td>\n",
       "      <td>39</td>\n",
       "      <td>14</td>\n",
       "      <td>13.843255</td>\n",
       "      <td>1.193361e-18</td>\n",
       "      <td>3.556710e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7614</th>\n",
       "      <td>3_133</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>13.741720</td>\n",
       "      <td>5.782126e-09</td>\n",
       "      <td>5.730219e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9271</th>\n",
       "      <td>3_2855</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>14.343321</td>\n",
       "      <td>8.559985e-06</td>\n",
       "      <td>7.725087e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7059</th>\n",
       "      <td>2_827</td>\n",
       "      <td>141</td>\n",
       "      <td>82</td>\n",
       "      <td>22.186910</td>\n",
       "      <td>1.602818e-124</td>\n",
       "      <td>8.449797e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>1_485</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>12.925697</td>\n",
       "      <td>5.580758e-03</td>\n",
       "      <td>1.722963e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>12.340603</td>\n",
       "      <td>1.346841e-22</td>\n",
       "      <td>1.942342e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>1_4505</td>\n",
       "      <td>96</td>\n",
       "      <td>9</td>\n",
       "      <td>17.562307</td>\n",
       "      <td>5.923819e-07</td>\n",
       "      <td>4.216780e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>13.132881</td>\n",
       "      <td>1.703152e-32</td>\n",
       "      <td>5.361920e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15696</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>1598</td>\n",
       "      <td>10</td>\n",
       "      <td>18.230969</td>\n",
       "      <td>9.570234e-01</td>\n",
       "      <td>1.710446e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16421</th>\n",
       "      <td>5_780</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "      <td>11.961668</td>\n",
       "      <td>1.818823e-18</td>\n",
       "      <td>3.341260e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5244</th>\n",
       "      <td>2_1116</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "      <td>10.800096</td>\n",
       "      <td>1.438617e-40</td>\n",
       "      <td>3.594995e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7658</th>\n",
       "      <td>3_137</td>\n",
       "      <td>65</td>\n",
       "      <td>22</td>\n",
       "      <td>13.610011</td>\n",
       "      <td>8.035261e-28</td>\n",
       "      <td>6.002880e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8884</th>\n",
       "      <td>3_250</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>11.374437</td>\n",
       "      <td>1.254594e-10</td>\n",
       "      <td>8.289905e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13384</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9785</th>\n",
       "      <td>3_3325</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>12.259003</td>\n",
       "      <td>4.129632e-01</td>\n",
       "      <td>1.366619e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14638</th>\n",
       "      <td>5_1947</td>\n",
       "      <td>215</td>\n",
       "      <td>59</td>\n",
       "      <td>23.067357</td>\n",
       "      <td>9.586678e-66</td>\n",
       "      <td>1.399768e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7399</th>\n",
       "      <td>3_1136</td>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>10.000855</td>\n",
       "      <td>1.141790e-12</td>\n",
       "      <td>1.574742e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>64</td>\n",
       "      <td>12</td>\n",
       "      <td>12.798694</td>\n",
       "      <td>2.029109e-12</td>\n",
       "      <td>2.191944e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>1_4758</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>11.020971</td>\n",
       "      <td>6.073662e-02</td>\n",
       "      <td>2.316135e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        block  SNPs  hits          Z  top_candidate_p        Z_pVal\n",
       "7214    2_970  1716    55  12.847247     2.080748e-13  2.243761e-13\n",
       "4789    1_702  1699   267  12.236921    1.282245e-221  8.324132e-10\n",
       "14961  5_2244    39    14  13.843255     1.193361e-18  3.556710e-09\n",
       "7614    3_133    40     8  13.741720     5.782126e-09  5.730219e-09\n",
       "9271   3_2855    48     6  14.343321     8.559985e-06  7.725087e-09\n",
       "7059    2_827   141    82  22.186910    1.602818e-124  8.449797e-09\n",
       "5407   2_1265    16    15  11.151806     1.585000e-29  1.569795e-08\n",
       "4227    1_485    36     3  12.925697     5.580758e-03  1.722963e-08\n",
       "654    1_1594    30    15  12.340603     1.346841e-22  1.942342e-08\n",
       "3850   1_4505    96     9  17.562307     5.923819e-07  4.216780e-08\n",
       "350    1_1318    44    22  13.132881     1.703152e-32  5.361920e-08\n",
       "15696  5_2922  1598    10  18.230969     9.570234e-01  1.710446e-07\n",
       "16421   5_780    40    14  11.961668     1.818823e-18  3.341260e-07\n",
       "5244   2_1116    26    22  10.800096     1.438617e-40  3.594995e-07\n",
       "7658    3_137    65    22  13.610011     8.035261e-28  6.002880e-07\n",
       "8884    3_250    38     9  11.374437     1.254594e-10  8.289905e-07\n",
       "13384   4_801   225   130  23.960528    7.309601e-196  1.040595e-06\n",
       "9785   3_3325    53     1  12.259003     4.129632e-01  1.366619e-06\n",
       "14638  5_1947   215    59  23.067357     9.586678e-66  1.399768e-06\n",
       "7399   3_1136    24     9  10.000855     1.141790e-12  1.574742e-06\n",
       "2046   1_2867    64    12  12.798694     2.029109e-12  2.191944e-06\n",
       "4126   1_4758    40     2  11.020971     6.073662e-02  2.316135e-06"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm.sort_values('Z_pVal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "52773b95-9544-4244-9b7d-0ef3de545730",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>13.132881</td>\n",
       "      <td>1.703152e-32</td>\n",
       "      <td>5.361920e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>12.340603</td>\n",
       "      <td>1.346841e-22</td>\n",
       "      <td>1.942342e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>64</td>\n",
       "      <td>12</td>\n",
       "      <td>12.798694</td>\n",
       "      <td>2.029109e-12</td>\n",
       "      <td>2.191944e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>1_4505</td>\n",
       "      <td>96</td>\n",
       "      <td>9</td>\n",
       "      <td>17.562307</td>\n",
       "      <td>5.923819e-07</td>\n",
       "      <td>4.216780e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>1_4758</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>11.020971</td>\n",
       "      <td>6.073662e-02</td>\n",
       "      <td>2.316135e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>1_485</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>12.925697</td>\n",
       "      <td>5.580758e-03</td>\n",
       "      <td>1.722963e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4789</th>\n",
       "      <td>1_702</td>\n",
       "      <td>1699</td>\n",
       "      <td>267</td>\n",
       "      <td>12.236921</td>\n",
       "      <td>1.282245e-221</td>\n",
       "      <td>8.324132e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5244</th>\n",
       "      <td>2_1116</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "      <td>10.800096</td>\n",
       "      <td>1.438617e-40</td>\n",
       "      <td>3.594995e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7059</th>\n",
       "      <td>2_827</td>\n",
       "      <td>141</td>\n",
       "      <td>82</td>\n",
       "      <td>22.186910</td>\n",
       "      <td>1.602818e-124</td>\n",
       "      <td>8.449797e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7214</th>\n",
       "      <td>2_970</td>\n",
       "      <td>1716</td>\n",
       "      <td>55</td>\n",
       "      <td>12.847247</td>\n",
       "      <td>2.080748e-13</td>\n",
       "      <td>2.243761e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7399</th>\n",
       "      <td>3_1136</td>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>10.000855</td>\n",
       "      <td>1.141790e-12</td>\n",
       "      <td>1.574742e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7614</th>\n",
       "      <td>3_133</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>13.741720</td>\n",
       "      <td>5.782126e-09</td>\n",
       "      <td>5.730219e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7658</th>\n",
       "      <td>3_137</td>\n",
       "      <td>65</td>\n",
       "      <td>22</td>\n",
       "      <td>13.610011</td>\n",
       "      <td>8.035261e-28</td>\n",
       "      <td>6.002880e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8884</th>\n",
       "      <td>3_250</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>11.374437</td>\n",
       "      <td>1.254594e-10</td>\n",
       "      <td>8.289905e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9271</th>\n",
       "      <td>3_2855</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>14.343321</td>\n",
       "      <td>8.559985e-06</td>\n",
       "      <td>7.725087e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9785</th>\n",
       "      <td>3_3325</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>12.259003</td>\n",
       "      <td>4.129632e-01</td>\n",
       "      <td>1.366619e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13384</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14638</th>\n",
       "      <td>5_1947</td>\n",
       "      <td>215</td>\n",
       "      <td>59</td>\n",
       "      <td>23.067357</td>\n",
       "      <td>9.586678e-66</td>\n",
       "      <td>1.399768e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14961</th>\n",
       "      <td>5_2244</td>\n",
       "      <td>39</td>\n",
       "      <td>14</td>\n",
       "      <td>13.843255</td>\n",
       "      <td>1.193361e-18</td>\n",
       "      <td>3.556710e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15696</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>1598</td>\n",
       "      <td>10</td>\n",
       "      <td>18.230969</td>\n",
       "      <td>9.570234e-01</td>\n",
       "      <td>1.710446e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16421</th>\n",
       "      <td>5_780</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "      <td>11.961668</td>\n",
       "      <td>1.818823e-18</td>\n",
       "      <td>3.341260e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        block  SNPs  hits          Z  top_candidate_p        Z_pVal\n",
       "350    1_1318    44    22  13.132881     1.703152e-32  5.361920e-08\n",
       "654    1_1594    30    15  12.340603     1.346841e-22  1.942342e-08\n",
       "2046   1_2867    64    12  12.798694     2.029109e-12  2.191944e-06\n",
       "3850   1_4505    96     9  17.562307     5.923819e-07  4.216780e-08\n",
       "4126   1_4758    40     2  11.020971     6.073662e-02  2.316135e-06\n",
       "4227    1_485    36     3  12.925697     5.580758e-03  1.722963e-08\n",
       "4789    1_702  1699   267  12.236921    1.282245e-221  8.324132e-10\n",
       "5244   2_1116    26    22  10.800096     1.438617e-40  3.594995e-07\n",
       "5407   2_1265    16    15  11.151806     1.585000e-29  1.569795e-08\n",
       "7059    2_827   141    82  22.186910    1.602818e-124  8.449797e-09\n",
       "7214    2_970  1716    55  12.847247     2.080748e-13  2.243761e-13\n",
       "7399   3_1136    24     9  10.000855     1.141790e-12  1.574742e-06\n",
       "7614    3_133    40     8  13.741720     5.782126e-09  5.730219e-09\n",
       "7658    3_137    65    22  13.610011     8.035261e-28  6.002880e-07\n",
       "8884    3_250    38     9  11.374437     1.254594e-10  8.289905e-07\n",
       "9271   3_2855    48     6  14.343321     8.559985e-06  7.725087e-09\n",
       "9785   3_3325    53     1  12.259003     4.129632e-01  1.366619e-06\n",
       "13384   4_801   225   130  23.960528    7.309601e-196  1.040595e-06\n",
       "14638  5_1947   215    59  23.067357     9.586678e-66  1.399768e-06\n",
       "14961  5_2244    39    14  13.843255     1.193361e-18  3.556710e-09\n",
       "15696  5_2922  1598    10  18.230969     9.570234e-01  1.710446e-07\n",
       "16421   5_780    40    14  11.961668     1.818823e-18  3.341260e-07"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3066410-c677-4d32-8b6e-9240c9955c74",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "c5dae16c-5346-469a-8fdb-b51ae1979794",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the BED file for writing\n",
    "with open(\"wza_lfmm_bio1_sign_blocks.bed\", \"w\") as bed_file:\n",
    "    for block in sign_wza_lfmm['block']:\n",
    "        if block in dict_blocks:\n",
    "            # Extract SNPs for the block\n",
    "            snps = dict_blocks[block]\n",
    "            \n",
    "            # Assuming the format of SNP is like '1_291' where '1' is the chromosome and '291' is the position\n",
    "            # We need to find the min and max positions for the block to create the start-end range\n",
    "            chrom_positions = [snp.split('_') for snp in snps]\n",
    "            chrom = chrom_positions[0][0]  # The chromosome, assuming all SNPs are from the same chromosome\n",
    "            \n",
    "            # Get the minimum and maximum position for the block\n",
    "            start = min(int(pos) for chrom, pos in chrom_positions)\n",
    "            end = max(int(pos) for chrom, pos in chrom_positions)\n",
    "            \n",
    "            # Write the block information in BED format: chrom, start, end, block name\n",
    "            bed_file.write(f\"{chrom}\\t{start}\\t{end}\\t{block}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "88cb60e3-d24e-4b08-83fe-8652c8c64b47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BED file has been created with 'Chr' prefix in chromosome names.\n"
     ]
    }
   ],
   "source": [
    "# Open the BED file for writing\n",
    "with open(\"wza_lfmm_bio1_sign_blocks.bed\", \"w\") as bed_file:\n",
    "    for block in sign_wza_lfmm['block']:\n",
    "        if block in dict_blocks:\n",
    "            # Extract SNPs for the block\n",
    "            snps = dict_blocks[block]\n",
    "            \n",
    "            # Assuming the format of SNP is like '1_291' where '1' is the chromosome and '291' is the position\n",
    "            # We need to find the min and max positions for the block to create the start-end range\n",
    "            chrom_positions = [snp.split('_') for snp in snps]\n",
    "            chrom = chrom_positions[0][0]  # The chromosome, assuming all SNPs are from the same chromosome\n",
    "            \n",
    "            # Prepend 'Chr' to the chromosome name to match GFF naming convention\n",
    "            chrom = f\"Chr{chrom}\"\n",
    "            \n",
    "            # Get the minimum and maximum position for the block\n",
    "            start = min(int(pos) for chrom, pos in chrom_positions)\n",
    "            end = max(int(pos) for chrom, pos in chrom_positions)\n",
    "            \n",
    "            # Write the block information in BED format: chrom, start, end, block name\n",
    "            bed_file.write(f\"{chrom}\\t{start}\\t{end}\\t{block}\\n\")\n",
    "\n",
    "print(\"BED file has been created with 'Chr' prefix in chromosome names.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7a16dc5-dfc2-4272-929a-94237679e62b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "6eff397f-bd12-4dd4-9d49-75cfae27d663",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intersection complete. Results saved in wza_lfmm_bio1_sign_blocks_annotated.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File wza_lfmm_bio1_sign_blocks.bed has inconsistent naming convention for record:\n",
      "1\t9219501\t9221073\t1_1318\n",
      "\n",
      "***** WARNING: File wza_lfmm_bio1_sign_blocks.bed has inconsistent naming convention for record:\n",
      "1\t9219501\t9221073\t1_1318\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Assuming the bed file is already created by the previous code\n",
    "bed_file = \"wza_lfmm_bio1_sign_blocks.bed\"\n",
    "gff_file = \"../key_files/TAIR10_GFF3_genes_transposons.gff\"  # Replace with your actual GFF3 file path\n",
    "output_file = \"wza_lfmm_bio1_sign_blocks_annotated.txt\" \n",
    "\n",
    "# Construct the command to load the bedtools module and run the intersect command\n",
    "cmd = \"\"\"\n",
    "module load BEDTools/2.29.2; \\\n",
    "bedtools intersect -wb -a {bed_file} -b {gff_file} > {output_file}\n",
    "\"\"\".format(gff_file=gff_file, bed_file=bed_file, output_file=output_file)\n",
    "\n",
    "# Run the command\n",
    "try:\n",
    "    subprocess.run(cmd, shell=True, check=True, executable=\"/bin/bash\")  # Use bash to run the command\n",
    "    print(f\"Intersection complete. Results saved in {output_file}\")\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"Error running bedtools: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "b85b5c83-0dcc-4f60-bde7-1d7e30f3bb11",
   "metadata": {},
   "outputs": [
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[101], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m annot \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mwza_lfmm_bio1_sign_blocks_annotated.txt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1679\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1676\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m   1678\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1679\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmapping\u001b[49m\u001b[43m[\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m]\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1680\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1681\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/pandas/io/parsers/c_parser_wrapper.py:93\u001b[0m, in \u001b[0;36mCParserWrapper.__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype_backend\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m     91\u001b[0m     \u001b[38;5;66;03m# Fail here loudly instead of in cython after reading\u001b[39;00m\n\u001b[1;32m     92\u001b[0m     import_optional_dependency(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 93\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader \u001b[38;5;241m=\u001b[39m \u001b[43mparsers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mTextReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39munnamed_cols \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader\u001b[38;5;241m.\u001b[39munnamed_cols\n\u001b[1;32m     97\u001b[0m \u001b[38;5;66;03m# error: Cannot determine type of 'names'\u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge3/envs/pipeline_snakemake/lib/python3.9/site-packages/pandas/_libs/parsers.pyx:555\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    }
   ],
   "source": [
    "annot = pd.read_csv('wza_lfmm_bio1_sign_blocks_annotated.txt', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdbd38cd-60d8-40ec-9f40-de19f1e23331",
   "metadata": {},
   "outputs": [],
   "source": [
    "#annot = annot[annot[6]=='gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "60e2ef93-b99d-409d-a23b-146423d6f441",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot = annot[annot[12].str.contains('protein_coding_gene')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "9d131195-78ba-4c8a-b918-0f072ff62782",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot['gene_id'] = annot[12].str.split(';').str[0].str.replace('ID=', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "23fe8999-643e-447d-9edb-860d00b8618e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>9219501</td>\n",
       "      <td>9221073</td>\n",
       "      <td>1_1318</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>9219463</td>\n",
       "      <td>9223649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G26680;Note=protein_coding_gene;Name=AT1...</td>\n",
       "      <td>AT1G26680</td>\n",
       "      <td>AT1G26680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>10898171</td>\n",
       "      <td>10898797</td>\n",
       "      <td>1_1594</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>10898172</td>\n",
       "      <td>10899912</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G30720;Note=protein_coding_gene;Name=AT1...</td>\n",
       "      <td>AT1G30720</td>\n",
       "      <td>AT1G30720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>10895628</td>\n",
       "      <td>10897109</td>\n",
       "      <td>1_1594</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>10895280</td>\n",
       "      <td>10897109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G30710;Note=protein_coding_gene;Name=AT1...</td>\n",
       "      <td>AT1G30710</td>\n",
       "      <td>AT1G30710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>20800693</td>\n",
       "      <td>20802340</td>\n",
       "      <td>1_2867</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>20800694</td>\n",
       "      <td>20802340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G55660;Note=protein_coding_gene;Name=AT1...</td>\n",
       "      <td>AT1G55660</td>\n",
       "      <td>AT1G55660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>20802669</td>\n",
       "      <td>20803449</td>\n",
       "      <td>1_2867</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>20802670</td>\n",
       "      <td>20803449</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G55670;Note=protein_coding_gene;Name=AT1...</td>\n",
       "      <td>AT1G55670</td>\n",
       "      <td>AT1G55670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4173</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>24227414</td>\n",
       "      <td>24232873</td>\n",
       "      <td>5_2922</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>24227415</td>\n",
       "      <td>24232873</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT5G60170;Note=protein_coding_gene;Name=AT5...</td>\n",
       "      <td>AT5G60170</td>\n",
       "      <td>AT5G60170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4182</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>24264621</td>\n",
       "      <td>24268029</td>\n",
       "      <td>5_2922</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>24264622</td>\n",
       "      <td>24268029</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT5G60300;Note=protein_coding_gene;Name=AT5...</td>\n",
       "      <td>AT5G60300</td>\n",
       "      <td>AT5G60300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4189</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>24279896</td>\n",
       "      <td>24282397</td>\n",
       "      <td>5_2922</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>24279897</td>\n",
       "      <td>24282397</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT5G60360;Note=protein_coding_gene;Name=AT5...</td>\n",
       "      <td>AT5G60360</td>\n",
       "      <td>AT5G60360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4197</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>9425150</td>\n",
       "      <td>9426711</td>\n",
       "      <td>5_780</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>9425151</td>\n",
       "      <td>9426711</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT5G26800;Note=protein_coding_gene;Name=AT5...</td>\n",
       "      <td>AT5G26800</td>\n",
       "      <td>AT5G26800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4208</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>9427542</td>\n",
       "      <td>9427864</td>\n",
       "      <td>5_780</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>TAIR10</td>\n",
       "      <td>gene</td>\n",
       "      <td>9427543</td>\n",
       "      <td>9428067</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT5G26805;Note=protein_coding_gene;Name=AT5...</td>\n",
       "      <td>AT5G26805</td>\n",
       "      <td>AT5G26805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0         1         2       3     4       5     6         7  \\\n",
       "0     Chr1   9219501   9221073  1_1318  Chr1  TAIR10  gene   9219463   \n",
       "16    Chr1  10898171  10898797  1_1594  Chr1  TAIR10  gene  10898172   \n",
       "22    Chr1  10895628  10897109  1_1594  Chr1  TAIR10  gene  10895280   \n",
       "28    Chr1  20800693  20802340  1_2867  Chr1  TAIR10  gene  20800694   \n",
       "42    Chr1  20802669  20803449  1_2867  Chr1  TAIR10  gene  20802670   \n",
       "...    ...       ...       ...     ...   ...     ...   ...       ...   \n",
       "4173  Chr5  24227414  24232873  5_2922  Chr5  TAIR10  gene  24227415   \n",
       "4182  Chr5  24264621  24268029  5_2922  Chr5  TAIR10  gene  24264622   \n",
       "4189  Chr5  24279896  24282397  5_2922  Chr5  TAIR10  gene  24279897   \n",
       "4197  Chr5   9425150   9426711   5_780  Chr5  TAIR10  gene   9425151   \n",
       "4208  Chr5   9427542   9427864   5_780  Chr5  TAIR10  gene   9427543   \n",
       "\n",
       "             8  9 10 11                                                 12  \\\n",
       "0      9223649  .  +  .  ID=AT1G26680;Note=protein_coding_gene;Name=AT1...   \n",
       "16    10899912  .  +  .  ID=AT1G30720;Note=protein_coding_gene;Name=AT1...   \n",
       "22    10897109  .  +  .  ID=AT1G30710;Note=protein_coding_gene;Name=AT1...   \n",
       "28    20802340  .  +  .  ID=AT1G55660;Note=protein_coding_gene;Name=AT1...   \n",
       "42    20803449  .  -  .  ID=AT1G55670;Note=protein_coding_gene;Name=AT1...   \n",
       "...        ... .. .. ..                                                ...   \n",
       "4173  24232873  .  +  .  ID=AT5G60170;Note=protein_coding_gene;Name=AT5...   \n",
       "4182  24268029  .  +  .  ID=AT5G60300;Note=protein_coding_gene;Name=AT5...   \n",
       "4189  24282397  .  +  .  ID=AT5G60360;Note=protein_coding_gene;Name=AT5...   \n",
       "4197   9426711  .  +  .  ID=AT5G26800;Note=protein_coding_gene;Name=AT5...   \n",
       "4208   9428067  .  +  .  ID=AT5G26805;Note=protein_coding_gene;Name=AT5...   \n",
       "\n",
       "        GENE_ID    gene_id  \n",
       "0     AT1G26680  AT1G26680  \n",
       "16    AT1G30720  AT1G30720  \n",
       "22    AT1G30710  AT1G30710  \n",
       "28    AT1G55660  AT1G55660  \n",
       "42    AT1G55670  AT1G55670  \n",
       "...         ...        ...  \n",
       "4173  AT5G60170  AT5G60170  \n",
       "4182  AT5G60300  AT5G60300  \n",
       "4189  AT5G60360  AT5G60360  \n",
       "4197  AT5G26800  AT5G26800  \n",
       "4208  AT5G26805  AT5G26805  \n",
       "\n",
       "[178 rows x 15 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "71894598-c8d4-429d-b4d5-27531681f543",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene IDs saved to gene_ids.txt\n"
     ]
    }
   ],
   "source": [
    "# Assuming your DataFrame has a column for gene IDs, e.g., 'gene_id'\n",
    "gene_ids = annot['gene_id'].unique()  # Get unique gene IDs\n",
    "\n",
    "# Write these gene IDs to a file to use for filtering the GFF3 file\n",
    "with open('wza_lfmm_bio1_sign_blocks_gene_ids.txt', 'w') as f:\n",
    "    for gene_id in gene_ids:\n",
    "        f.write(f\"{gene_id}\\n\")\n",
    "        \n",
    "print(\"Gene IDs saved to gene_ids.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cfe8f39-032c-4e0a-8389-b8d9d1d18a25",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "70b6ca83-d193-46fa-b8f7-12d22a3da0dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='grep -Ff wza_lfmm_bio1_sign_blocks_gene_ids.txt ../key_files/TAIR10_GFF3_genes_transposons.gff > filtered_annotations.gff', returncode=0)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 2: Filter the GFF3 file using the list of gene IDs\n",
    "filter_cmd = \"grep -Ff wza_lfmm_bio1_sign_blocks_gene_ids.txt ../key_files/TAIR10_GFF3_genes_transposons.gff > filtered_annotations.gff\"\n",
    "subprocess.run(filter_cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b72858c-8218-4711-b7ff-9c2c561396d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "GffRead/0.12.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c465a29-3ff6-4998-933f-b9eb17aaa58b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = \"\"\"\n",
    "module load BEDTools/2.29.2; \\\n",
    "bedtools intersect -wb -a {bed_file} -b {gff_file} > {output_file}\n",
    "\"\"\".format(gff_file=gff_file, bed_file=bed_file, output_file=output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "86290505-9000-406a-8bf4-189135b84ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the command\n",
    "try:\n",
    "    subprocess.run(cmd, shell=True, check=True, executable=\"/bin/bash\")  # Use bash to run the command\n",
    "    print(f\"Intersection complete. Results saved in {output_file}\")\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"Error running bedtools: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f05541df-d429-4f9c-9ce5-b50f13c9e57d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d848c66c-5ef5-4990-8fa0-dc07d39149af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome names replaced successfully.\n"
     ]
    }
   ],
   "source": [
    "# Define a function to replace chromosome names in the GFF file\n",
    "def replace_chromosomes(gff_file):\n",
    "    # Read the GFF file\n",
    "    with open(gff_file, 'r') as file:\n",
    "        gff_content = file.read()\n",
    "    \n",
    "    # Replace Chr1, Chr2, ..., with 1, 2, ..., and leave ChrC and ChrM unchanged\n",
    "    gff_content = gff_content.replace('Chr1', '1')\n",
    "    gff_content = gff_content.replace('Chr2', '2')\n",
    "    gff_content = gff_content.replace('Chr3', '3')\n",
    "    gff_content = gff_content.replace('Chr4', '4')\n",
    "    gff_content = gff_content.replace('Chr5', '5')\n",
    "    # ChrC and ChrM remain unchanged, no need for replacements\n",
    "    \n",
    "    # Write the modified content back to the GFF file\n",
    "    with open(gff_file, 'w') as file:\n",
    "        file.write(gff_content)\n",
    "\n",
    "# Specify your GFF file\n",
    "gff_file = \"filtered_annotations.gff\"\n",
    "\n",
    "# Call the function to replace chromosome names\n",
    "replace_chromosomes(gff_file)\n",
    "\n",
    "print(\"Chromosome names replaced successfully.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c98aee8-547f-4d4d-8b86-47e1c67c5524",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "44dffde3-c4ca-4ba5-b816-57c83693df09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nucleotide sequences saved to output_transcripts.fa\n"
     ]
    }
   ],
   "source": [
    "# Step 3: Use gffread to extract sequences\n",
    "gffread_cmd = \"\"\"\n",
    "module load GffRead/0.12.3; \\\n",
    "gffread -w output_transcripts.fa -g ../key_files/arabidopsis_reference/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa filtered_annotations.gff\n",
    "\"\"\"\n",
    "\n",
    "subprocess.run(gffread_cmd, shell=True, check=True, executable=\"/bin/bash\")\n",
    "print(\"Nucleotide sequences saved to output_transcripts.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2d04ae7-6197-47e1-97d1-06d900d27e63",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5c42ed5-47ba-4c8e-8658-ab290fe76e38",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "5e15e960-e0bf-42af-81c7-64ddcdd90e2f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4033925/243824316.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sign_wza_lfmm['matching_gene_id'] = sign_wza_lfmm.apply(lambda row: match_gene_id(row, tair), axis=1)\n"
     ]
    }
   ],
   "source": [
    "# Step 2: Apply the function to df_genes to create a new column with the matching gene_id from tair\n",
    "sign_wza_lfmm['matching_gene_id'] = sign_wza_lfmm.apply(lambda row: match_gene_id(row, tair), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "9910f597-5534-44f6-8d08-aa4e8e2def7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "      <th>matching_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>13.132881</td>\n",
       "      <td>1.703152e-32</td>\n",
       "      <td>5.361920e-08</td>\n",
       "      <td>AT1G26680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>12.340603</td>\n",
       "      <td>1.346841e-22</td>\n",
       "      <td>1.942342e-08</td>\n",
       "      <td>AT1G30710,AT1G30720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>64</td>\n",
       "      <td>12</td>\n",
       "      <td>12.798694</td>\n",
       "      <td>2.029109e-12</td>\n",
       "      <td>2.191944e-06</td>\n",
       "      <td>AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>1_4505</td>\n",
       "      <td>96</td>\n",
       "      <td>9</td>\n",
       "      <td>17.562307</td>\n",
       "      <td>5.923819e-07</td>\n",
       "      <td>4.216780e-08</td>\n",
       "      <td>AT1G67560,AT1G67570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>1_4758</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>11.020971</td>\n",
       "      <td>6.073662e-02</td>\n",
       "      <td>2.316135e-06</td>\n",
       "      <td>AT1G70020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>1_485</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>12.925697</td>\n",
       "      <td>5.580758e-03</td>\n",
       "      <td>1.722963e-08</td>\n",
       "      <td>AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4789</th>\n",
       "      <td>1_702</td>\n",
       "      <td>1699</td>\n",
       "      <td>267</td>\n",
       "      <td>12.236921</td>\n",
       "      <td>1.282245e-221</td>\n",
       "      <td>8.324132e-10</td>\n",
       "      <td>AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5244</th>\n",
       "      <td>2_1116</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "      <td>10.800096</td>\n",
       "      <td>1.438617e-40</td>\n",
       "      <td>3.594995e-07</td>\n",
       "      <td>AT2G24230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "      <td>AT2G27030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7059</th>\n",
       "      <td>2_827</td>\n",
       "      <td>141</td>\n",
       "      <td>82</td>\n",
       "      <td>22.186910</td>\n",
       "      <td>1.602818e-124</td>\n",
       "      <td>8.449797e-09</td>\n",
       "      <td>AT2G17380,AT2G17390,AT2G17410,AT2G17420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7214</th>\n",
       "      <td>2_970</td>\n",
       "      <td>1716</td>\n",
       "      <td>55</td>\n",
       "      <td>12.847247</td>\n",
       "      <td>2.080748e-13</td>\n",
       "      <td>2.243761e-13</td>\n",
       "      <td>AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7399</th>\n",
       "      <td>3_1136</td>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>10.000855</td>\n",
       "      <td>1.141790e-12</td>\n",
       "      <td>1.574742e-06</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7614</th>\n",
       "      <td>3_133</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>13.741720</td>\n",
       "      <td>5.782126e-09</td>\n",
       "      <td>5.730219e-09</td>\n",
       "      <td>AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7658</th>\n",
       "      <td>3_137</td>\n",
       "      <td>65</td>\n",
       "      <td>22</td>\n",
       "      <td>13.610011</td>\n",
       "      <td>8.035261e-28</td>\n",
       "      <td>6.002880e-07</td>\n",
       "      <td>AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8884</th>\n",
       "      <td>3_250</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>11.374437</td>\n",
       "      <td>1.254594e-10</td>\n",
       "      <td>8.289905e-07</td>\n",
       "      <td>AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9271</th>\n",
       "      <td>3_2855</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>14.343321</td>\n",
       "      <td>8.559985e-06</td>\n",
       "      <td>7.725087e-09</td>\n",
       "      <td>AT3G50750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9785</th>\n",
       "      <td>3_3325</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>12.259003</td>\n",
       "      <td>4.129632e-01</td>\n",
       "      <td>1.366619e-06</td>\n",
       "      <td>AT3G62360,AT3G62370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13384</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "      <td>AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14638</th>\n",
       "      <td>5_1947</td>\n",
       "      <td>215</td>\n",
       "      <td>59</td>\n",
       "      <td>23.067357</td>\n",
       "      <td>9.586678e-66</td>\n",
       "      <td>1.399768e-06</td>\n",
       "      <td>AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14961</th>\n",
       "      <td>5_2244</td>\n",
       "      <td>39</td>\n",
       "      <td>14</td>\n",
       "      <td>13.843255</td>\n",
       "      <td>1.193361e-18</td>\n",
       "      <td>3.556710e-09</td>\n",
       "      <td>AT5G48560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15696</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>1598</td>\n",
       "      <td>10</td>\n",
       "      <td>18.230969</td>\n",
       "      <td>9.570234e-01</td>\n",
       "      <td>1.710446e-07</td>\n",
       "      <td>AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16421</th>\n",
       "      <td>5_780</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "      <td>11.961668</td>\n",
       "      <td>1.818823e-18</td>\n",
       "      <td>3.341260e-07</td>\n",
       "      <td>AT5G26800,AT5G26805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        block  SNPs  hits          Z  top_candidate_p        Z_pVal  \\\n",
       "350    1_1318    44    22  13.132881     1.703152e-32  5.361920e-08   \n",
       "654    1_1594    30    15  12.340603     1.346841e-22  1.942342e-08   \n",
       "2046   1_2867    64    12  12.798694     2.029109e-12  2.191944e-06   \n",
       "3850   1_4505    96     9  17.562307     5.923819e-07  4.216780e-08   \n",
       "4126   1_4758    40     2  11.020971     6.073662e-02  2.316135e-06   \n",
       "4227    1_485    36     3  12.925697     5.580758e-03  1.722963e-08   \n",
       "4789    1_702  1699   267  12.236921    1.282245e-221  8.324132e-10   \n",
       "5244   2_1116    26    22  10.800096     1.438617e-40  3.594995e-07   \n",
       "5407   2_1265    16    15  11.151806     1.585000e-29  1.569795e-08   \n",
       "7059    2_827   141    82  22.186910    1.602818e-124  8.449797e-09   \n",
       "7214    2_970  1716    55  12.847247     2.080748e-13  2.243761e-13   \n",
       "7399   3_1136    24     9  10.000855     1.141790e-12  1.574742e-06   \n",
       "7614    3_133    40     8  13.741720     5.782126e-09  5.730219e-09   \n",
       "7658    3_137    65    22  13.610011     8.035261e-28  6.002880e-07   \n",
       "8884    3_250    38     9  11.374437     1.254594e-10  8.289905e-07   \n",
       "9271   3_2855    48     6  14.343321     8.559985e-06  7.725087e-09   \n",
       "9785   3_3325    53     1  12.259003     4.129632e-01  1.366619e-06   \n",
       "13384   4_801   225   130  23.960528    7.309601e-196  1.040595e-06   \n",
       "14638  5_1947   215    59  23.067357     9.586678e-66  1.399768e-06   \n",
       "14961  5_2244    39    14  13.843255     1.193361e-18  3.556710e-09   \n",
       "15696  5_2922  1598    10  18.230969     9.570234e-01  1.710446e-07   \n",
       "16421   5_780    40    14  11.961668     1.818823e-18  3.341260e-07   \n",
       "\n",
       "                                        matching_gene_id  \n",
       "350                                            AT1G26680  \n",
       "654                                  AT1G30710,AT1G30720  \n",
       "2046   AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690  \n",
       "3850                                 AT1G67560,AT1G67570  \n",
       "4126                                           AT1G70020  \n",
       "4227   AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...  \n",
       "4789   AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...  \n",
       "5244                                           AT2G24230  \n",
       "5407                                           AT2G27030  \n",
       "7059             AT2G17380,AT2G17390,AT2G17410,AT2G17420  \n",
       "7214   AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...  \n",
       "7399                                                None  \n",
       "7614   AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...  \n",
       "7658   AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...  \n",
       "8884   AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420  \n",
       "9271                                           AT3G50750  \n",
       "9785                                 AT3G62360,AT3G62370  \n",
       "13384  AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150  \n",
       "14638  AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120  \n",
       "14961                                          AT5G48560  \n",
       "15696  AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...  \n",
       "16421                                AT5G26800,AT5G26805  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9b3b6341-dec5-4818-a790-f0954f9fe3f9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "      <th>matching_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "      <td>AT2G27030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       block  SNPs  hits          Z  top_candidate_p        Z_pVal  \\\n",
       "5407  2_1265    16    15  11.151806     1.585000e-29  1.569795e-08   \n",
       "\n",
       "     matching_gene_id  \n",
       "5407        AT2G27030  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm[sign_wza_lfmm['block'] == '2_1265']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d6a4b438-91f0-4920-b8d4-c85fdf0897e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>biotype</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>exon_chromstart</th>\n",
       "      <th>exon_chromend</th>\n",
       "      <th>blocks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3631</td>\n",
       "      <td>5899</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>3631,3996,4486,4706,5174,5439</td>\n",
       "      <td>3913,4276,4605,5095,5326,5899</td>\n",
       "      <td>['1_0']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5928</td>\n",
       "      <td>8737</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>8571,8417,8236,7942,7762,7564,7384,7157,6437,5...</td>\n",
       "      <td>8737,8464,8325,7987,7835,7649,7450,7232,7069,6...</td>\n",
       "      <td>['1_0']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>11649</td>\n",
       "      <td>13714</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>13335,11649</td>\n",
       "      <td>13714,13173</td>\n",
       "      <td>['1_0']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>23146</td>\n",
       "      <td>31227</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>23146,24542,24752,25041,25524,25825,26081,2629...</td>\n",
       "      <td>24451,24655,24962,25435,25743,25997,26203,2645...</td>\n",
       "      <td>['1_0']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>31170</td>\n",
       "      <td>33153</td>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>33029,32547,32431,32282,32088,31933,31693,3152...</td>\n",
       "      <td>33153,32670,32459,32347,32195,31998,31813,3160...</td>\n",
       "      <td>['1_0']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27201</th>\n",
       "      <td>207189</td>\n",
       "      <td>5</td>\n",
       "      <td>26959569</td>\n",
       "      <td>26960323</td>\n",
       "      <td>AT5G67600</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67600</td>\n",
       "      <td>26960181,26959937,26959569</td>\n",
       "      <td>26960323,26960093,26959799</td>\n",
       "      <td>['5_3109']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27202</th>\n",
       "      <td>207192</td>\n",
       "      <td>5</td>\n",
       "      <td>26960931</td>\n",
       "      <td>26963638</td>\n",
       "      <td>AT5G67610</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67610</td>\n",
       "      <td>26963509,26963122,26962863,26962648,26962503,2...</td>\n",
       "      <td>26963638,26963335,26963020,26962793,26962562,2...</td>\n",
       "      <td>['5_3109']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27203</th>\n",
       "      <td>207212</td>\n",
       "      <td>5</td>\n",
       "      <td>26964770</td>\n",
       "      <td>26965996</td>\n",
       "      <td>AT5G67620</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67620</td>\n",
       "      <td>26965468,26965198,26964770</td>\n",
       "      <td>26965996,26965379,26965004</td>\n",
       "      <td>['5_3109']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27204</th>\n",
       "      <td>207215</td>\n",
       "      <td>5</td>\n",
       "      <td>26967381</td>\n",
       "      <td>26969394</td>\n",
       "      <td>AT5G67630</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67630</td>\n",
       "      <td>26968558,26967381</td>\n",
       "      <td>26969394,26968195</td>\n",
       "      <td>['5_3109']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27205</th>\n",
       "      <td>207217</td>\n",
       "      <td>5</td>\n",
       "      <td>26969546</td>\n",
       "      <td>26970641</td>\n",
       "      <td>AT5G67640</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67640</td>\n",
       "      <td>26970444,26970312,26969915,26969546</td>\n",
       "      <td>26970641,26970360,26970241,26969670</td>\n",
       "      <td>['5_3109']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27206 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  chrom  gene_start  gene_end gene_symbol  \\\n",
       "0               0      1        3631      5899   AT1G01010   \n",
       "1               6      1        5928      8737   AT1G01020   \n",
       "2              24      1       11649     13714   AT1G01030   \n",
       "3              26      1       23146     31227   AT1G01040   \n",
       "4              66      1       31170     33153   AT1G01050   \n",
       "...           ...    ...         ...       ...         ...   \n",
       "27201      207189      5    26959569  26960323   AT5G67600   \n",
       "27202      207192      5    26960931  26963638   AT5G67610   \n",
       "27203      207212      5    26964770  26965996   AT5G67620   \n",
       "27204      207215      5    26967381  26969394   AT5G67630   \n",
       "27205      207217      5    26969546  26970641   AT5G67640   \n",
       "\n",
       "                   biotype    gene_id  \\\n",
       "0      protein_coding_gene  AT1G01010   \n",
       "1      protein_coding_gene  AT1G01020   \n",
       "2      protein_coding_gene  AT1G01030   \n",
       "3      protein_coding_gene  AT1G01040   \n",
       "4      protein_coding_gene  AT1G01050   \n",
       "...                    ...        ...   \n",
       "27201  protein_coding_gene  AT5G67600   \n",
       "27202  protein_coding_gene  AT5G67610   \n",
       "27203  protein_coding_gene  AT5G67620   \n",
       "27204  protein_coding_gene  AT5G67630   \n",
       "27205  protein_coding_gene  AT5G67640   \n",
       "\n",
       "                                         exon_chromstart  \\\n",
       "0                          3631,3996,4486,4706,5174,5439   \n",
       "1      8571,8417,8236,7942,7762,7564,7384,7157,6437,5...   \n",
       "2                                            13335,11649   \n",
       "3      23146,24542,24752,25041,25524,25825,26081,2629...   \n",
       "4      33029,32547,32431,32282,32088,31933,31693,3152...   \n",
       "...                                                  ...   \n",
       "27201                         26960181,26959937,26959569   \n",
       "27202  26963509,26963122,26962863,26962648,26962503,2...   \n",
       "27203                         26965468,26965198,26964770   \n",
       "27204                                  26968558,26967381   \n",
       "27205                26970444,26970312,26969915,26969546   \n",
       "\n",
       "                                           exon_chromend      blocks  \n",
       "0                          3913,4276,4605,5095,5326,5899     ['1_0']  \n",
       "1      8737,8464,8325,7987,7835,7649,7450,7232,7069,6...     ['1_0']  \n",
       "2                                            13714,13173     ['1_0']  \n",
       "3      24451,24655,24962,25435,25743,25997,26203,2645...     ['1_0']  \n",
       "4      33153,32670,32459,32347,32195,31998,31813,3160...     ['1_0']  \n",
       "...                                                  ...         ...  \n",
       "27201                         26960323,26960093,26959799  ['5_3109']  \n",
       "27202  26963638,26963335,26963020,26962793,26962562,2...  ['5_3109']  \n",
       "27203                         26965996,26965379,26965004  ['5_3109']  \n",
       "27204                                  26969394,26968195  ['5_3109']  \n",
       "27205                26970641,26970360,26970241,26969670  ['5_3109']  \n",
       "\n",
       "[27206 rows x 10 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "454f8eac-aaad-4df4-9969-fb1ab283af76",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "      <th>matching_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>13.132881</td>\n",
       "      <td>1.703152e-32</td>\n",
       "      <td>5.361920e-08</td>\n",
       "      <td>AT1G26680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>12.340603</td>\n",
       "      <td>1.346841e-22</td>\n",
       "      <td>1.942342e-08</td>\n",
       "      <td>AT1G30710,AT1G30720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>64</td>\n",
       "      <td>12</td>\n",
       "      <td>12.798694</td>\n",
       "      <td>2.029109e-12</td>\n",
       "      <td>2.191944e-06</td>\n",
       "      <td>AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>1_4505</td>\n",
       "      <td>96</td>\n",
       "      <td>9</td>\n",
       "      <td>17.562307</td>\n",
       "      <td>5.923819e-07</td>\n",
       "      <td>4.216780e-08</td>\n",
       "      <td>AT1G67560,AT1G67570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>1_4758</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>11.020971</td>\n",
       "      <td>6.073662e-02</td>\n",
       "      <td>2.316135e-06</td>\n",
       "      <td>AT1G70020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>1_485</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>12.925697</td>\n",
       "      <td>5.580758e-03</td>\n",
       "      <td>1.722963e-08</td>\n",
       "      <td>AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4789</th>\n",
       "      <td>1_702</td>\n",
       "      <td>1699</td>\n",
       "      <td>267</td>\n",
       "      <td>12.236921</td>\n",
       "      <td>1.282245e-221</td>\n",
       "      <td>8.324132e-10</td>\n",
       "      <td>AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5244</th>\n",
       "      <td>2_1116</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "      <td>10.800096</td>\n",
       "      <td>1.438617e-40</td>\n",
       "      <td>3.594995e-07</td>\n",
       "      <td>AT2G24230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "      <td>AT2G27030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7059</th>\n",
       "      <td>2_827</td>\n",
       "      <td>141</td>\n",
       "      <td>82</td>\n",
       "      <td>22.186910</td>\n",
       "      <td>1.602818e-124</td>\n",
       "      <td>8.449797e-09</td>\n",
       "      <td>AT2G17380,AT2G17390,AT2G17410,AT2G17420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7214</th>\n",
       "      <td>2_970</td>\n",
       "      <td>1716</td>\n",
       "      <td>55</td>\n",
       "      <td>12.847247</td>\n",
       "      <td>2.080748e-13</td>\n",
       "      <td>2.243761e-13</td>\n",
       "      <td>AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7399</th>\n",
       "      <td>3_1136</td>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>10.000855</td>\n",
       "      <td>1.141790e-12</td>\n",
       "      <td>1.574742e-06</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7614</th>\n",
       "      <td>3_133</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>13.741720</td>\n",
       "      <td>5.782126e-09</td>\n",
       "      <td>5.730219e-09</td>\n",
       "      <td>AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7658</th>\n",
       "      <td>3_137</td>\n",
       "      <td>65</td>\n",
       "      <td>22</td>\n",
       "      <td>13.610011</td>\n",
       "      <td>8.035261e-28</td>\n",
       "      <td>6.002880e-07</td>\n",
       "      <td>AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8884</th>\n",
       "      <td>3_250</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>11.374437</td>\n",
       "      <td>1.254594e-10</td>\n",
       "      <td>8.289905e-07</td>\n",
       "      <td>AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9271</th>\n",
       "      <td>3_2855</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>14.343321</td>\n",
       "      <td>8.559985e-06</td>\n",
       "      <td>7.725087e-09</td>\n",
       "      <td>AT3G50750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9785</th>\n",
       "      <td>3_3325</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>12.259003</td>\n",
       "      <td>4.129632e-01</td>\n",
       "      <td>1.366619e-06</td>\n",
       "      <td>AT3G62360,AT3G62370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13384</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "      <td>AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14638</th>\n",
       "      <td>5_1947</td>\n",
       "      <td>215</td>\n",
       "      <td>59</td>\n",
       "      <td>23.067357</td>\n",
       "      <td>9.586678e-66</td>\n",
       "      <td>1.399768e-06</td>\n",
       "      <td>AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14961</th>\n",
       "      <td>5_2244</td>\n",
       "      <td>39</td>\n",
       "      <td>14</td>\n",
       "      <td>13.843255</td>\n",
       "      <td>1.193361e-18</td>\n",
       "      <td>3.556710e-09</td>\n",
       "      <td>AT5G48560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15696</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>1598</td>\n",
       "      <td>10</td>\n",
       "      <td>18.230969</td>\n",
       "      <td>9.570234e-01</td>\n",
       "      <td>1.710446e-07</td>\n",
       "      <td>AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16421</th>\n",
       "      <td>5_780</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "      <td>11.961668</td>\n",
       "      <td>1.818823e-18</td>\n",
       "      <td>3.341260e-07</td>\n",
       "      <td>AT5G26800,AT5G26805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        block  SNPs  hits          Z  top_candidate_p        Z_pVal  \\\n",
       "350    1_1318    44    22  13.132881     1.703152e-32  5.361920e-08   \n",
       "654    1_1594    30    15  12.340603     1.346841e-22  1.942342e-08   \n",
       "2046   1_2867    64    12  12.798694     2.029109e-12  2.191944e-06   \n",
       "3850   1_4505    96     9  17.562307     5.923819e-07  4.216780e-08   \n",
       "4126   1_4758    40     2  11.020971     6.073662e-02  2.316135e-06   \n",
       "4227    1_485    36     3  12.925697     5.580758e-03  1.722963e-08   \n",
       "4789    1_702  1699   267  12.236921    1.282245e-221  8.324132e-10   \n",
       "5244   2_1116    26    22  10.800096     1.438617e-40  3.594995e-07   \n",
       "5407   2_1265    16    15  11.151806     1.585000e-29  1.569795e-08   \n",
       "7059    2_827   141    82  22.186910    1.602818e-124  8.449797e-09   \n",
       "7214    2_970  1716    55  12.847247     2.080748e-13  2.243761e-13   \n",
       "7399   3_1136    24     9  10.000855     1.141790e-12  1.574742e-06   \n",
       "7614    3_133    40     8  13.741720     5.782126e-09  5.730219e-09   \n",
       "7658    3_137    65    22  13.610011     8.035261e-28  6.002880e-07   \n",
       "8884    3_250    38     9  11.374437     1.254594e-10  8.289905e-07   \n",
       "9271   3_2855    48     6  14.343321     8.559985e-06  7.725087e-09   \n",
       "9785   3_3325    53     1  12.259003     4.129632e-01  1.366619e-06   \n",
       "13384   4_801   225   130  23.960528    7.309601e-196  1.040595e-06   \n",
       "14638  5_1947   215    59  23.067357     9.586678e-66  1.399768e-06   \n",
       "14961  5_2244    39    14  13.843255     1.193361e-18  3.556710e-09   \n",
       "15696  5_2922  1598    10  18.230969     9.570234e-01  1.710446e-07   \n",
       "16421   5_780    40    14  11.961668     1.818823e-18  3.341260e-07   \n",
       "\n",
       "                                        matching_gene_id  \n",
       "350                                            AT1G26680  \n",
       "654                                  AT1G30710,AT1G30720  \n",
       "2046   AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690  \n",
       "3850                                 AT1G67560,AT1G67570  \n",
       "4126                                           AT1G70020  \n",
       "4227   AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...  \n",
       "4789   AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...  \n",
       "5244                                           AT2G24230  \n",
       "5407                                           AT2G27030  \n",
       "7059             AT2G17380,AT2G17390,AT2G17410,AT2G17420  \n",
       "7214   AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...  \n",
       "7399                                                None  \n",
       "7614   AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...  \n",
       "7658   AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...  \n",
       "8884   AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420  \n",
       "9271                                           AT3G50750  \n",
       "9785                                 AT3G62360,AT3G62370  \n",
       "13384  AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150  \n",
       "14638  AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120  \n",
       "14961                                          AT5G48560  \n",
       "15696  AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...  \n",
       "16421                                AT5G26800,AT5G26805  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "011667dc-2626-4e92-92a6-d5a806875b97",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d092e8a7-9959-44e8-8d0c-ab96d2ff075d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to retrieve gene information from Ensembl\n",
    "def get_gene_info(gene_id):\n",
    "    base_url = f\"https://rest.ensembl.org/lookup/symbol/{gene_id}\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    # Send the request\n",
    "    response = requests.get(base_url, headers=headers)\n",
    "    \n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        return response.json()  # Return the JSON response\n",
    "    else:\n",
    "        return None  # Return None if the request failed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "635fa91f-0346-48d5-8dca-13a2abfe86f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_gene_info('AT2G27030')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b0f554c5-fa21-4588-81c7-5d56587502da",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "# Function to retrieve gene information from Ensembl\n",
    "def get_gene_info(gene_id):\n",
    "    base_url = f\"https://rest.ensembl.org/lookup/symbol/arabidopsis_thaliana/{gene_id}\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    # Send the request\n",
    "    response = requests.get(base_url, headers=headers)\n",
    "    \n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        return response.json()  # Return the JSON response\n",
    "    else:\n",
    "        return None  # Return None if the request failed\n",
    "\n",
    "# Initialize an empty list to store gene information and the corresponding block\n",
    "gene_info_list = []\n",
    "\n",
    "# Loop over the rows of sign_wza_lfmm\n",
    "for index, row in sign_wza_lfmm.iterrows():\n",
    "    # Get the block and matching gene IDs from the row\n",
    "    block = row['block']  # Assuming there's a column 'block' in your dataframe\n",
    "    matching_gene_ids = row['matching_gene_id']  # List or comma-separated gene_ids\n",
    "    \n",
    "    # Skip rows where matching_gene_ids is empty or None\n",
    "    if not matching_gene_ids:\n",
    "        continue  # Skip this iteration if matching_gene_ids is None or empty\n",
    "    \n",
    "    # If matching_gene_ids is a string, split by commas into a list\n",
    "    if isinstance(matching_gene_ids, str):\n",
    "        matching_gene_ids = matching_gene_ids.split(',')\n",
    "    \n",
    "    # Loop through each gene_id and retrieve information from the API\n",
    "    for gene_id in matching_gene_ids:\n",
    "        gene_info = get_gene_info(gene_id.strip())  # Ensure any extra spaces are removed\n",
    "        if gene_info:\n",
    "            # Add the gene information and the corresponding block to the list\n",
    "            gene_info_list.append({\n",
    "                \"block\": block,\n",
    "                \"gene_id\": gene_id,\n",
    "                \"gene_name\": gene_info.get('display_name'),\n",
    "                \"start\": gene_info.get('start'),\n",
    "                \"end\": gene_info.get('end'),\n",
    "                \"biotype\": gene_info.get('biotype'),\n",
    "                \"description\": gene_info.get('description')\n",
    "            })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "def355f5-e6df-48e6-89d2-24a9435ddcd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "      <th>matching_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>13.132881</td>\n",
       "      <td>1.703152e-32</td>\n",
       "      <td>5.361920e-08</td>\n",
       "      <td>AT1G26680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>12.340603</td>\n",
       "      <td>1.346841e-22</td>\n",
       "      <td>1.942342e-08</td>\n",
       "      <td>AT1G30710,AT1G30720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>64</td>\n",
       "      <td>12</td>\n",
       "      <td>12.798694</td>\n",
       "      <td>2.029109e-12</td>\n",
       "      <td>2.191944e-06</td>\n",
       "      <td>AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>1_4505</td>\n",
       "      <td>96</td>\n",
       "      <td>9</td>\n",
       "      <td>17.562307</td>\n",
       "      <td>5.923819e-07</td>\n",
       "      <td>4.216780e-08</td>\n",
       "      <td>AT1G67560,AT1G67570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>1_4758</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>11.020971</td>\n",
       "      <td>6.073662e-02</td>\n",
       "      <td>2.316135e-06</td>\n",
       "      <td>AT1G70020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>1_485</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>12.925697</td>\n",
       "      <td>5.580758e-03</td>\n",
       "      <td>1.722963e-08</td>\n",
       "      <td>AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4789</th>\n",
       "      <td>1_702</td>\n",
       "      <td>1699</td>\n",
       "      <td>267</td>\n",
       "      <td>12.236921</td>\n",
       "      <td>1.282245e-221</td>\n",
       "      <td>8.324132e-10</td>\n",
       "      <td>AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5244</th>\n",
       "      <td>2_1116</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "      <td>10.800096</td>\n",
       "      <td>1.438617e-40</td>\n",
       "      <td>3.594995e-07</td>\n",
       "      <td>AT2G24230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "      <td>AT2G27030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7059</th>\n",
       "      <td>2_827</td>\n",
       "      <td>141</td>\n",
       "      <td>82</td>\n",
       "      <td>22.186910</td>\n",
       "      <td>1.602818e-124</td>\n",
       "      <td>8.449797e-09</td>\n",
       "      <td>AT2G17380,AT2G17390,AT2G17410,AT2G17420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7214</th>\n",
       "      <td>2_970</td>\n",
       "      <td>1716</td>\n",
       "      <td>55</td>\n",
       "      <td>12.847247</td>\n",
       "      <td>2.080748e-13</td>\n",
       "      <td>2.243761e-13</td>\n",
       "      <td>AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7399</th>\n",
       "      <td>3_1136</td>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>10.000855</td>\n",
       "      <td>1.141790e-12</td>\n",
       "      <td>1.574742e-06</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7614</th>\n",
       "      <td>3_133</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>13.741720</td>\n",
       "      <td>5.782126e-09</td>\n",
       "      <td>5.730219e-09</td>\n",
       "      <td>AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7658</th>\n",
       "      <td>3_137</td>\n",
       "      <td>65</td>\n",
       "      <td>22</td>\n",
       "      <td>13.610011</td>\n",
       "      <td>8.035261e-28</td>\n",
       "      <td>6.002880e-07</td>\n",
       "      <td>AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8884</th>\n",
       "      <td>3_250</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>11.374437</td>\n",
       "      <td>1.254594e-10</td>\n",
       "      <td>8.289905e-07</td>\n",
       "      <td>AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9271</th>\n",
       "      <td>3_2855</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>14.343321</td>\n",
       "      <td>8.559985e-06</td>\n",
       "      <td>7.725087e-09</td>\n",
       "      <td>AT3G50750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9785</th>\n",
       "      <td>3_3325</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>12.259003</td>\n",
       "      <td>4.129632e-01</td>\n",
       "      <td>1.366619e-06</td>\n",
       "      <td>AT3G62360,AT3G62370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13384</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "      <td>AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14638</th>\n",
       "      <td>5_1947</td>\n",
       "      <td>215</td>\n",
       "      <td>59</td>\n",
       "      <td>23.067357</td>\n",
       "      <td>9.586678e-66</td>\n",
       "      <td>1.399768e-06</td>\n",
       "      <td>AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14961</th>\n",
       "      <td>5_2244</td>\n",
       "      <td>39</td>\n",
       "      <td>14</td>\n",
       "      <td>13.843255</td>\n",
       "      <td>1.193361e-18</td>\n",
       "      <td>3.556710e-09</td>\n",
       "      <td>AT5G48560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15696</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>1598</td>\n",
       "      <td>10</td>\n",
       "      <td>18.230969</td>\n",
       "      <td>9.570234e-01</td>\n",
       "      <td>1.710446e-07</td>\n",
       "      <td>AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16421</th>\n",
       "      <td>5_780</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "      <td>11.961668</td>\n",
       "      <td>1.818823e-18</td>\n",
       "      <td>3.341260e-07</td>\n",
       "      <td>AT5G26800,AT5G26805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        block  SNPs  hits          Z  top_candidate_p        Z_pVal  \\\n",
       "350    1_1318    44    22  13.132881     1.703152e-32  5.361920e-08   \n",
       "654    1_1594    30    15  12.340603     1.346841e-22  1.942342e-08   \n",
       "2046   1_2867    64    12  12.798694     2.029109e-12  2.191944e-06   \n",
       "3850   1_4505    96     9  17.562307     5.923819e-07  4.216780e-08   \n",
       "4126   1_4758    40     2  11.020971     6.073662e-02  2.316135e-06   \n",
       "4227    1_485    36     3  12.925697     5.580758e-03  1.722963e-08   \n",
       "4789    1_702  1699   267  12.236921    1.282245e-221  8.324132e-10   \n",
       "5244   2_1116    26    22  10.800096     1.438617e-40  3.594995e-07   \n",
       "5407   2_1265    16    15  11.151806     1.585000e-29  1.569795e-08   \n",
       "7059    2_827   141    82  22.186910    1.602818e-124  8.449797e-09   \n",
       "7214    2_970  1716    55  12.847247     2.080748e-13  2.243761e-13   \n",
       "7399   3_1136    24     9  10.000855     1.141790e-12  1.574742e-06   \n",
       "7614    3_133    40     8  13.741720     5.782126e-09  5.730219e-09   \n",
       "7658    3_137    65    22  13.610011     8.035261e-28  6.002880e-07   \n",
       "8884    3_250    38     9  11.374437     1.254594e-10  8.289905e-07   \n",
       "9271   3_2855    48     6  14.343321     8.559985e-06  7.725087e-09   \n",
       "9785   3_3325    53     1  12.259003     4.129632e-01  1.366619e-06   \n",
       "13384   4_801   225   130  23.960528    7.309601e-196  1.040595e-06   \n",
       "14638  5_1947   215    59  23.067357     9.586678e-66  1.399768e-06   \n",
       "14961  5_2244    39    14  13.843255     1.193361e-18  3.556710e-09   \n",
       "15696  5_2922  1598    10  18.230969     9.570234e-01  1.710446e-07   \n",
       "16421   5_780    40    14  11.961668     1.818823e-18  3.341260e-07   \n",
       "\n",
       "                                        matching_gene_id  \n",
       "350                                            AT1G26680  \n",
       "654                                  AT1G30710,AT1G30720  \n",
       "2046   AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690  \n",
       "3850                                 AT1G67560,AT1G67570  \n",
       "4126                                           AT1G70020  \n",
       "4227   AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...  \n",
       "4789   AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...  \n",
       "5244                                           AT2G24230  \n",
       "5407                                           AT2G27030  \n",
       "7059             AT2G17380,AT2G17390,AT2G17410,AT2G17420  \n",
       "7214   AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...  \n",
       "7399                                                None  \n",
       "7614   AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...  \n",
       "7658   AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...  \n",
       "8884   AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420  \n",
       "9271                                           AT3G50750  \n",
       "9785                                 AT3G62360,AT3G62370  \n",
       "13384  AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150  \n",
       "14638  AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120  \n",
       "14961                                          AT5G48560  \n",
       "15696  AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...  \n",
       "16421                                AT5G26800,AT5G26805  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "60b711e8-0dac-4054-b5da-5a3f6ff426bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the gene information list to a DataFrame for further analysis\n",
    "df_gene_info_ensembl = pd.DataFrame(gene_info_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6df0e0ce-8fc8-46ae-b3de-545caaafbc6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>biotype</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>AT1G26680</td>\n",
       "      <td>AT1G26680</td>\n",
       "      <td>9219460</td>\n",
       "      <td>9223795</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcriptional factor B3 family protein [Sour...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>AT1G30710</td>\n",
       "      <td>AT1G30710</td>\n",
       "      <td>10895252</td>\n",
       "      <td>10897127</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FAD-binding Berberine family protein [Source:N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>AT1G30720</td>\n",
       "      <td>AT1G30720</td>\n",
       "      <td>10897925</td>\n",
       "      <td>10899975</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FAD-binding Berberine family protein [Source:N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>AT1G55660</td>\n",
       "      <td>AT1G55660</td>\n",
       "      <td>20800694</td>\n",
       "      <td>20802377</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FBD, F-box and Leucine Rich Repeat domains con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>AT1G55675</td>\n",
       "      <td>AT1G55675</td>\n",
       "      <td>20803661</td>\n",
       "      <td>20804574</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transmembrane protein [Source:NCBI gene (forme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60335</td>\n",
       "      <td>AT5G60335</td>\n",
       "      <td>24272995</td>\n",
       "      <td>24274860</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>hydroxyacyl-thioester dehydratase type-like pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60350</td>\n",
       "      <td>AT5G60350</td>\n",
       "      <td>24277730</td>\n",
       "      <td>24279147</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60370</td>\n",
       "      <td>AT5G60370</td>\n",
       "      <td>24282666</td>\n",
       "      <td>24285061</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonuclease V-like protein [Source:NCBI gene (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60380</td>\n",
       "      <td>AT5G60380</td>\n",
       "      <td>24285920</td>\n",
       "      <td>24287808</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transmembrane protein, putative (DUF239) [Sour...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>5_780</td>\n",
       "      <td>AT5G26805</td>\n",
       "      <td>AT5G26805</td>\n",
       "      <td>9427543</td>\n",
       "      <td>9428067</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>B3 domain protein [Source:NCBI gene (formerly ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      block    gene_id  gene_name     start       end         biotype  \\\n",
       "0    1_1318  AT1G26680  AT1G26680   9219460   9223795  protein_coding   \n",
       "1    1_1594  AT1G30710  AT1G30710  10895252  10897127  protein_coding   \n",
       "2    1_1594  AT1G30720  AT1G30720  10897925  10899975  protein_coding   \n",
       "3    1_2867  AT1G55660  AT1G55660  20800694  20802377  protein_coding   \n",
       "4    1_2867  AT1G55675  AT1G55675  20803661  20804574  protein_coding   \n",
       "..      ...        ...        ...       ...       ...             ...   \n",
       "116  5_2922  AT5G60335  AT5G60335  24272995  24274860  protein_coding   \n",
       "117  5_2922  AT5G60350  AT5G60350  24277730  24279147  protein_coding   \n",
       "118  5_2922  AT5G60370  AT5G60370  24282666  24285061  protein_coding   \n",
       "119  5_2922  AT5G60380  AT5G60380  24285920  24287808  protein_coding   \n",
       "120   5_780  AT5G26805  AT5G26805   9427543   9428067  protein_coding   \n",
       "\n",
       "                                           description  \n",
       "0    transcriptional factor B3 family protein [Sour...  \n",
       "1    FAD-binding Berberine family protein [Source:N...  \n",
       "2    FAD-binding Berberine family protein [Source:N...  \n",
       "3    FBD, F-box and Leucine Rich Repeat domains con...  \n",
       "4    transmembrane protein [Source:NCBI gene (forme...  \n",
       "..                                                 ...  \n",
       "116  hydroxyacyl-thioester dehydratase type-like pr...  \n",
       "117                                               None  \n",
       "118  exonuclease V-like protein [Source:NCBI gene (...  \n",
       "119  transmembrane protein, putative (DUF239) [Sour...  \n",
       "120  B3 domain protein [Source:NCBI gene (formerly ...  \n",
       "\n",
       "[121 rows x 7 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_info_ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d13c6b11-ab25-4ad7-a9d5-206aea61fa55",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8421daf6-1540-4618-a5df-65a599d1e314",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2def42d3-9024-47ba-a5c4-675cdd10a2f9",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'biomart'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mbiomart\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BiomartServer\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'biomart'"
     ]
    }
   ],
   "source": [
    "from biomart import BiomartServer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bb0e7b4d-3475-48f3-9908-16918988ef30",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'biomart'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Example using BioMart to map gene IDs between Ensembl and RefSeq\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mbiomart\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Connect to the Ensembl Plants BioMart database\u001b[39;00m\n\u001b[1;32m      5\u001b[0m server \u001b[38;5;241m=\u001b[39m biomart\u001b[38;5;241m.\u001b[39mBioMartServer(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttp://plants.ensembl.org/biomart\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'biomart'"
     ]
    }
   ],
   "source": [
    "# Example using BioMart to map gene IDs between Ensembl and RefSeq\n",
    "import biomart\n",
    "\n",
    "# Connect to the Ensembl Plants BioMart database\n",
    "server = biomart.BioMartServer(\"http://plants.ensembl.org/biomart\")\n",
    "\n",
    "# Use the appropriate dataset for Arabidopsis thaliana\n",
    "dataset = server.datasets['athaliana_eg_gene']\n",
    "\n",
    "# Query to map TAIR ID to RefSeq ID\n",
    "response = dataset.search({\n",
    "    'filters': {\n",
    "        'ensembl_gene_id': ['AT2G27030']  # TAIR ID or Ensembl ID\n",
    "    },\n",
    "    'attributes': [\n",
    "        'ensembl_gene_id', 'external_gene_name', 'refseq_mrna', 'refseq_peptide'\n",
    "    ]\n",
    "})\n",
    "\n",
    "print(response.text)  # This will show the mapped gene IDs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ad29040-9856-4f91-9140-43e7220c7c7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6b1cbe5-6920-40d1-ac38-d749eab84700",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "315490ae-82dd-4ad6-ac6b-dcdd77375eab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d2f70e0-248f-4705-9572-f79fb844309e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"https://rest.ensembl.org/lookup/symbol/arabidopsis_thaliana/AT2G27030\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "0278579f-c29f-4737-8aed-8d1ac2c34af9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_info(gene_id):\n",
    "    base_url = f\"https://rest.ensembl.org/lookup/symbol/arabidopsis_thaliana/{gene_id}\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    \n",
    "    # Send the request\n",
    "    response = requests.get(base_url, headers=headers)\n",
    "    \n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        return response.json()  # Return the JSON response\n",
    "    else:\n",
    "        return None  # Return None if the request failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfc7974e-f5e3-4be7-8eb4-8e73e5c638b5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "475722c9-7895-4936-80bc-0a3876de9e45",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (440813899.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[87], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    curl -X GET \"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/gene/id/AT2G27030/download_summary\" \\\u001b[0m\n\u001b[0m            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "curl -X GET \"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/gene/id/AT2G27030/download_summary\" \\\n",
    " -H \"accept: application/json\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "38bedbd7-5a08-42ec-892f-4329a60fd294",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id = 'AT2G27030'\n",
    "gene_info = get_gene_info(gene_id.strip())  # Ensure any extra spaces are removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "d87bd604-32ee-4466-9192-b4883e6598b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "print(gene_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "6666f73c-c0fa-4056-b61e-1e4a8f4dccf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_gene_info('AT2G27030')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "89e6e4c7-fab6-4e7d-9ed1-3ea1ee5c1967",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_info = pd.DataFrame(gene_info_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "f8351e54-3546-445f-9ef5-a16f275ff00a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>biotype</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>exon_chromstart</th>\n",
       "      <th>exon_chromend</th>\n",
       "      <th>blocks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9060</th>\n",
       "      <td>69982</td>\n",
       "      <td>2</td>\n",
       "      <td>11532004</td>\n",
       "      <td>11534333</td>\n",
       "      <td>AT2G27030</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT2G27030</td>\n",
       "      <td>11532004,11532687,11532007,11532004,11532687,1...</td>\n",
       "      <td>11532144,11533276,11534272,11532144,11533056,1...</td>\n",
       "      <td>['2_1264', '2_1265', '2_1266']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0  chrom  gene_start  gene_end gene_symbol  \\\n",
       "9060       69982      2    11532004  11534333   AT2G27030   \n",
       "\n",
       "                  biotype    gene_id  \\\n",
       "9060  protein_coding_gene  AT2G27030   \n",
       "\n",
       "                                        exon_chromstart  \\\n",
       "9060  11532004,11532687,11532007,11532004,11532687,1...   \n",
       "\n",
       "                                          exon_chromend  \\\n",
       "9060  11532144,11533276,11534272,11532144,11533056,1...   \n",
       "\n",
       "                              blocks  \n",
       "9060  ['2_1264', '2_1265', '2_1266']  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair[tair['gene_id']== 'AT2G27030']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abc0decf-e064-43d1-adb1-ec0ecd5c3661",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "442e405b-7586-4769-a222-8e7eafce4645",
   "metadata": {},
   "outputs": [],
   "source": [
    "AT2G27030"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "7bb77a0a-7427-4d93-b20b-556f61eeb2f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>biotype</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [block, gene_id, gene_name, start, end, biotype, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_info[df_gene_info['gene_id'] == 'AT2G27030']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "17599175-60f0-423e-88cc-416a3e585605",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_info.to_csv('blocks_annotation_wza_results_lfmm_bio1.csv',index =None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8c9049e5-3ee3-459f-88f0-32293d5cc77f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1_10898797'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_blocks['1_1594'][-1] # 1_10895628  1_10898797"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "99d7dc14-3b40-4723-ac9c-3dbc7ba2cc2a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "fa4d83f7-ad16-4a7c-96bf-495b2df3d6e8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error fetching data for gene AT1G26680: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G30710: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G30720: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G55660: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G55675: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G55680: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G55690: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G67560: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT1G67570: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G70020: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G70590: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G70600: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G70620: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G70630: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G70640: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G70650: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17500: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17510: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17520: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17540: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17545: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17590: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT1G17600: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17610: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17615: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17620: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17630: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17640: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17665: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17680: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17690: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT1G17700: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT1G17744: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17780: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17790: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17800: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT1G17820: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17830: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17850: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17860: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17880: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT1G17890: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT1G17910: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17930: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17940: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT1G17960: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G24230: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G17410: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22480: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT2G22510: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22520: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22530: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22560: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT2G22590: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22600: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22610: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22620: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22650: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22660: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22680: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT2G22690: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22720: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22730: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22750: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22760: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22790: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22795: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22802: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT2G22805: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G05327: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28580: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28590: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28600: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28610: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28620: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28630: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28640: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G05380: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT3G05390: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28917: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT3G28918: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28920: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT3G28925: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28940: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28950: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G28956: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G07350: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G47400: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G47410: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G50750: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT3G62360: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT3G62370: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT4G12130: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT4G12140: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT4G12150: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G44080: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G44090: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G44100: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT5G48560: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59830: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59840: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59845: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59850: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59860: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59900: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59910: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT5G59930: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59940: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59950: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59960: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59970: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G59990: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60000: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60010: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60030: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60050: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60060: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60070: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60080: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60090: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60120: 'Entrezgene_summary'\n",
      "Error fetching data for gene AT5G60130: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60140: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60142: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60150: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60160: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60170: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60190: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60240: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60250: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60260: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60270: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60280: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60290: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60310: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60320: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60330: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60335: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60350: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60370: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G60380: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G26800: 'Gene-ref_locus'\n",
      "Error fetching data for gene AT5G26805: 'Gene-ref_locus'\n",
      "     block    gene_id  gene_name  \\\n",
      "0   1_2867  AT1G55670       PSAG   \n",
      "1    1_485  AT1G11840       GLX1   \n",
      "2    1_485  AT1G70580      AOAT2   \n",
      "3    1_485  AT1G70610     ABCB26   \n",
      "4    1_485  AT1G70660       MMZ2   \n",
      "..     ...        ...        ...   \n",
      "59  5_2922  AT5G60220       TET4   \n",
      "60  5_2922  AT5G60230       SEN2   \n",
      "61  5_2922  AT5G60300  LecRK-I.9   \n",
      "62  5_2922  AT5G60340       AAK6   \n",
      "63  5_2922  AT5G60360        ALP   \n",
      "\n",
      "                                          description  \n",
      "0   Encodes subunit G of photosystem I, an 11-kDa ...  \n",
      "1              Encodes a glyoxalase I homolog ATGLX1.  \n",
      "2   Encodes a protein with glyoxylate aminotransfe...  \n",
      "3                             member of TAP subfamily  \n",
      "4   MMZ2/UEV1B encodes a protein that may play a r...  \n",
      "..                                                ...  \n",
      "59                       Member of TETRASPANIN family  \n",
      "60     putative subunit of tRNA splicing endonuclease  \n",
      "61  Lectin Receptor Kinase involved in protein-pro...  \n",
      "62  Encodes a nuclear adenylate kinase that intera...  \n",
      "63    Encodes a senescence-associated thiol protease.  \n",
      "\n",
      "[64 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "# Provide your email, NCBI requires this for queries\n",
    "Entrez.email = \"tbellg@berkeley.edu\"\n",
    "\n",
    "# Function to retrieve gene information using NCBI Entrez API\n",
    "def get_gene_info_ncbi(gene_id):\n",
    "    try:\n",
    "        # Search for the gene in the NCBI database\n",
    "        search_handle = Entrez.esearch(db=\"gene\", term=f\"{gene_id}[Gene]\", retmax=1)\n",
    "        search_results = Entrez.read(search_handle)\n",
    "        search_handle.close()\n",
    "        \n",
    "        if search_results[\"IdList\"]:\n",
    "            gene_ncbi_id = search_results[\"IdList\"][0]  # Get the first matching gene ID\n",
    "            \n",
    "            # Fetch detailed gene information using the gene ID\n",
    "            fetch_handle = Entrez.efetch(db=\"gene\", id=gene_ncbi_id, retmode=\"xml\")\n",
    "            gene_data = Entrez.read(fetch_handle)\n",
    "            fetch_handle.close()\n",
    "            \n",
    "            # Extract relevant information from the gene data\n",
    "            gene_info = gene_data[0]\n",
    "            gene_name = gene_info['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']\n",
    "            gene_desc = gene_info['Entrezgene_summary']\n",
    "            return {\n",
    "                \"gene_id\": gene_id,\n",
    "                \"gene_name\": gene_name,\n",
    "                \"description\": gene_desc\n",
    "            }\n",
    "        else:\n",
    "            return None\n",
    "    except Exception as e:\n",
    "        print(f\"Error fetching data for gene {gene_id}: {e}\")\n",
    "        return None\n",
    "\n",
    "# Initialize an empty list to store gene information and the corresponding block\n",
    "gene_info_list = []\n",
    "\n",
    "# Loop over the rows of sign_wza_lfmm\n",
    "for index, row in sign_wza_lfmm.iterrows():\n",
    "    # Get the block and matching gene IDs from the row\n",
    "    block = row['block']  # Assuming there's a column 'block' in your dataframe\n",
    "    matching_gene_ids = row['matching_gene_id']  # List or comma-separated gene_ids\n",
    "    \n",
    "    # Skip rows where matching_gene_ids is empty or None\n",
    "    if not matching_gene_ids:\n",
    "        continue  # Skip this iteration if matching_gene_ids is None or empty\n",
    "    \n",
    "    # If matching_gene_ids is a string, split by commas into a list\n",
    "    if isinstance(matching_gene_ids, str):\n",
    "        matching_gene_ids = matching_gene_ids.split(',')\n",
    "    \n",
    "    # Loop through each gene_id and retrieve information from the API\n",
    "    for gene_id in matching_gene_ids:\n",
    "        gene_info = get_gene_info_ncbi(gene_id.strip())  # Ensure any extra spaces are removed\n",
    "        if gene_info:\n",
    "            # Add the gene information and the corresponding block to the list\n",
    "            gene_info_list.append({\n",
    "                \"block\": block,\n",
    "                \"gene_id\": gene_info.get('gene_id'),\n",
    "                \"gene_name\": gene_info.get('gene_name'),\n",
    "                \"description\": gene_info.get('description')\n",
    "            })\n",
    "\n",
    "# Convert the gene information list to a DataFrame for further analysis\n",
    "df_gene_info = pd.DataFrame(gene_info_list)\n",
    "\n",
    "# Display the gene information DataFrame\n",
    "print(df_gene_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "6136f230-46e5-4e3a-8017-898297ff4f36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "      <th>matching_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>13.132881</td>\n",
       "      <td>1.703152e-32</td>\n",
       "      <td>5.361920e-08</td>\n",
       "      <td>AT1G26680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>12.340603</td>\n",
       "      <td>1.346841e-22</td>\n",
       "      <td>1.942342e-08</td>\n",
       "      <td>AT1G30710,AT1G30720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>64</td>\n",
       "      <td>12</td>\n",
       "      <td>12.798694</td>\n",
       "      <td>2.029109e-12</td>\n",
       "      <td>2.191944e-06</td>\n",
       "      <td>AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>1_4505</td>\n",
       "      <td>96</td>\n",
       "      <td>9</td>\n",
       "      <td>17.562307</td>\n",
       "      <td>5.923819e-07</td>\n",
       "      <td>4.216780e-08</td>\n",
       "      <td>AT1G67560,AT1G67570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>1_4758</td>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>11.020971</td>\n",
       "      <td>6.073662e-02</td>\n",
       "      <td>2.316135e-06</td>\n",
       "      <td>AT1G70020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>1_485</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>12.925697</td>\n",
       "      <td>5.580758e-03</td>\n",
       "      <td>1.722963e-08</td>\n",
       "      <td>AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4789</th>\n",
       "      <td>1_702</td>\n",
       "      <td>1699</td>\n",
       "      <td>267</td>\n",
       "      <td>12.236921</td>\n",
       "      <td>1.282245e-221</td>\n",
       "      <td>8.324132e-10</td>\n",
       "      <td>AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5244</th>\n",
       "      <td>2_1116</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "      <td>10.800096</td>\n",
       "      <td>1.438617e-40</td>\n",
       "      <td>3.594995e-07</td>\n",
       "      <td>AT2G24230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "      <td>AT2G27030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7059</th>\n",
       "      <td>2_827</td>\n",
       "      <td>141</td>\n",
       "      <td>82</td>\n",
       "      <td>22.186910</td>\n",
       "      <td>1.602818e-124</td>\n",
       "      <td>8.449797e-09</td>\n",
       "      <td>AT2G17380,AT2G17390,AT2G17410,AT2G17420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7214</th>\n",
       "      <td>2_970</td>\n",
       "      <td>1716</td>\n",
       "      <td>55</td>\n",
       "      <td>12.847247</td>\n",
       "      <td>2.080748e-13</td>\n",
       "      <td>2.243761e-13</td>\n",
       "      <td>AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7399</th>\n",
       "      <td>3_1136</td>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>10.000855</td>\n",
       "      <td>1.141790e-12</td>\n",
       "      <td>1.574742e-06</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7614</th>\n",
       "      <td>3_133</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>13.741720</td>\n",
       "      <td>5.782126e-09</td>\n",
       "      <td>5.730219e-09</td>\n",
       "      <td>AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7658</th>\n",
       "      <td>3_137</td>\n",
       "      <td>65</td>\n",
       "      <td>22</td>\n",
       "      <td>13.610011</td>\n",
       "      <td>8.035261e-28</td>\n",
       "      <td>6.002880e-07</td>\n",
       "      <td>AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8884</th>\n",
       "      <td>3_250</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>11.374437</td>\n",
       "      <td>1.254594e-10</td>\n",
       "      <td>8.289905e-07</td>\n",
       "      <td>AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9271</th>\n",
       "      <td>3_2855</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>14.343321</td>\n",
       "      <td>8.559985e-06</td>\n",
       "      <td>7.725087e-09</td>\n",
       "      <td>AT3G50750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9785</th>\n",
       "      <td>3_3325</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>12.259003</td>\n",
       "      <td>4.129632e-01</td>\n",
       "      <td>1.366619e-06</td>\n",
       "      <td>AT3G62360,AT3G62370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13384</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "      <td>AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14638</th>\n",
       "      <td>5_1947</td>\n",
       "      <td>215</td>\n",
       "      <td>59</td>\n",
       "      <td>23.067357</td>\n",
       "      <td>9.586678e-66</td>\n",
       "      <td>1.399768e-06</td>\n",
       "      <td>AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14961</th>\n",
       "      <td>5_2244</td>\n",
       "      <td>39</td>\n",
       "      <td>14</td>\n",
       "      <td>13.843255</td>\n",
       "      <td>1.193361e-18</td>\n",
       "      <td>3.556710e-09</td>\n",
       "      <td>AT5G48560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15696</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>1598</td>\n",
       "      <td>10</td>\n",
       "      <td>18.230969</td>\n",
       "      <td>9.570234e-01</td>\n",
       "      <td>1.710446e-07</td>\n",
       "      <td>AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16421</th>\n",
       "      <td>5_780</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "      <td>11.961668</td>\n",
       "      <td>1.818823e-18</td>\n",
       "      <td>3.341260e-07</td>\n",
       "      <td>AT5G26800,AT5G26805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        block  SNPs  hits          Z  top_candidate_p        Z_pVal  \\\n",
       "350    1_1318    44    22  13.132881     1.703152e-32  5.361920e-08   \n",
       "654    1_1594    30    15  12.340603     1.346841e-22  1.942342e-08   \n",
       "2046   1_2867    64    12  12.798694     2.029109e-12  2.191944e-06   \n",
       "3850   1_4505    96     9  17.562307     5.923819e-07  4.216780e-08   \n",
       "4126   1_4758    40     2  11.020971     6.073662e-02  2.316135e-06   \n",
       "4227    1_485    36     3  12.925697     5.580758e-03  1.722963e-08   \n",
       "4789    1_702  1699   267  12.236921    1.282245e-221  8.324132e-10   \n",
       "5244   2_1116    26    22  10.800096     1.438617e-40  3.594995e-07   \n",
       "5407   2_1265    16    15  11.151806     1.585000e-29  1.569795e-08   \n",
       "7059    2_827   141    82  22.186910    1.602818e-124  8.449797e-09   \n",
       "7214    2_970  1716    55  12.847247     2.080748e-13  2.243761e-13   \n",
       "7399   3_1136    24     9  10.000855     1.141790e-12  1.574742e-06   \n",
       "7614    3_133    40     8  13.741720     5.782126e-09  5.730219e-09   \n",
       "7658    3_137    65    22  13.610011     8.035261e-28  6.002880e-07   \n",
       "8884    3_250    38     9  11.374437     1.254594e-10  8.289905e-07   \n",
       "9271   3_2855    48     6  14.343321     8.559985e-06  7.725087e-09   \n",
       "9785   3_3325    53     1  12.259003     4.129632e-01  1.366619e-06   \n",
       "13384   4_801   225   130  23.960528    7.309601e-196  1.040595e-06   \n",
       "14638  5_1947   215    59  23.067357     9.586678e-66  1.399768e-06   \n",
       "14961  5_2244    39    14  13.843255     1.193361e-18  3.556710e-09   \n",
       "15696  5_2922  1598    10  18.230969     9.570234e-01  1.710446e-07   \n",
       "16421   5_780    40    14  11.961668     1.818823e-18  3.341260e-07   \n",
       "\n",
       "                                        matching_gene_id  \n",
       "350                                            AT1G26680  \n",
       "654                                  AT1G30710,AT1G30720  \n",
       "2046   AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690  \n",
       "3850                                 AT1G67560,AT1G67570  \n",
       "4126                                           AT1G70020  \n",
       "4227   AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70...  \n",
       "4789   AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17...  \n",
       "5244                                           AT2G24230  \n",
       "5407                                           AT2G27030  \n",
       "7059             AT2G17380,AT2G17390,AT2G17410,AT2G17420  \n",
       "7214   AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22...  \n",
       "7399                                                None  \n",
       "7614   AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28...  \n",
       "7658   AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28...  \n",
       "8884   AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420  \n",
       "9271                                           AT3G50750  \n",
       "9785                                 AT3G62360,AT3G62370  \n",
       "13384  AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150  \n",
       "14638  AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120  \n",
       "14961                                          AT5G48560  \n",
       "15696  AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59...  \n",
       "16421                                AT5G26800,AT5G26805  "
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "0ffc5b22-7f42-4200-a88b-86eb263458a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>AT1G55670</td>\n",
       "      <td>PSAG</td>\n",
       "      <td>Encodes subunit G of photosystem I, an 11-kDa ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G11840</td>\n",
       "      <td>GLX1</td>\n",
       "      <td>Encodes a glyoxalase I homolog ATGLX1.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70580</td>\n",
       "      <td>AOAT2</td>\n",
       "      <td>Encodes a protein with glyoxylate aminotransfe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70610</td>\n",
       "      <td>ABCB26</td>\n",
       "      <td>member of TAP subfamily</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70660</td>\n",
       "      <td>MMZ2</td>\n",
       "      <td>MMZ2/UEV1B encodes a protein that may play a r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60220</td>\n",
       "      <td>TET4</td>\n",
       "      <td>Member of TETRASPANIN family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60230</td>\n",
       "      <td>SEN2</td>\n",
       "      <td>putative subunit of tRNA splicing endonuclease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60300</td>\n",
       "      <td>LecRK-I.9</td>\n",
       "      <td>Lectin Receptor Kinase involved in protein-pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60340</td>\n",
       "      <td>AAK6</td>\n",
       "      <td>Encodes a nuclear adenylate kinase that intera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60360</td>\n",
       "      <td>ALP</td>\n",
       "      <td>Encodes a senescence-associated thiol protease.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     block    gene_id  gene_name  \\\n",
       "0   1_2867  AT1G55670       PSAG   \n",
       "1    1_485  AT1G11840       GLX1   \n",
       "2    1_485  AT1G70580      AOAT2   \n",
       "3    1_485  AT1G70610     ABCB26   \n",
       "4    1_485  AT1G70660       MMZ2   \n",
       "..     ...        ...        ...   \n",
       "59  5_2922  AT5G60220       TET4   \n",
       "60  5_2922  AT5G60230       SEN2   \n",
       "61  5_2922  AT5G60300  LecRK-I.9   \n",
       "62  5_2922  AT5G60340       AAK6   \n",
       "63  5_2922  AT5G60360        ALP   \n",
       "\n",
       "                                          description  \n",
       "0   Encodes subunit G of photosystem I, an 11-kDa ...  \n",
       "1              Encodes a glyoxalase I homolog ATGLX1.  \n",
       "2   Encodes a protein with glyoxylate aminotransfe...  \n",
       "3                             member of TAP subfamily  \n",
       "4   MMZ2/UEV1B encodes a protein that may play a r...  \n",
       "..                                                ...  \n",
       "59                       Member of TETRASPANIN family  \n",
       "60     putative subunit of tRNA splicing endonuclease  \n",
       "61  Lectin Receptor Kinase involved in protein-pro...  \n",
       "62  Encodes a nuclear adenylate kinase that intera...  \n",
       "63    Encodes a senescence-associated thiol protease.  \n",
       "\n",
       "[64 rows x 4 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_info_ncbi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "8bef99b5-92e6-4dbe-8771-1446bca6bbe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_info_ncbi = df_gene_info.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e454473f-e331-405e-8329-9857fe9ea84c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4804d55-fcdb-480e-880f-5122d50c09fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "5f47f7dc-2445-49c8-96b0-9069d14bc669",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_info_ncbi.to_csv('df_lfmm_wza_block_gene_info_ncbi.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de495784-39f1-4ad5-9b4f-376ae5a32a4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "72404332-0180-426c-8914-8892ecbe2eaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1_2867', '1_485', '1_702', '2_1265', '2_827', '2_970', '3_137',\n",
       "       '3_250', '4_801', '5_1947', '5_2922'], dtype=object)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_info_ncbi['block'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "f6c45fea-5104-4793-aae0-8e50f4818e19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1_485',\n",
       " '1_702',\n",
       " '1_2867',\n",
       " '2_827',\n",
       " '2_970',\n",
       " '2_1265',\n",
       " '3_137',\n",
       " '3_250',\n",
       " '4_801',\n",
       " '5_1947',\n",
       " '5_2922']"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['1_485', '1_702', '1_2867', '2_827', '2_970','2_1265', '3_137',\n",
    "       '3_250', '4_801', '5_1947', '5_2922']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "dc0b9f63-86b8-41d9-af8c-69df2a48358f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_485\n",
      "1_3995653\n",
      "1_3997916\n",
      "1_702\n",
      "1_6018870\n",
      "1_6181134\n",
      "1_2867\n",
      "1_20800345\n",
      "1_20814896\n",
      "2_827\n",
      "2_7553538\n",
      "2_7567816\n",
      "2_970\n",
      "2_9540996\n",
      "2_9714921\n",
      "2_1265\n",
      "2_11533904\n",
      "2_11534263\n",
      "3_137\n",
      "3_1540490\n",
      "3_1548153\n",
      "3_250\n",
      "3_2346807\n",
      "3_2350128\n",
      "4_801\n",
      "4_7255726\n",
      "4_7269426\n",
      "5_1947\n",
      "5_17738817\n",
      "5_17759685\n",
      "5_2922\n",
      "5_24100585\n",
      "5_24286474\n"
     ]
    }
   ],
   "source": [
    "for i in ['1_485', '1_702', '1_2867', '2_827', '2_970','2_1265', '3_137',\n",
    "       '3_250', '4_801', '5_1947', '5_2922']:\n",
    "    print(i)\n",
    "    print(dict_blocks[i][0])\n",
    "    print(dict_blocks[i][-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd3710d-7dad-4546-abe7-a9e0dcdf84e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fbd62a0-ad75-444e-b9a1-156b649026c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c7645dd4-8e94-4aa8-8217-e4ba298f3b6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>biotype</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1318</td>\n",
       "      <td>AT1G26680</td>\n",
       "      <td>AT1G26680</td>\n",
       "      <td>9219460</td>\n",
       "      <td>9223795</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcriptional factor B3 family protein [Sour...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>AT1G30710</td>\n",
       "      <td>AT1G30710</td>\n",
       "      <td>10895252</td>\n",
       "      <td>10897127</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FAD-binding Berberine family protein [Source:N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_1594</td>\n",
       "      <td>AT1G30720</td>\n",
       "      <td>AT1G30720</td>\n",
       "      <td>10897925</td>\n",
       "      <td>10899975</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FAD-binding Berberine family protein [Source:N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>AT1G55660</td>\n",
       "      <td>AT1G55660</td>\n",
       "      <td>20800694</td>\n",
       "      <td>20802377</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FBD, F-box and Leucine Rich Repeat domains con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>AT1G55675</td>\n",
       "      <td>AT1G55675</td>\n",
       "      <td>20803661</td>\n",
       "      <td>20804574</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transmembrane protein [Source:NCBI gene (forme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60335</td>\n",
       "      <td>AT5G60335</td>\n",
       "      <td>24272995</td>\n",
       "      <td>24274860</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>hydroxyacyl-thioester dehydratase type-like pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60350</td>\n",
       "      <td>AT5G60350</td>\n",
       "      <td>24277730</td>\n",
       "      <td>24279147</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60370</td>\n",
       "      <td>AT5G60370</td>\n",
       "      <td>24282666</td>\n",
       "      <td>24285061</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonuclease V-like protein [Source:NCBI gene (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60380</td>\n",
       "      <td>AT5G60380</td>\n",
       "      <td>24285920</td>\n",
       "      <td>24287808</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transmembrane protein, putative (DUF239) [Sour...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>5_780</td>\n",
       "      <td>AT5G26805</td>\n",
       "      <td>AT5G26805</td>\n",
       "      <td>9427543</td>\n",
       "      <td>9428067</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>B3 domain protein [Source:NCBI gene (formerly ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      block    gene_id  gene_name     start       end         biotype  \\\n",
       "0    1_1318  AT1G26680  AT1G26680   9219460   9223795  protein_coding   \n",
       "1    1_1594  AT1G30710  AT1G30710  10895252  10897127  protein_coding   \n",
       "2    1_1594  AT1G30720  AT1G30720  10897925  10899975  protein_coding   \n",
       "3    1_2867  AT1G55660  AT1G55660  20800694  20802377  protein_coding   \n",
       "4    1_2867  AT1G55675  AT1G55675  20803661  20804574  protein_coding   \n",
       "..      ...        ...        ...       ...       ...             ...   \n",
       "116  5_2922  AT5G60335  AT5G60335  24272995  24274860  protein_coding   \n",
       "117  5_2922  AT5G60350  AT5G60350  24277730  24279147  protein_coding   \n",
       "118  5_2922  AT5G60370  AT5G60370  24282666  24285061  protein_coding   \n",
       "119  5_2922  AT5G60380  AT5G60380  24285920  24287808  protein_coding   \n",
       "120   5_780  AT5G26805  AT5G26805   9427543   9428067  protein_coding   \n",
       "\n",
       "                                           description  \n",
       "0    transcriptional factor B3 family protein [Sour...  \n",
       "1    FAD-binding Berberine family protein [Source:N...  \n",
       "2    FAD-binding Berberine family protein [Source:N...  \n",
       "3    FBD, F-box and Leucine Rich Repeat domains con...  \n",
       "4    transmembrane protein [Source:NCBI gene (forme...  \n",
       "..                                                 ...  \n",
       "116  hydroxyacyl-thioester dehydratase type-like pr...  \n",
       "117                                               None  \n",
       "118  exonuclease V-like protein [Source:NCBI gene (...  \n",
       "119  transmembrane protein, putative (DUF239) [Sour...  \n",
       "120  B3 domain protein [Source:NCBI gene (formerly ...  \n",
       "\n",
       "[121 rows x 7 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_info_ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5991c0cc-3b49-4979-9ba3-db7e4d86a985",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_2867</td>\n",
       "      <td>AT1G55670</td>\n",
       "      <td>PSAG</td>\n",
       "      <td>Encodes subunit G of photosystem I, an 11-kDa ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G11840</td>\n",
       "      <td>GLX1</td>\n",
       "      <td>Encodes a glyoxalase I homolog ATGLX1.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70580</td>\n",
       "      <td>AOAT2</td>\n",
       "      <td>Encodes a protein with glyoxylate aminotransfe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70610</td>\n",
       "      <td>ABCB26</td>\n",
       "      <td>member of TAP subfamily</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_485</td>\n",
       "      <td>AT1G70660</td>\n",
       "      <td>MMZ2</td>\n",
       "      <td>MMZ2/UEV1B encodes a protein that may play a r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60220</td>\n",
       "      <td>TET4</td>\n",
       "      <td>Member of TETRASPANIN family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60230</td>\n",
       "      <td>SEN2</td>\n",
       "      <td>putative subunit of tRNA splicing endonuclease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60300</td>\n",
       "      <td>LecRK-I.9</td>\n",
       "      <td>Lectin Receptor Kinase involved in protein-pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60340</td>\n",
       "      <td>AAK6</td>\n",
       "      <td>Encodes a nuclear adenylate kinase that intera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>5_2922</td>\n",
       "      <td>AT5G60360</td>\n",
       "      <td>ALP</td>\n",
       "      <td>Encodes a senescence-associated thiol protease.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     block    gene_id  gene_name  \\\n",
       "0   1_2867  AT1G55670       PSAG   \n",
       "1    1_485  AT1G11840       GLX1   \n",
       "2    1_485  AT1G70580      AOAT2   \n",
       "3    1_485  AT1G70610     ABCB26   \n",
       "4    1_485  AT1G70660       MMZ2   \n",
       "..     ...        ...        ...   \n",
       "59  5_2922  AT5G60220       TET4   \n",
       "60  5_2922  AT5G60230       SEN2   \n",
       "61  5_2922  AT5G60300  LecRK-I.9   \n",
       "62  5_2922  AT5G60340       AAK6   \n",
       "63  5_2922  AT5G60360        ALP   \n",
       "\n",
       "                                          description  \n",
       "0   Encodes subunit G of photosystem I, an 11-kDa ...  \n",
       "1              Encodes a glyoxalase I homolog ATGLX1.  \n",
       "2   Encodes a protein with glyoxylate aminotransfe...  \n",
       "3                             member of TAP subfamily  \n",
       "4   MMZ2/UEV1B encodes a protein that may play a r...  \n",
       "..                                                ...  \n",
       "59                       Member of TETRASPANIN family  \n",
       "60     putative subunit of tRNA splicing endonuclease  \n",
       "61  Lectin Receptor Kinase involved in protein-pro...  \n",
       "62  Encodes a nuclear adenylate kinase that intera...  \n",
       "63    Encodes a senescence-associated thiol protease.  \n",
       "\n",
       "[64 rows x 4 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_info_ncbi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "68a265c3-e613-4260-b034-fefc6cc83691",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "207"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(flattened_gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "995521c0-5a98-4fa5-9205-fe099d72b362",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['protein_coding_gene'], dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair['biotype'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6c92c2a9-97cd-4723-8454-76c4bba9bda5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27206"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair['gene_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "684907cf-5ba4-40a0-81cc-501b8f197e00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AT1G26680',\n",
       " 'AT1G30710',\n",
       " 'AT1G30720',\n",
       " 'AT1G55660',\n",
       " 'AT1G55670',\n",
       " 'AT1G55675',\n",
       " 'AT1G55680',\n",
       " 'AT1G55690',\n",
       " 'AT1G67560',\n",
       " 'AT1G67570',\n",
       " 'AT1G70020',\n",
       " 'AT1G11840',\n",
       " 'AT1G70580',\n",
       " 'AT1G70590',\n",
       " 'AT1G70600',\n",
       " 'AT1G70610',\n",
       " 'AT1G70620',\n",
       " 'AT1G70630',\n",
       " 'AT1G70640',\n",
       " 'AT1G70650',\n",
       " 'AT1G70660',\n",
       " 'AT1G17500',\n",
       " 'AT1G17510',\n",
       " 'AT1G17520',\n",
       " 'AT1G17530',\n",
       " 'AT1G17540',\n",
       " 'AT1G17545',\n",
       " 'AT1G17550',\n",
       " 'AT1G17560',\n",
       " 'AT1G17580',\n",
       " 'AT1G17590',\n",
       " 'AT1G17600',\n",
       " 'AT1G17610',\n",
       " 'AT1G17615',\n",
       " 'AT1G17620',\n",
       " 'AT1G17630',\n",
       " 'AT1G17640',\n",
       " 'AT1G17650',\n",
       " 'AT1G17665',\n",
       " 'AT1G17680',\n",
       " 'AT1G17690',\n",
       " 'AT1G17700',\n",
       " 'AT1G17710',\n",
       " 'AT1G17720',\n",
       " 'AT1G17730',\n",
       " 'AT1G17744',\n",
       " 'AT1G17745',\n",
       " 'AT1G17750',\n",
       " 'AT1G17760',\n",
       " 'AT1G17770',\n",
       " 'AT1G17780',\n",
       " 'AT1G17790',\n",
       " 'AT1G17800',\n",
       " 'AT1G17810',\n",
       " 'AT1G17820',\n",
       " 'AT1G17830',\n",
       " 'AT1G17840',\n",
       " 'AT1G17850',\n",
       " 'AT1G17860',\n",
       " 'AT1G17870',\n",
       " 'AT1G17880',\n",
       " 'AT1G17890',\n",
       " 'AT1G17910',\n",
       " 'AT1G17920',\n",
       " 'AT1G17930',\n",
       " 'AT1G17940',\n",
       " 'AT1G17950',\n",
       " 'AT1G17960',\n",
       " 'AT2G24230',\n",
       " 'AT2G27030',\n",
       " 'AT2G17380',\n",
       " 'AT2G17390',\n",
       " 'AT2G17410',\n",
       " 'AT2G17420',\n",
       " 'AT2G22475',\n",
       " 'AT2G22480',\n",
       " 'AT2G22490',\n",
       " 'AT2G22500',\n",
       " 'AT2G22510',\n",
       " 'AT2G22520',\n",
       " 'AT2G22530',\n",
       " 'AT2G22540',\n",
       " 'AT2G22560',\n",
       " 'AT2G22570',\n",
       " 'AT2G22590',\n",
       " 'AT2G22600',\n",
       " 'AT2G22610',\n",
       " 'AT2G22620',\n",
       " 'AT2G22630',\n",
       " 'AT2G22640',\n",
       " 'AT2G22650',\n",
       " 'AT2G22660',\n",
       " 'AT2G22670',\n",
       " 'AT2G22680',\n",
       " 'AT2G22690',\n",
       " 'AT2G22720',\n",
       " 'AT2G22730',\n",
       " 'AT2G22740',\n",
       " 'AT2G22750',\n",
       " 'AT2G22760',\n",
       " 'AT2G22770',\n",
       " 'AT2G22780',\n",
       " 'AT2G22790',\n",
       " 'AT2G22795',\n",
       " 'AT2G22800',\n",
       " 'AT2G22802',\n",
       " 'AT2G22805',\n",
       " 'AT3G05327',\n",
       " 'AT3G28580',\n",
       " 'AT3G28590',\n",
       " 'AT3G28600',\n",
       " 'AT3G28610',\n",
       " 'AT3G28620',\n",
       " 'AT3G28630',\n",
       " 'AT3G28640',\n",
       " 'AT3G05380',\n",
       " 'AT3G05390',\n",
       " 'AT3G28910',\n",
       " 'AT3G28917',\n",
       " 'AT3G28918',\n",
       " 'AT3G28920',\n",
       " 'AT3G28925',\n",
       " 'AT3G28930',\n",
       " 'AT3G28940',\n",
       " 'AT3G28950',\n",
       " 'AT3G28956',\n",
       " 'AT3G07350',\n",
       " 'AT3G47390',\n",
       " 'AT3G47400',\n",
       " 'AT3G47410',\n",
       " 'AT3G47420',\n",
       " 'AT3G50750',\n",
       " 'AT3G62360',\n",
       " 'AT3G62370',\n",
       " 'AT4G12110',\n",
       " 'AT4G12120',\n",
       " 'AT4G12130',\n",
       " 'AT4G12140',\n",
       " 'AT4G12150',\n",
       " 'AT5G44080',\n",
       " 'AT5G44090',\n",
       " 'AT5G44100',\n",
       " 'AT5G44110',\n",
       " 'AT5G44120',\n",
       " 'AT5G48560',\n",
       " 'AT5G59820',\n",
       " 'AT5G59830',\n",
       " 'AT5G59840',\n",
       " 'AT5G59845',\n",
       " 'AT5G59850',\n",
       " 'AT5G59860',\n",
       " 'AT5G59870',\n",
       " 'AT5G59880',\n",
       " 'AT5G59890',\n",
       " 'AT5G59900',\n",
       " 'AT5G59910',\n",
       " 'AT5G59920',\n",
       " 'AT5G59930',\n",
       " 'AT5G59940',\n",
       " 'AT5G59950',\n",
       " 'AT5G59960',\n",
       " 'AT5G59970',\n",
       " 'AT5G59980',\n",
       " 'AT5G59990',\n",
       " 'AT5G60000',\n",
       " 'AT5G60010',\n",
       " 'AT5G60020',\n",
       " 'AT5G60030',\n",
       " 'AT5G60040',\n",
       " 'AT5G60050',\n",
       " 'AT5G60060',\n",
       " 'AT5G60070',\n",
       " 'AT5G60080',\n",
       " 'AT5G60090',\n",
       " 'AT5G60100',\n",
       " 'AT5G60110',\n",
       " 'AT5G60120',\n",
       " 'AT5G60130',\n",
       " 'AT5G60140',\n",
       " 'AT5G60142',\n",
       " 'AT5G60150',\n",
       " 'AT5G60160',\n",
       " 'AT5G60170',\n",
       " 'AT5G60180',\n",
       " 'AT5G60190',\n",
       " 'AT5G60200',\n",
       " 'AT5G60210',\n",
       " 'AT5G60220',\n",
       " 'AT5G60230',\n",
       " 'AT5G60240',\n",
       " 'AT5G60250',\n",
       " 'AT5G60260',\n",
       " 'AT5G60270',\n",
       " 'AT5G60280',\n",
       " 'AT5G60290',\n",
       " 'AT5G60300',\n",
       " 'AT5G60310',\n",
       " 'AT5G60320',\n",
       " 'AT5G60330',\n",
       " 'AT5G60335',\n",
       " 'AT5G60340',\n",
       " 'AT5G60350',\n",
       " 'AT5G60360',\n",
       " 'AT5G60370',\n",
       " 'AT5G60380',\n",
       " 'AT5G26800',\n",
       " 'AT5G26805']"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flattened_gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "130035f5-e53a-4259-913f-6cb78d938adf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_gene_info_ncbi['gene_id']).intersection(df_gene_info_ensembl['gene_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "793be534-5e2f-4134-ae99-6cddfba23dac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(flattened_gene_ids).intersection(set(df_gene_info_ncbi['gene_id'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b8e39adc-fe7f-49fc-bd82-a9b5c78da731",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "121"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(flattened_gene_ids).intersection(set(df_gene_info_ensembl['gene_id'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4acad464-e98c-41ec-993d-ba916c6f4b7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "185"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "121 + 64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "0ae6cee3-8831-45df-9269-54d3f6ac6ce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "207"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(flattened_gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32064662-53d4-4275-b93f-cfac141e5af5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9c13c24d-b48a-4612-ad0a-badce8ca0109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>block_x</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>biotype</th>\n",
       "      <th>description_x</th>\n",
       "      <th>block_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>description_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [block_x, gene_id, gene_name_x, start, end, biotype, description_x, block_y, gene_name_y, description_y]\n",
       "Index: []"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_info_ensembl.merge(df_gene_info_ncbi, on = 'gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af6c1ac-d728-4276-b4c4-7a09ceb9f91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_info_ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdfac6b6-7c36-4ce8-9249-091dd4aef886",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6a94336-e077-4996-932d-f023c7de90e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "# Provide your email, NCBI requires this for queries\n",
    "Entrez.email = \"tbellg@berkeley.edu\"\n",
    "\n",
    "# Function to retrieve gene information using NCBI Entrez API\n",
    "def get_gene_info_ncbi(gene_id):\n",
    "    try:\n",
    "        # Search for the gene in the NCBI database\n",
    "        search_handle = Entrez.esearch(db=\"gene\", term=f\"{gene_id}[Gene]\", retmax=1)\n",
    "        search_results = Entrez.read(search_handle)\n",
    "        search_handle.close()\n",
    "        \n",
    "        if search_results[\"IdList\"]:\n",
    "            gene_ncbi_id = search_results[\"IdList\"][0]  # Get the first matching gene ID\n",
    "            \n",
    "            # Fetch detailed gene information using the gene ID\n",
    "            fetch_handle = Entrez.efetch(db=\"gene\", id=gene_ncbi_id, retmode=\"xml\")\n",
    "            gene_data = Entrez.read(fetch_handle)\n",
    "            fetch_handle.close()\n",
    "            \n",
    "            # Extract relevant information from the gene data\n",
    "            gene_info = gene_data[0]\n",
    "            gene_name = gene_info['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']\n",
    "            gene_desc = gene_info['Entrezgene_summary']\n",
    "            return {\n",
    "                \"gene_id\": gene_id,\n",
    "                \"gene_name\": gene_name,\n",
    "                \"description\": gene_desc\n",
    "            }\n",
    "        else:\n",
    "            return None\n",
    "    except Exception as e:\n",
    "        print(f\"Error fetching data for gene {gene_id}: {e}\")\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "24d1c97b-d531-406d-b5df-5f50573e51cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AT1G26680', 'AT1G30710,AT1G30720',\n",
       "       'AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690',\n",
       "       'AT1G67560,AT1G67570', 'AT1G70020',\n",
       "       'AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70610,AT1G70620,AT1G70630,AT1G70640,AT1G70650,AT1G70660',\n",
       "       'AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17540,AT1G17545,AT1G17550,AT1G17560,AT1G17580,AT1G17590,AT1G17600,AT1G17610,AT1G17615,AT1G17620,AT1G17630,AT1G17640,AT1G17650,AT1G17665,AT1G17680,AT1G17690,AT1G17700,AT1G17710,AT1G17720,AT1G17730,AT1G17744,AT1G17745,AT1G17750,AT1G17760,AT1G17770,AT1G17780,AT1G17790,AT1G17800,AT1G17810,AT1G17820,AT1G17830,AT1G17840,AT1G17850,AT1G17860,AT1G17870,AT1G17880,AT1G17890,AT1G17910,AT1G17920,AT1G17930,AT1G17940,AT1G17950,AT1G17960',\n",
       "       'AT2G24230', 'AT2G27030',\n",
       "       'AT2G17380,AT2G17390,AT2G17410,AT2G17420',\n",
       "       'AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22510,AT2G22520,AT2G22530,AT2G22540,AT2G22560,AT2G22570,AT2G22590,AT2G22600,AT2G22610,AT2G22620,AT2G22630,AT2G22640,AT2G22650,AT2G22660,AT2G22670,AT2G22680,AT2G22690,AT2G22720,AT2G22730,AT2G22740,AT2G22750,AT2G22760,AT2G22770,AT2G22780,AT2G22790,AT2G22795,AT2G22800,AT2G22802,AT2G22805',\n",
       "       None,\n",
       "       'AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28610,AT3G28620,AT3G28630,AT3G28640',\n",
       "       'AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28918,AT3G28920,AT3G28925,AT3G28930,AT3G28940,AT3G28950,AT3G28956',\n",
       "       'AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420', 'AT3G50750',\n",
       "       'AT3G62360,AT3G62370',\n",
       "       'AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150',\n",
       "       'AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120', 'AT5G48560',\n",
       "       'AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59850,AT5G59860,AT5G59870,AT5G59880,AT5G59890,AT5G59900,AT5G59910,AT5G59920,AT5G59930,AT5G59940,AT5G59950,AT5G59960,AT5G59970,AT5G59980,AT5G59990,AT5G60000,AT5G60010,AT5G60020,AT5G60030,AT5G60040,AT5G60050,AT5G60060,AT5G60070,AT5G60080,AT5G60090,AT5G60100,AT5G60110,AT5G60120,AT5G60130,AT5G60140,AT5G60142,AT5G60150,AT5G60160,AT5G60170,AT5G60180,AT5G60190,AT5G60200,AT5G60210,AT5G60220,AT5G60230,AT5G60240,AT5G60250,AT5G60260,AT5G60270,AT5G60280,AT5G60290,AT5G60300,AT5G60310,AT5G60320,AT5G60330,AT5G60335,AT5G60340,AT5G60350,AT5G60360,AT5G60370,AT5G60380',\n",
       "       'AT5G26800,AT5G26805'], dtype=object)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm['matching_gene_id'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e711f6c8-1b84-4669-8002-78a62561c23b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the values from 'matching_gene_id' column\n",
    "matching_gene_ids = sign_wza_lfmm['matching_gene_id'].values\n",
    "\n",
    "# Filter out None or NaN values\n",
    "filtered_gene_ids = [gene_id for gene_id in matching_gene_ids if gene_id]\n",
    "\n",
    "# Flatten the gene IDs (since some are comma-separated) into a single list\n",
    "flattened_gene_ids = []\n",
    "for gene_ids in filtered_gene_ids:\n",
    "    flattened_gene_ids.extend(gene_ids.split(','))  # Split the comma-separated strings and extend the list\n",
    "\n",
    "# Join all the gene IDs into a single string, separated by comma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "5eb07065-75d1-4908-85e5-7e04390faaae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "207"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(flattened_gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "35ce5a57-149d-40d0-87aa-60ad57739d0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AT1G26680,AT1G30710,AT1G30720,AT1G55660,AT1G55670,AT1G55675,AT1G55680,AT1G55690,AT1G67560,AT1G67570,AT1G70020,AT1G11840,AT1G70580,AT1G70590,AT1G70600,AT1G70610,AT1G70620,AT1G70630,AT1G70640,AT1G70650,AT1G70660,AT1G17500,AT1G17510,AT1G17520,AT1G17530,AT1G17540,AT1G17545,AT1G17550,AT1G17560,AT1G17580,AT1G17590,AT1G17600,AT1G17610,AT1G17615,AT1G17620,AT1G17630,AT1G17640,AT1G17650,AT1G17665,AT1G17680,AT1G17690,AT1G17700,AT1G17710,AT1G17720,AT1G17730,AT1G17744,AT1G17745,AT1G17750,AT1G17760,AT1G17770,AT1G17780,AT1G17790,AT1G17800,AT1G17810,AT1G17820,AT1G17830,AT1G17840,AT1G17850,AT1G17860,AT1G17870,AT1G17880,AT1G17890,AT1G17910,AT1G17920,AT1G17930,AT1G17940,AT1G17950,AT1G17960,AT2G24230,AT2G27030,AT2G17380,AT2G17390,AT2G17410,AT2G17420,AT2G22475,AT2G22480,AT2G22490,AT2G22500,AT2G22510,AT2G22520,AT2G22530,AT2G22540,AT2G22560,AT2G22570,AT2G22590,AT2G22600,AT2G22610,AT2G22620,AT2G22630,AT2G22640,AT2G22650,AT2G22660,AT2G22670,AT2G22680,AT2G22690,AT2G22720,AT2G22730,AT2G22740,AT2G22750,AT2G22760,AT2G22770,AT2G22780,AT2G22790,AT2G22795,AT2G22800,AT2G22802,AT2G22805,AT3G05327,AT3G28580,AT3G28590,AT3G28600,AT3G28610,AT3G28620,AT3G28630,AT3G28640,AT3G05380,AT3G05390,AT3G28910,AT3G28917,AT3G28918,AT3G28920,AT3G28925,AT3G28930,AT3G28940,AT3G28950,AT3G28956,AT3G07350,AT3G47390,AT3G47400,AT3G47410,AT3G47420,AT3G50750,AT3G62360,AT3G62370,AT4G12110,AT4G12120,AT4G12130,AT4G12140,AT4G12150,AT5G44080,AT5G44090,AT5G44100,AT5G44110,AT5G44120,AT5G48560,AT5G59820,AT5G59830,AT5G59840,AT5G59845,AT5G59850,AT5G59860,AT5G59870,AT5G59880,AT5G59890,AT5G59900,AT5G59910,AT5G59920,AT5G59930,AT5G59940,AT5G59950,AT5G59960,AT5G59970,AT5G59980,AT5G59990,AT5G60000,AT5G60010,AT5G60020,AT5G60030,AT5G60040,AT5G60050,AT5G60060,AT5G60070,AT5G60080,AT5G60090,AT5G60100,AT5G60110,AT5G60120,AT5G60130,AT5G60140,AT5G60142,AT5G60150,AT5G60160,AT5G60170,AT5G60180,AT5G60190,AT5G60200,AT5G60210,AT5G60220,AT5G60230,AT5G60240,AT5G60250,AT5G60260,AT5G60270,AT5G60280,AT5G60290,AT5G60300,AT5G60310,AT5G60320,AT5G60330,AT5G60335,AT5G60340,AT5G60350,AT5G60360,AT5G60370,AT5G60380,AT5G26800,AT5G26805'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_gene_ids_joined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ceab912-9332-4243-80db-00795dca55a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ee8e936-7c0f-43f4-8cf8-0a1953cf3682",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63dd4eea-6ccd-4b2e-b20a-d25a2f6b6f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "# Function to retrieve gene information from NCBI using the API\n",
    "def get_gene_info_ncbi(gene_id):\n",
    "    url = f\"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/gene/id/{gene_id}/download_summary\"\n",
    "    headers = {\"accept\": \"application/json\"}\n",
    "    \n",
    "    response = requests.get(url, headers=headers)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        # Return the JSON response\n",
    "        return response.json()\n",
    "    else:\n",
    "        return None  # Return None if the request failed\n",
    "\n",
    "# Example gene IDs (from your dataset)\n",
    "gene_ids = ['59067', '7157']  # Replace these with the actual gene IDs from your dataset\n",
    "\n",
    "# Fetch information for each gene\n",
    "gene_info_list = []\n",
    "\n",
    "for gene_id in gene_ids:\n",
    "    gene_info = get_gene_info_ncbi(gene_id)\n",
    "    if gene_info:\n",
    "        # Extract relevant data from the response\n",
    "        gene_data = {\n",
    "            \"gene_id\": gene_id,\n",
    "            \"gene_name\": gene_info['gene']['name'],\n",
    "            \"description\": gene_info['gene']['description']\n",
    "        }\n",
    "        gene_info_list.append(gene_data)\n",
    "\n",
    "# Convert the list of gene information to a DataFrame for easier handling\n",
    "df_gene_info = pd.DataFrame(gene_info_list)\n",
    "\n",
    "# Display the gene information DataFrame\n",
    "print(df_gene_info)\n",
    "\n",
    "# Now, assuming we have significant blocks from your data\n",
    "# Let's integrate the gene information with the plot\n",
    "\n",
    "# Significance threshold\n",
    "threshold_value = 0.05 / len(wza_lfmm)\n",
    "threshold = -np.log10(threshold_value)\n",
    "biovar = 'bio1'\n",
    "\n",
    "# Create chrom_pos in wza_lfmm by combining 'chrom' and 'pos'\n",
    "wza_lfmm['chrom_pos'] = wza_lfmm['chrom'].astype(str) + '_' + wza_lfmm['pos'].astype(str)\n",
    "\n",
    "# Copy the relevant columns for plotting\n",
    "df = wza_lfmm[['Z_pVal', 'pos', 'chrom', 'chrom_pos']].copy()\n",
    "\n",
    "# Parse chromosome number and position\n",
    "df['chromosome'] = df['chrom']\n",
    "df['position'] = df['pos']\n",
    "df['-log10(pvalue)'] = -np.log10(df['Z_pVal'])\n",
    "\n",
    "# Define colors for each chromosome or block\n",
    "colors = sns.color_palette(\"crest\", n_colors=len(df['chromosome'].unique()))  # Assign a unique color per chromosome\n",
    "\n",
    "# Calculate chromosome offsets to prevent overlap\n",
    "chromosome_offsets = {}\n",
    "offset = 0\n",
    "for chrom in sorted(df['chromosome'].unique()):\n",
    "    chromosome_offsets[chrom] = offset\n",
    "    max_position = df[df['chromosome'] == chrom]['position'].max()\n",
    "    offset += max_position + 200  # Add buffer to prevent overlap\n",
    "\n",
    "# Apply offsets to the position\n",
    "df['adjusted_position'] = df.apply(lambda row: row['position'] + chromosome_offsets[row['chromosome']], axis=1)\n",
    "\n",
    "# Create the Manhattan plot\n",
    "plt.figure(figsize=(20, 6))\n",
    "\n",
    "# Plot each chromosome separately\n",
    "for chrom in sorted(df['chromosome'].unique()):\n",
    "    subset = df[df['chromosome'] == chrom]\n",
    "    plt.scatter(\n",
    "        subset['adjusted_position'],\n",
    "        subset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc53f0ec-ecfd-4d26-add9-d1ed61b3a69e",
   "metadata": {},
   "outputs": [],
   "source": [
    "1_1594"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3955f8-de65-448a-abe7-69a0276aec9c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6eedf6c2-271f-499e-9f08-3add8638fb78",
   "metadata": {},
   "outputs": [],
   "source": [
    "## all samples wza on kendalltau\n",
    "wza_kendall = pd.read_csv('../wza/wza_kendalltau_results_bio1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "27327310-82fc-4bb4-a9bb-61dfa1cf5317",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sign_wza_kendall = wza_kendall[wza_kendall['Z_pVal'] <= 0.05/len(wza_kendall)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c94b8a68-05ed-498e-b9ee-b918897fb7eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5406</th>\n",
       "      <td>2_1264</td>\n",
       "      <td>1885</td>\n",
       "      <td>0</td>\n",
       "      <td>2.057459</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7217</th>\n",
       "      <td>2_973</td>\n",
       "      <td>1843</td>\n",
       "      <td>5</td>\n",
       "      <td>10.919003</td>\n",
       "      <td>9.999435e-01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10099</th>\n",
       "      <td>3_551</td>\n",
       "      <td>1849</td>\n",
       "      <td>4</td>\n",
       "      <td>5.318958</td>\n",
       "      <td>9.999891e-01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12569</th>\n",
       "      <td>4_2808</td>\n",
       "      <td>1729</td>\n",
       "      <td>220</td>\n",
       "      <td>20.956158</td>\n",
       "      <td>1.191537e-162</td>\n",
       "      <td>1.110223e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13384</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>146</td>\n",
       "      <td>31.615317</td>\n",
       "      <td>5.445274e-231</td>\n",
       "      <td>2.907737e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9133</th>\n",
       "      <td>3_2730</td>\n",
       "      <td>53</td>\n",
       "      <td>30</td>\n",
       "      <td>15.113784</td>\n",
       "      <td>4.984762e-46</td>\n",
       "      <td>2.548200e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7135</th>\n",
       "      <td>2_897</td>\n",
       "      <td>30</td>\n",
       "      <td>18</td>\n",
       "      <td>11.773300</td>\n",
       "      <td>7.715812e-29</td>\n",
       "      <td>4.589528e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>1_1391</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>11.351576</td>\n",
       "      <td>2.056867e-32</td>\n",
       "      <td>6.736171e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3851</th>\n",
       "      <td>1_4506</td>\n",
       "      <td>109</td>\n",
       "      <td>8</td>\n",
       "      <td>17.976827</td>\n",
       "      <td>1.550062e-05</td>\n",
       "      <td>1.090242e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6199</th>\n",
       "      <td>2_199</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>9.224959</td>\n",
       "      <td>9.910000e-18</td>\n",
       "      <td>3.637041e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>9.028052</td>\n",
       "      <td>3.984317e-07</td>\n",
       "      <td>2.119224e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>1_1040</td>\n",
       "      <td>93</td>\n",
       "      <td>37</td>\n",
       "      <td>14.712225</td>\n",
       "      <td>6.834746e-49</td>\n",
       "      <td>2.455139e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene  SNPs  hits          Z  top_candidate_p        Z_pVal\n",
       "5406   2_1264  1885     0   2.057459     1.000000e+00  0.000000e+00\n",
       "7217    2_973  1843     5  10.919003     9.999435e-01  0.000000e+00\n",
       "10099   3_551  1849     4   5.318958     9.999891e-01  0.000000e+00\n",
       "12569  4_2808  1729   220  20.956158    1.191537e-162  1.110223e-16\n",
       "13384   4_801   225   146  31.615317    5.445274e-231  2.907737e-10\n",
       "9133   3_2730    53    30  15.113784     4.984762e-46  2.548200e-09\n",
       "7135    2_897    30    18  11.773300     7.715812e-29  4.589528e-08\n",
       "431    1_1391    27    19  11.351576     2.056867e-32  6.736171e-08\n",
       "3851   1_4506   109     8  17.976827     1.550062e-05  1.090242e-07\n",
       "6199    2_199    10     9   9.224959     9.910000e-18  3.637041e-07\n",
       "5407   2_1265    16     5   9.028052     3.984317e-07  2.119224e-06\n",
       "48     1_1040    93    37  14.712225     6.834746e-49  2.455139e-06"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_kendall.sort_values('Z_pVal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bff0c1c7-cfbc-40d5-a222-474c53244b3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "kendall_sign = ['2_1264', '2_973' '3_551', '4_2808']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc670da6-e238-4247-92a3-27844d14cb74",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e521774-4baf-4235-b490-95c27d99736a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2db0c6eb-05be-41a9-91cb-2e9c023098a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>SNPs_x</th>\n",
       "      <th>hits_x</th>\n",
       "      <th>Z_x</th>\n",
       "      <th>top_candidate_p_x</th>\n",
       "      <th>Z_pVal_x</th>\n",
       "      <th>SNPs_y</th>\n",
       "      <th>hits_y</th>\n",
       "      <th>Z_y</th>\n",
       "      <th>top_candidate_p_y</th>\n",
       "      <th>Z_pVal_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>9.028052</td>\n",
       "      <td>3.984317e-07</td>\n",
       "      <td>2.119224e-06</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>146</td>\n",
       "      <td>31.615317</td>\n",
       "      <td>5.445274e-231</td>\n",
       "      <td>2.907737e-10</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene  SNPs_x  hits_x        Z_x  top_candidate_p_x      Z_pVal_x  SNPs_y  \\\n",
       "0  2_1265      16       5   9.028052       3.984317e-07  2.119224e-06      16   \n",
       "1   4_801     225     146  31.615317      5.445274e-231  2.907737e-10     225   \n",
       "\n",
       "   hits_y        Z_y  top_candidate_p_y      Z_pVal_y  \n",
       "0      15  11.151806       1.585000e-29  1.569795e-08  \n",
       "1     130  23.960528      7.309601e-196  1.040595e-06  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(sign_wza_kendall,sign_wza_lfmm, on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a55f3283-069a-4515-a5e3-9b52b93bf2bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['2_1265', '4_801']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d19028b0-f89f-49f4-8e7e-83dafbcf7c99",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a42a76d6-b6b6-4804-9c81-0315288ddf26",
   "metadata": {},
   "outputs": [],
   "source": [
    "linages_wza_picmin = pd.read_csv('../linages_wza_picmin/linage_based_kendall_wza_picmin.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b144aac9-877e-46bc-be93-67b45010f502",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>numLin</th>\n",
       "      <th>p</th>\n",
       "      <th>q</th>\n",
       "      <th>n_est</th>\n",
       "      <th>locus</th>\n",
       "      <th>redundan</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>start</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4172</th>\n",
       "      <td>4172</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.001955</td>\n",
       "      <td>12</td>\n",
       "      <td>1_4847</td>\n",
       "      <td>1</td>\n",
       "      <td>4847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>4847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12438</th>\n",
       "      <td>12438</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.002633</td>\n",
       "      <td>12</td>\n",
       "      <td>4_2817</td>\n",
       "      <td>4</td>\n",
       "      <td>2817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>2817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13736</th>\n",
       "      <td>13736</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.015449</td>\n",
       "      <td>12</td>\n",
       "      <td>5_1272</td>\n",
       "      <td>5</td>\n",
       "      <td>1272</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>1272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5818</th>\n",
       "      <td>5818</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>0.022242</td>\n",
       "      <td>12</td>\n",
       "      <td>2_170</td>\n",
       "      <td>2</td>\n",
       "      <td>170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>259</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000574</td>\n",
       "      <td>0.023699</td>\n",
       "      <td>12</td>\n",
       "      <td>1_1239</td>\n",
       "      <td>1</td>\n",
       "      <td>1239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14707</th>\n",
       "      <td>14707</td>\n",
       "      <td>12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12</td>\n",
       "      <td>5_2166</td>\n",
       "      <td>5</td>\n",
       "      <td>2166</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>2166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12717</th>\n",
       "      <td>12717</td>\n",
       "      <td>12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12</td>\n",
       "      <td>4_325</td>\n",
       "      <td>4</td>\n",
       "      <td>325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9730</th>\n",
       "      <td>9730</td>\n",
       "      <td>12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12</td>\n",
       "      <td>3_3372</td>\n",
       "      <td>3</td>\n",
       "      <td>3372</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8996</th>\n",
       "      <td>8996</td>\n",
       "      <td>12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12</td>\n",
       "      <td>3_2698</td>\n",
       "      <td>3</td>\n",
       "      <td>2698</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>2698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12714</th>\n",
       "      <td>12714</td>\n",
       "      <td>12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12</td>\n",
       "      <td>4_322</td>\n",
       "      <td>4</td>\n",
       "      <td>322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6417 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index  numLin         p         q  n_est   locus  redundan  scaffold  \\\n",
       "4172    4172      12  0.000024  0.001955     12  1_4847         1      4847   \n",
       "12438  12438      12  0.000036  0.002633     12  4_2817         4      2817   \n",
       "13736  13736      12  0.000337  0.015449     12  5_1272         5      1272   \n",
       "5818    5818      12  0.000523  0.022242     12   2_170         2       170   \n",
       "259      259      12  0.000574  0.023699     12  1_1239         1      1239   \n",
       "...      ...     ...       ...       ...    ...     ...       ...       ...   \n",
       "14707  14707      12  1.000000  1.000000     12  5_2166         5      2166   \n",
       "12717  12717      12  1.000000  1.000000     12   4_325         4       325   \n",
       "9730    9730      12  1.000000  1.000000     12  3_3372         3      3372   \n",
       "8996    8996      12  1.000000  1.000000     12  3_2698         3      2698   \n",
       "12714  12714      12  1.000000  1.000000     12   4_322         4       322   \n",
       "\n",
       "       start  chrom   pos  \n",
       "4172     NaN      1  4847  \n",
       "12438    NaN      4  2817  \n",
       "13736    NaN      5  1272  \n",
       "5818     NaN      2   170  \n",
       "259      NaN      1  1239  \n",
       "...      ...    ...   ...  \n",
       "14707    NaN      5  2166  \n",
       "12717    NaN      4   325  \n",
       "9730     NaN      3  3372  \n",
       "8996     NaN      3  2698  \n",
       "12714    NaN      4   322  \n",
       "\n",
       "[6417 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linages_wza_picmin[linages_wza_picmin['n_est'] ==12].sort_values('p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "34b6798a-7dac-4671-85e8-cf3cee560d61",
   "metadata": {},
   "outputs": [],
   "source": [
    "th = 0.05/len(linages_wza_picmin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2af7715c-2159-4e01-88b9-5058f311d43d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_linages_wza_picmin = linages_wza_picmin[linages_wza_picmin['p'] <= th]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5bc012ea-0f4b-4913-8d9f-f532a3349102",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_linages_wza_picmin = sign_linages_wza_picmin.rename(columns = {'locus':'gene'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e9e62a9b-95b3-4f78-b975-6d95fab4f2aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>numLin</th>\n",
       "      <th>p</th>\n",
       "      <th>q</th>\n",
       "      <th>n_est</th>\n",
       "      <th>gene</th>\n",
       "      <th>redundan</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>start</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>48</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>1_1040</td>\n",
       "      <td>1</td>\n",
       "      <td>1040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9684</th>\n",
       "      <td>9684</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>3_3330</td>\n",
       "      <td>3</td>\n",
       "      <td>3330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9683</th>\n",
       "      <td>9683</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>3_333</td>\n",
       "      <td>3</td>\n",
       "      <td>333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9440</th>\n",
       "      <td>9440</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>3_3104</td>\n",
       "      <td>3</td>\n",
       "      <td>3104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9033</th>\n",
       "      <td>9033</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>3_2730</td>\n",
       "      <td>3</td>\n",
       "      <td>2730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>2730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14579</th>\n",
       "      <td>14579</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>7</td>\n",
       "      <td>5_2048</td>\n",
       "      <td>5</td>\n",
       "      <td>2048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>2048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3058</th>\n",
       "      <td>3058</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>7</td>\n",
       "      <td>1_3821</td>\n",
       "      <td>1</td>\n",
       "      <td>3821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>3821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863</th>\n",
       "      <td>3863</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>8</td>\n",
       "      <td>1_4562</td>\n",
       "      <td>1</td>\n",
       "      <td>4562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>4562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15070</th>\n",
       "      <td>15070</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>8</td>\n",
       "      <td>5_2505</td>\n",
       "      <td>5</td>\n",
       "      <td>2505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>2505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11347</th>\n",
       "      <td>11347</td>\n",
       "      <td>12</td>\n",
       "      <td>2.999997e-06</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>8</td>\n",
       "      <td>4_1803</td>\n",
       "      <td>4</td>\n",
       "      <td>1803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index  numLin             p         q  n_est    gene  redundan  \\\n",
       "48        48      12  9.999990e-07  0.000158      2  1_1040         1   \n",
       "9684    9684      12  9.999990e-07  0.000158      2  3_3330         3   \n",
       "9683    9683      12  9.999990e-07  0.000158      2   3_333         3   \n",
       "9440    9440      12  9.999990e-07  0.000158      2  3_3104         3   \n",
       "9033    9033      12  9.999990e-07  0.000158      2  3_2730         3   \n",
       "...      ...     ...           ...       ...    ...     ...       ...   \n",
       "14579  14579      12  9.999990e-07  0.000158      7  5_2048         5   \n",
       "3058    3058      12  9.999990e-07  0.000158      7  1_3821         1   \n",
       "3863    3863      12  9.999990e-07  0.000158      8  1_4562         1   \n",
       "15070  15070      12  9.999990e-07  0.000158      8  5_2505         5   \n",
       "11347  11347      12  2.999997e-06  0.000380      8  4_1803         4   \n",
       "\n",
       "       scaffold  start  chrom   pos  \n",
       "48         1040    NaN      1  1040  \n",
       "9684       3330    NaN      3  3330  \n",
       "9683        333    NaN      3   333  \n",
       "9440       3104    NaN      3  3104  \n",
       "9033       2730    NaN      3  2730  \n",
       "...         ...    ...    ...   ...  \n",
       "14579      2048    NaN      5  2048  \n",
       "3058       3821    NaN      1  3821  \n",
       "3863       4562    NaN      1  4562  \n",
       "15070      2505    NaN      5  2505  \n",
       "11347      1803    NaN      4  1803  \n",
       "\n",
       "[130 rows x 11 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_linages_wza_picmin.sort_values('n_est')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0df920d9-6e45-4a6e-adef-9361ed552fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 8 lineages adn significant \n",
    "lineages = ['4_1803', '5_2505', '1_4562' ]\n",
    "\n",
    "# 12 lineages most significant \n",
    "lineages = ['1_4847']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7c6bb67-555e-4e34-8616-e8cd12d557cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bb2ed9a8-9e3e-492d-a70c-b9c2ef9bcf0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>numLin</th>\n",
       "      <th>p</th>\n",
       "      <th>q</th>\n",
       "      <th>n_est</th>\n",
       "      <th>gene</th>\n",
       "      <th>redundan</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>start</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>48</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>1_1040</td>\n",
       "      <td>1</td>\n",
       "      <td>1040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1040</td>\n",
       "      <td>93</td>\n",
       "      <td>37</td>\n",
       "      <td>14.712225</td>\n",
       "      <td>6.834746e-49</td>\n",
       "      <td>2.455139e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>425</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>3</td>\n",
       "      <td>1_1391</td>\n",
       "      <td>1</td>\n",
       "      <td>1391</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1391</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>11.351576</td>\n",
       "      <td>2.056867e-32</td>\n",
       "      <td>6.736171e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3802</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>1_4506</td>\n",
       "      <td>1</td>\n",
       "      <td>4506</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>4506</td>\n",
       "      <td>109</td>\n",
       "      <td>8</td>\n",
       "      <td>17.976827</td>\n",
       "      <td>1.550062e-05</td>\n",
       "      <td>1.090242e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5348</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>5</td>\n",
       "      <td>2_1265</td>\n",
       "      <td>2</td>\n",
       "      <td>1265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1265</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>9.028052</td>\n",
       "      <td>3.984317e-07</td>\n",
       "      <td>2.119224e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6132</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>6</td>\n",
       "      <td>2_199</td>\n",
       "      <td>2</td>\n",
       "      <td>199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>199</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>9.224959</td>\n",
       "      <td>9.910000e-18</td>\n",
       "      <td>3.637041e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7058</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>6</td>\n",
       "      <td>2_897</td>\n",
       "      <td>2</td>\n",
       "      <td>897</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>897</td>\n",
       "      <td>30</td>\n",
       "      <td>18</td>\n",
       "      <td>11.773300</td>\n",
       "      <td>7.715812e-29</td>\n",
       "      <td>4.589528e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9033</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>3_2730</td>\n",
       "      <td>3</td>\n",
       "      <td>2730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>2730</td>\n",
       "      <td>53</td>\n",
       "      <td>30</td>\n",
       "      <td>15.113784</td>\n",
       "      <td>4.984762e-46</td>\n",
       "      <td>2.548200e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9989</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>3_551</td>\n",
       "      <td>3</td>\n",
       "      <td>551</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>551</td>\n",
       "      <td>1849</td>\n",
       "      <td>4</td>\n",
       "      <td>5.318958</td>\n",
       "      <td>9.999891e-01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>12428</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>2</td>\n",
       "      <td>4_2808</td>\n",
       "      <td>4</td>\n",
       "      <td>2808</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>2808</td>\n",
       "      <td>1729</td>\n",
       "      <td>220</td>\n",
       "      <td>20.956158</td>\n",
       "      <td>1.191537e-162</td>\n",
       "      <td>1.110223e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13229</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>4</td>\n",
       "      <td>4_801</td>\n",
       "      <td>4</td>\n",
       "      <td>801</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>801</td>\n",
       "      <td>225</td>\n",
       "      <td>146</td>\n",
       "      <td>31.615317</td>\n",
       "      <td>5.445274e-231</td>\n",
       "      <td>2.907737e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  numLin             p         q  n_est    gene  redundan  scaffold  \\\n",
       "0     48      12  9.999990e-07  0.000158      2  1_1040         1      1040   \n",
       "1    425      12  9.999990e-07  0.000158      3  1_1391         1      1391   \n",
       "2   3802      12  9.999990e-07  0.000158      2  1_4506         1      4506   \n",
       "3   5348      12  9.999990e-07  0.000158      5  2_1265         2      1265   \n",
       "4   6132      12  9.999990e-07  0.000158      6   2_199         2       199   \n",
       "5   7058      12  9.999990e-07  0.000158      6   2_897         2       897   \n",
       "6   9033      12  9.999990e-07  0.000158      2  3_2730         3      2730   \n",
       "7   9989      12  9.999990e-07  0.000158      2   3_551         3       551   \n",
       "8  12428      12  9.999990e-07  0.000158      2  4_2808         4      2808   \n",
       "9  13229      12  9.999990e-07  0.000158      4   4_801         4       801   \n",
       "\n",
       "   start  chrom   pos  SNPs  hits          Z  top_candidate_p        Z_pVal  \n",
       "0    NaN      1  1040    93    37  14.712225     6.834746e-49  2.455139e-06  \n",
       "1    NaN      1  1391    27    19  11.351576     2.056867e-32  6.736171e-08  \n",
       "2    NaN      1  4506   109     8  17.976827     1.550062e-05  1.090242e-07  \n",
       "3    NaN      2  1265    16     5   9.028052     3.984317e-07  2.119224e-06  \n",
       "4    NaN      2   199    10     9   9.224959     9.910000e-18  3.637041e-07  \n",
       "5    NaN      2   897    30    18  11.773300     7.715812e-29  4.589528e-08  \n",
       "6    NaN      3  2730    53    30  15.113784     4.984762e-46  2.548200e-09  \n",
       "7    NaN      3   551  1849     4   5.318958     9.999891e-01  0.000000e+00  \n",
       "8    NaN      4  2808  1729   220  20.956158    1.191537e-162  1.110223e-16  \n",
       "9    NaN      4   801   225   146  31.615317    5.445274e-231  2.907737e-10  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_linages_wza_picmin.merge(sign_wza_kendall, on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d271b89e-2dd6-4e15-8a8b-0aca9325c924",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['2_1265', '4_801']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "55140fc3-9fd9-407a-8910-55d02dc44782",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>SNPs_x</th>\n",
       "      <th>hits_x</th>\n",
       "      <th>Z_x</th>\n",
       "      <th>top_candidate_p_x</th>\n",
       "      <th>Z_pVal_x</th>\n",
       "      <th>index</th>\n",
       "      <th>numLin</th>\n",
       "      <th>p</th>\n",
       "      <th>q</th>\n",
       "      <th>...</th>\n",
       "      <th>redundan</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>start</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>SNPs_y</th>\n",
       "      <th>hits_y</th>\n",
       "      <th>Z_y</th>\n",
       "      <th>top_candidate_p_y</th>\n",
       "      <th>Z_pVal_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2_1265</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11.151806</td>\n",
       "      <td>1.585000e-29</td>\n",
       "      <td>1.569795e-08</td>\n",
       "      <td>5348</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1265</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>9.028052</td>\n",
       "      <td>3.984317e-07</td>\n",
       "      <td>2.119224e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4_801</td>\n",
       "      <td>225</td>\n",
       "      <td>130</td>\n",
       "      <td>23.960528</td>\n",
       "      <td>7.309601e-196</td>\n",
       "      <td>1.040595e-06</td>\n",
       "      <td>13229</td>\n",
       "      <td>12</td>\n",
       "      <td>9.999990e-07</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>801</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>801</td>\n",
       "      <td>225</td>\n",
       "      <td>146</td>\n",
       "      <td>31.615317</td>\n",
       "      <td>5.445274e-231</td>\n",
       "      <td>2.907737e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene  SNPs_x  hits_x        Z_x  top_candidate_p_x      Z_pVal_x  index  \\\n",
       "0  2_1265      16      15  11.151806       1.585000e-29  1.569795e-08   5348   \n",
       "1   4_801     225     130  23.960528      7.309601e-196  1.040595e-06  13229   \n",
       "\n",
       "   numLin             p         q  ...  redundan  scaffold  start  chrom  \\\n",
       "0      12  9.999990e-07  0.000158  ...         2      1265    NaN      2   \n",
       "1      12  9.999990e-07  0.000158  ...         4       801    NaN      4   \n",
       "\n",
       "    pos  SNPs_y  hits_y        Z_y  top_candidate_p_y      Z_pVal_y  \n",
       "0  1265      16       5   9.028052       3.984317e-07  2.119224e-06  \n",
       "1   801     225     146  31.615317      5.445274e-231  2.907737e-10  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_wza_lfmm.merge(sign_linages_wza_picmin, on = 'gene').merge(sign_wza_kendall, on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "56f9995b-fb0c-4289-9f81-19f13869edbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "gemma = pd.read_csv('/carnegie/nobackup/scratch/tbellagio/gea_grene-net/gwas/allele_assoc_runs/lmm_gemma/bio1/output/results_lmm.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "76a8ae23-cd80-40e2-bf2b-59f775d08d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gemma_sign_blocks = gemma.groupby('blocks')['significant'].sum()[gemma.groupby('blocks')['significant'].sum() !=0].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5023cff0-0033-486f-ac0f-da86e2707e92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>blocks</th>\n",
       "      <th>significant</th>\n",
       "      <th>gene</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [blocks, significant, gene, SNPs, hits, Z, top_candidate_p, Z_pVal]\n",
       "Index: []"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gemma_sign_blocks.merge(sign_wza_lfmm, left_on = 'blocks', right_on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fffe189b-095e-4db8-ae18-f18510a6d19d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>blocks</th>\n",
       "      <th>significant</th>\n",
       "      <th>index</th>\n",
       "      <th>numLin</th>\n",
       "      <th>p</th>\n",
       "      <th>q</th>\n",
       "      <th>n_est</th>\n",
       "      <th>gene</th>\n",
       "      <th>redundan</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>start</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [blocks, significant, index, numLin, p, q, n_est, gene, redundan, scaffold, start, chrom, pos]\n",
       "Index: []"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gemma_sign_blocks.merge(sign_linages_wza_picmin, left_on = 'blocks', right_on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "bb864cfb-71ef-4ae8-bebf-5fc11a13d3fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>blocks</th>\n",
       "      <th>significant</th>\n",
       "      <th>gene</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>hits</th>\n",
       "      <th>Z</th>\n",
       "      <th>top_candidate_p</th>\n",
       "      <th>Z_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2_973</td>\n",
       "      <td>4</td>\n",
       "      <td>2_973</td>\n",
       "      <td>1843</td>\n",
       "      <td>5</td>\n",
       "      <td>10.919003</td>\n",
       "      <td>0.999944</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  blocks  significant   gene  SNPs  hits          Z  top_candidate_p  Z_pVal\n",
       "0  2_973            4  2_973  1843     5  10.919003         0.999944     0.0"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gemma_sign_blocks.merge(sign_wza_kendall, left_on = 'blocks', right_on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87e4a12e-737b-44de-a88b-ac8d979e6b83",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e49b9cf-66b3-4028-8b89-188b19ed1b7a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7120e76-1e36-4f28-813c-c0fd87e47a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### addd to the annotated genes the blocks "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8baa044d-b767-4e01-96ce-4120a6c14ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "dict_blocks = '../key_files/blocks_snpsid_dict.pkl'\n",
    "\n",
    "with open(dict_blocks, 'rb') as file:\n",
    "    dict_blocks = pickle.load(file)\n",
    "\n",
    "reverse_mapping = {item: key for key, values in dict_blocks.items() for item in values}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ed16113a-bd63-4afd-a7aa-2d00a0026168",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff3_formatted = pd.read_csv('../key_files/TAIR10_GFF3_genes_transposons_formatted4topr.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5cae79a3-591d-4463-a1e3-97bcb5183f25",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "3beeff58-5be1-4151-a7d5-354617fc4989",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create clock ranges\n",
    "blocks_ranges = {}\n",
    "for key, values in dict_blocks.items():\n",
    "    chromosome = dict_blocks[key][0].split('_')[0]\n",
    "    start = dict_blocks[key][0].split('_')[1]\n",
    "    end = dict_blocks[key][-1].split('_')[1]\n",
    "    blocks_ranges[key] = [int(chromosome), int(start), int(end)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "15e777fb-5ad8-49c2-8ed7-5c760d73c6f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Step 2: Function to check if a gene falls within one or more blocks\n",
    "def find_gene_blocks(chrom, gene_start, gene_end):\n",
    "    matching_blocks = []\n",
    "    for block, values in blocks_ranges.items():\n",
    "        block_chrom, block_start, block_end = values\n",
    "        if chrom == block_chrom:\n",
    "            # Check if the gene overlaps with the block\n",
    "            if (block_start <= gene_start <= block_end) or (block_start <= gene_end <= block_end) or (gene_start <= block_start <= gene_end):\n",
    "                matching_blocks.append(block)\n",
    "    return matching_blocks if matching_blocks else 'No matching block'\n",
    "\n",
    "# Step 3: Apply the function to the dataframe\n",
    "\n",
    "gff3_formatted['blocks'] = gff3_formatted.apply(lambda row: find_gene_blocks(row['chrom'], row['gene_start'], row['gene_end']), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "7216b446-29c2-40b8-a0fe-596e51f74ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff3_formatted.to_csv('../key_files/TAIR10_GFF3_genes_transposons_formatted4topr_w_blocks',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "ea0d9023-0058-46ae-86c0-f239b40f2920",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>biotype</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>exon_chromstart</th>\n",
       "      <th>exon_chromend</th>\n",
       "      <th>blocks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3631</td>\n",
       "      <td>5899</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>3631,3996,4486,4706,5174,5439</td>\n",
       "      <td>3913,4276,4605,5095,5326,5899</td>\n",
       "      <td>[1_0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5928</td>\n",
       "      <td>8737</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>8571,8417,8236,7942,7762,7564,7384,7157,6437,5...</td>\n",
       "      <td>8737,8464,8325,7987,7835,7649,7450,7232,7069,6...</td>\n",
       "      <td>[1_0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>11649</td>\n",
       "      <td>13714</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>13335,11649</td>\n",
       "      <td>13714,13173</td>\n",
       "      <td>[1_0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>23146</td>\n",
       "      <td>31227</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>23146,24542,24752,25041,25524,25825,26081,2629...</td>\n",
       "      <td>24451,24655,24962,25435,25743,25997,26203,2645...</td>\n",
       "      <td>[1_0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>31170</td>\n",
       "      <td>33153</td>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>33029,32547,32431,32282,32088,31933,31693,3152...</td>\n",
       "      <td>33153,32670,32459,32347,32195,31998,31813,3160...</td>\n",
       "      <td>[1_0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27201</th>\n",
       "      <td>207189</td>\n",
       "      <td>5</td>\n",
       "      <td>26959569</td>\n",
       "      <td>26960323</td>\n",
       "      <td>AT5G67600</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67600</td>\n",
       "      <td>26960181,26959937,26959569</td>\n",
       "      <td>26960323,26960093,26959799</td>\n",
       "      <td>[5_3109]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27202</th>\n",
       "      <td>207192</td>\n",
       "      <td>5</td>\n",
       "      <td>26960931</td>\n",
       "      <td>26963638</td>\n",
       "      <td>AT5G67610</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67610</td>\n",
       "      <td>26963509,26963122,26962863,26962648,26962503,2...</td>\n",
       "      <td>26963638,26963335,26963020,26962793,26962562,2...</td>\n",
       "      <td>[5_3109]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27203</th>\n",
       "      <td>207212</td>\n",
       "      <td>5</td>\n",
       "      <td>26964770</td>\n",
       "      <td>26965996</td>\n",
       "      <td>AT5G67620</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67620</td>\n",
       "      <td>26965468,26965198,26964770</td>\n",
       "      <td>26965996,26965379,26965004</td>\n",
       "      <td>[5_3109]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27204</th>\n",
       "      <td>207215</td>\n",
       "      <td>5</td>\n",
       "      <td>26967381</td>\n",
       "      <td>26969394</td>\n",
       "      <td>AT5G67630</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67630</td>\n",
       "      <td>26968558,26967381</td>\n",
       "      <td>26969394,26968195</td>\n",
       "      <td>[5_3109]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27205</th>\n",
       "      <td>207217</td>\n",
       "      <td>5</td>\n",
       "      <td>26969546</td>\n",
       "      <td>26970641</td>\n",
       "      <td>AT5G67640</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>AT5G67640</td>\n",
       "      <td>26970444,26970312,26969915,26969546</td>\n",
       "      <td>26970641,26970360,26970241,26969670</td>\n",
       "      <td>[5_3109]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27206 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  chrom  gene_start  gene_end gene_symbol  \\\n",
       "0               0      1        3631      5899   AT1G01010   \n",
       "1               6      1        5928      8737   AT1G01020   \n",
       "2              24      1       11649     13714   AT1G01030   \n",
       "3              26      1       23146     31227   AT1G01040   \n",
       "4              66      1       31170     33153   AT1G01050   \n",
       "...           ...    ...         ...       ...         ...   \n",
       "27201      207189      5    26959569  26960323   AT5G67600   \n",
       "27202      207192      5    26960931  26963638   AT5G67610   \n",
       "27203      207212      5    26964770  26965996   AT5G67620   \n",
       "27204      207215      5    26967381  26969394   AT5G67630   \n",
       "27205      207217      5    26969546  26970641   AT5G67640   \n",
       "\n",
       "                   biotype    gene_id  \\\n",
       "0      protein_coding_gene  AT1G01010   \n",
       "1      protein_coding_gene  AT1G01020   \n",
       "2      protein_coding_gene  AT1G01030   \n",
       "3      protein_coding_gene  AT1G01040   \n",
       "4      protein_coding_gene  AT1G01050   \n",
       "...                    ...        ...   \n",
       "27201  protein_coding_gene  AT5G67600   \n",
       "27202  protein_coding_gene  AT5G67610   \n",
       "27203  protein_coding_gene  AT5G67620   \n",
       "27204  protein_coding_gene  AT5G67630   \n",
       "27205  protein_coding_gene  AT5G67640   \n",
       "\n",
       "                                         exon_chromstart  \\\n",
       "0                          3631,3996,4486,4706,5174,5439   \n",
       "1      8571,8417,8236,7942,7762,7564,7384,7157,6437,5...   \n",
       "2                                            13335,11649   \n",
       "3      23146,24542,24752,25041,25524,25825,26081,2629...   \n",
       "4      33029,32547,32431,32282,32088,31933,31693,3152...   \n",
       "...                                                  ...   \n",
       "27201                         26960181,26959937,26959569   \n",
       "27202  26963509,26963122,26962863,26962648,26962503,2...   \n",
       "27203                         26965468,26965198,26964770   \n",
       "27204                                  26968558,26967381   \n",
       "27205                26970444,26970312,26969915,26969546   \n",
       "\n",
       "                                           exon_chromend    blocks  \n",
       "0                          3913,4276,4605,5095,5326,5899     [1_0]  \n",
       "1      8737,8464,8325,7987,7835,7649,7450,7232,7069,6...     [1_0]  \n",
       "2                                            13714,13173     [1_0]  \n",
       "3      24451,24655,24962,25435,25743,25997,26203,2645...     [1_0]  \n",
       "4      33153,32670,32459,32347,32195,31998,31813,3160...     [1_0]  \n",
       "...                                                  ...       ...  \n",
       "27201                         26960323,26960093,26959799  [5_3109]  \n",
       "27202  26963638,26963335,26963020,26962793,26962562,2...  [5_3109]  \n",
       "27203                         26965996,26965379,26965004  [5_3109]  \n",
       "27204                                  26969394,26968195  [5_3109]  \n",
       "27205                26970641,26970360,26970241,26969670  [5_3109]  \n",
       "\n",
       "[27206 rows x 10 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff3_formatted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "961bf95a-88e1-4b50-a124-d9cedc28c823",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (pipeline_snakemake)",
   "language": "python",
   "name": "pipeline_snakemake"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
